<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
    <meta http-equiv="x-ua-compatible" content="IE=Edge"/>
    <style type="text/css">
        body {
            padding-bottom: 30px;
            margin: 0;
            padding: 0;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/table_sorter/style.css" />

    <script type="text/javascript">
        window.reporting = {};
        window.console = {log: function() {}};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <!--<script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>-->
    <!--<script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>-->
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <!--<script type="text/javascript" src="html_aux/jquery.align-column.js"></script>-->

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>

    <script type="text/javascript" src="html_aux/dragtable.js"></script>
    <script type="text/javascript" src="html_aux/table_sorter/tsort.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false });

            $('table.tableSorter').tableSort();
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for freebayes</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"report":{"metric_storage":{"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"general_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"sample_reports":[{"metric_storage":{"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"general_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-2","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-2/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-2/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":2156,"average":2156.0,"cosmic":2156,"dbsnp":2156},"known":{"1000genomes":380,"average":380.0,"cosmic":380,"dbsnp":380},"novel":{"1000genomes":1776,"average":1776.0,"cosmic":1776,"dbsnp":1776}},"value":2156,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":2156,"average":2156.0,"cosmic":2156,"dbsnp":2156},"known":{"1000genomes":380,"average":380.0,"cosmic":380,"dbsnp":380},"novel":{"1000genomes":1776,"average":1776.0,"cosmic":1776,"dbsnp":1776}},"value":2156,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":15,"average":170.66666666666666,"cosmic":117,"dbsnp":380},"known":{"1000genomes":15,"average":154.33333333333334,"cosmic":68,"dbsnp":380},"novel":{"1000genomes":0,"average":16.333333333333332,"cosmic":49,"dbsnp":0}},"value":117,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":0.7,"average":7.919999999999999,"cosmic":5.43,"dbsnp":17.63},"known":{"1000genomes":3.95,"average":40.61333333333334,"cosmic":17.89,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.9199999999999999,"cosmic":2.76,"dbsnp":0.0}},"value":5.43,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":14,"average":137.33333333333334,"cosmic":83,"dbsnp":315},"known":{"1000genomes":14,"average":130.0,"cosmic":61,"dbsnp":315},"novel":{"1000genomes":0,"average":7.333333333333333,"cosmic":22,"dbsnp":0}},"value":83,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":93.33,"average":82.38666666666667,"cosmic":70.94,"dbsnp":82.89},"known":{"1000genomes":93.33,"average":88.64333333333333,"cosmic":89.71,"dbsnp":82.89},"novel":{"1000genomes":0.0,"average":14.966666666666667,"cosmic":44.9,"dbsnp":0.0}},"value":70.94,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":6.9e-07,"average":6.899999999999999e-07,"cosmic":6.9e-07,"dbsnp":6.9e-07},"known":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07},"novel":{"1000genomes":5.7e-07,"average":5.7e-07,"cosmic":5.7e-07,"dbsnp":5.7e-07}},"value":6.9e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":1455084.0,"average":1455084.0,"cosmic":1455084.0,"dbsnp":1455084.0},"known":{"1000genomes":8255687.0,"average":8255687.0,"cosmic":8255687.0,"dbsnp":8255687.0},"novel":{"1000genomes":1766419.0,"average":1766419.0,"cosmic":1766419.0,"dbsnp":1766419.0}},"value":1455084.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":34.55,"average":34.55,"cosmic":34.55,"dbsnp":34.55},"known":{"1000genomes":7.02,"average":7.02,"cosmic":7.02,"dbsnp":7.02},"novel":{"1000genomes":196.67,"average":196.67,"cosmic":196.67,"dbsnp":196.67}},"value":34.55,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.16,"average":0.16,"cosmic":0.16,"dbsnp":0.16},"known":{"1000genomes":1.04,"average":1.04,"cosmic":1.04,"dbsnp":1.04},"novel":{"1000genomes":0.06,"average":0.06,"cosmic":0.06,"dbsnp":0.06}},"value":0.16,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":{"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"general_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-8","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-8","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-8","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-8","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-8/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-8","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-8/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":1136,"average":1136.0,"cosmic":1136,"dbsnp":1136},"known":{"1000genomes":436,"average":436.0,"cosmic":436,"dbsnp":436},"novel":{"1000genomes":700,"average":700.0,"cosmic":700,"dbsnp":700}},"value":1136,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":1136,"average":1136.0,"cosmic":1136,"dbsnp":1136},"known":{"1000genomes":436,"average":436.0,"cosmic":436,"dbsnp":436},"novel":{"1000genomes":700,"average":700.0,"cosmic":700,"dbsnp":700}},"value":1136,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":19,"average":185.33333333333334,"cosmic":101,"dbsnp":436},"known":{"1000genomes":19,"average":179.33333333333334,"cosmic":83,"dbsnp":436},"novel":{"1000genomes":0,"average":6.0,"cosmic":18,"dbsnp":0}},"value":101,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":1.67,"average":16.313333333333336,"cosmic":8.89,"dbsnp":38.38},"known":{"1000genomes":4.36,"average":41.13333333333333,"cosmic":19.04,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.8566666666666666,"cosmic":2.57,"dbsnp":0.0}},"value":8.89,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":19,"average":176.0,"cosmic":90,"dbsnp":419},"known":{"1000genomes":19,"average":172.0,"cosmic":78,"dbsnp":419},"novel":{"1000genomes":0,"average":4.0,"cosmic":12,"dbsnp":0}},"value":90,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":100.0,"average":95.07,"cosmic":89.11,"dbsnp":96.1},"known":{"1000genomes":100.0,"average":96.69333333333333,"cosmic":93.98,"dbsnp":96.1},"novel":{"1000genomes":0.0,"average":22.223333333333333,"cosmic":66.67,"dbsnp":0.0}},"value":89.11,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.6e-07,"average":3.6e-07,"cosmic":3.6e-07,"dbsnp":3.6e-07},"known":{"1000genomes":1.4e-07,"average":1.4e-07,"cosmic":1.4e-07,"dbsnp":1.4e-07},"novel":{"1000genomes":2.2e-07,"average":2.2e-07,"cosmic":2.2e-07,"dbsnp":2.2e-07}},"value":3.6e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":2761585.0,"average":2761585.0,"cosmic":2761585.0,"dbsnp":2761585.0},"known":{"1000genomes":7195324.0,"average":7195324.0,"cosmic":7195324.0,"dbsnp":7195324.0},"novel":{"1000genomes":4481658.0,"average":4481658.0,"cosmic":4481658.0,"dbsnp":4481658.0}},"value":2761585.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":28.85,"average":28.850000000000005,"cosmic":28.85,"dbsnp":28.85},"known":{"1000genomes":14.9,"average":14.9,"cosmic":14.9,"dbsnp":14.9},"novel":{"1000genomes":69.3,"average":69.3,"cosmic":69.3,"dbsnp":69.3}},"value":28.85,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.45,"average":0.45,"cosmic":0.45,"dbsnp":0.45},"known":{"1000genomes":1.58,"average":1.58,"cosmic":1.58,"dbsnp":1.58},"novel":{"1000genomes":0.14,"average":0.14,"cosmic":0.14,"dbsnp":0.14}},"value":0.45,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":{"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"general_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Sensitive-10","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-10","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-10","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-10","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-10/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-10","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-10/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":1564,"average":1564.0,"cosmic":1564,"dbsnp":1564},"known":{"1000genomes":1203,"average":1203.0,"cosmic":1203,"dbsnp":1203},"novel":{"1000genomes":361,"average":361.0,"cosmic":361,"dbsnp":361}},"value":1564,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":1433,"average":1433.0,"cosmic":1433,"dbsnp":1433},"known":{"1000genomes":1119,"average":1119.0,"cosmic":1119,"dbsnp":1119},"novel":{"1000genomes":314,"average":314.0,"cosmic":314,"dbsnp":314}},"value":1433,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":52,"average":52.0,"cosmic":52,"dbsnp":52},"known":{"1000genomes":45,"average":45.0,"cosmic":45,"dbsnp":45},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":52,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":60,"average":60.0,"cosmic":60,"dbsnp":60},"known":{"1000genomes":39,"average":39.0,"cosmic":39,"dbsnp":39},"novel":{"1000genomes":21,"average":21.0,"cosmic":21,"dbsnp":21}},"value":60,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":514,"average":624.3333333333334,"cosmic":156,"dbsnp":1203},"known":{"1000genomes":514,"average":619.3333333333334,"cosmic":141,"dbsnp":1203},"novel":{"1000genomes":0,"average":5.0,"cosmic":15,"dbsnp":0}},"value":156,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":32.86,"average":39.916666666666664,"cosmic":9.97,"dbsnp":76.92},"known":{"1000genomes":42.73,"average":51.48333333333333,"cosmic":11.72,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.3866666666666667,"cosmic":4.16,"dbsnp":0.0}},"value":9.97,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":513,"average":614.6666666666666,"cosmic":147,"dbsnp":1184},"known":{"1000genomes":513,"average":611.3333333333334,"cosmic":137,"dbsnp":1184},"novel":{"1000genomes":0,"average":3.3333333333333335,"cosmic":10,"dbsnp":0}},"value":147,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.81,"average":97.48666666666668,"cosmic":94.23,"dbsnp":98.42},"known":{"1000genomes":99.81,"average":98.46333333333332,"cosmic":97.16,"dbsnp":98.42},"novel":{"1000genomes":0.0,"average":22.223333333333333,"cosmic":66.67,"dbsnp":0.0}},"value":94.23,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":5e-07,"average":5e-07,"cosmic":5e-07,"dbsnp":5e-07},"known":{"1000genomes":3.8e-07,"average":3.8e-07,"cosmic":3.8e-07,"dbsnp":3.8e-07},"novel":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07}},"value":5e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":2005857.0,"average":2005857.0,"cosmic":2005857.0,"dbsnp":2005857.0},"known":{"1000genomes":2607781.0,"average":2607781.0,"cosmic":2607781.0,"dbsnp":2607781.0},"novel":{"1000genomes":8690197.0,"average":8690197.0,"cosmic":8690197.0,"dbsnp":8690197.0}},"value":2005857.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":44.93,"average":44.93,"cosmic":44.93,"dbsnp":44.93},"known":{"1000genomes":42.48,"average":42.48,"cosmic":42.48,"dbsnp":42.48},"novel":{"1000genomes":67.0,"average":67.0,"cosmic":67.0,"dbsnp":67.0}},"value":44.93,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":1.55,"average":1.55,"cosmic":1.55,"dbsnp":1.55},"known":{"1000genomes":2.45,"average":2.45,"cosmic":2.45,"dbsnp":2.45},"novel":{"1000genomes":0.31,"average":0.31,"cosmic":0.31,"dbsnp":0.31}},"value":1.55,"html_fpath":null}],"plots":[],"html_fpath":null}],"name":""},"data_outside_reports":{},"type_":null,"date":"14 November 2014, Friday, 11:37:44"}
            </div>
        </div>
    </div>
</div>
</body>
</html>
