<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
    <meta http-equiv="x-ua-compatible" content="IE=Edge"/>
    <style type="text/css">
        body {
            padding-bottom: 30px;
            margin: 0;
            padding: 0;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/table_sorter/style.css" />

    <script type="text/javascript">
        window.reporting = {};
        window.console = {log: function() {}};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <!--<script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>-->
    <!--<script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>-->
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <!--<script type="text/javascript" src="html_aux/jquery.align-column.js"></script>-->

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>

    <script type="text/javascript" src="html_aux/dragtable.js"></script>
    <script type="text/javascript" src="html_aux/table_sorter/tsort.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false });

            $('table.tableSorter').tableSort();
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"report":{"metric_storage":{"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"general_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"sample_reports":[{"metric_storage":null,"display_name":"Tissue-Resistant-1 mutect","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-1","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-1","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-1","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-1/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-1","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-1/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":183906,"average":183906.0,"cosmic":183906,"dbsnp":183906},"known":{"1000genomes":10649,"average":10649.0,"cosmic":10649,"dbsnp":10649},"novel":{"1000genomes":173257,"average":173257.0,"cosmic":173257,"dbsnp":173257}},"value":183906,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":183906,"average":183906.0,"cosmic":183906,"dbsnp":183906},"known":{"1000genomes":10649,"average":10649.0,"cosmic":10649,"dbsnp":10649},"novel":{"1000genomes":173257,"average":173257.0,"cosmic":173257,"dbsnp":173257}},"value":183906,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":126,"average":6470.333333333333,"cosmic":8636,"dbsnp":10649},"known":{"1000genomes":126,"average":3955.0,"cosmic":1090,"dbsnp":10649},"novel":{"1000genomes":0,"average":2515.3333333333335,"cosmic":7546,"dbsnp":0}},"value":8636,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":0.07,"average":3.52,"cosmic":4.7,"dbsnp":5.79},"known":{"1000genomes":1.18,"average":37.14,"cosmic":10.24,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4533333333333334,"cosmic":4.36,"dbsnp":0.0}},"value":4.7,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":13,"average":1296.3333333333333,"cosmic":2658,"dbsnp":1218},"known":{"1000genomes":13,"average":475.3333333333333,"cosmic":195,"dbsnp":1218},"novel":{"1000genomes":0,"average":821.0,"cosmic":2463,"dbsnp":0}},"value":2658,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":10.32,"average":17.513333333333332,"cosmic":30.78,"dbsnp":11.44},"known":{"1000genomes":10.32,"average":13.216666666666667,"cosmic":17.89,"dbsnp":11.44},"novel":{"1000genomes":0.0,"average":10.88,"cosmic":32.64,"dbsnp":0.0}},"value":30.78,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":5.862e-05,"average":5.862e-05,"cosmic":5.862e-05,"dbsnp":5.862e-05},"known":{"1000genomes":3.39e-06,"average":3.39e-06,"cosmic":3.39e-06,"dbsnp":3.39e-06},"novel":{"1000genomes":5.523e-05,"average":5.523e-05,"cosmic":5.523e-05,"dbsnp":5.523e-05}},"value":5.862e-05,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":17058.0,"average":17058.0,"cosmic":17058.0,"dbsnp":17058.0},"known":{"1000genomes":294596.0,"average":294596.0,"cosmic":294596.0,"dbsnp":294596.0},"novel":{"1000genomes":18106.0,"average":18106.0,"cosmic":18106.0,"dbsnp":18106.0}},"value":17058.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":183906.0,"average":183906.0,"cosmic":183906.0,"dbsnp":183906.0},"known":{"1000genomes":10649.0,"average":10649.0,"cosmic":10649.0,"dbsnp":10649.0},"novel":{"1000genomes":173257.0,"average":173257.0,"cosmic":173257.0,"dbsnp":173257.0}},"value":183906.0,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"known":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-1 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-1","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-1","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-1","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-1/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-1","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-1/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":20985,"average":20985.0,"cosmic":20985,"dbsnp":20985},"known":{"1000genomes":9408,"average":9408.0,"cosmic":9408,"dbsnp":9408},"novel":{"1000genomes":11577,"average":11577.0,"cosmic":11577,"dbsnp":11577}},"value":20985,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":20468,"average":20468.0,"cosmic":20468,"dbsnp":20468},"known":{"1000genomes":8984,"average":8984.0,"cosmic":8984,"dbsnp":8984},"novel":{"1000genomes":11484,"average":11484.0,"cosmic":11484,"dbsnp":11484}},"value":20468,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":202,"average":202.0,"cosmic":202,"dbsnp":202},"known":{"1000genomes":187,"average":187.0,"cosmic":187,"dbsnp":187},"novel":{"1000genomes":15,"average":15.0,"cosmic":15,"dbsnp":15}},"value":202,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":264,"average":264.0,"cosmic":264,"dbsnp":264},"known":{"1000genomes":229,"average":229.0,"cosmic":229,"dbsnp":229},"novel":{"1000genomes":35,"average":35.0,"cosmic":35,"dbsnp":35}},"value":264,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":4272,"average":5084.666666666667,"cosmic":1574,"dbsnp":9408},"known":{"1000genomes":4265,"average":4898.333333333333,"cosmic":1022,"dbsnp":9408},"novel":{"1000genomes":7,"average":186.33333333333334,"cosmic":552,"dbsnp":0}},"value":1574,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":20.36,"average":24.23,"cosmic":7.5,"dbsnp":44.83},"known":{"1000genomes":45.33,"average":52.06333333333333,"cosmic":10.86,"dbsnp":100.0},"novel":{"1000genomes":0.06,"average":1.6099999999999997,"cosmic":4.77,"dbsnp":0.0}},"value":7.5,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":4246,"average":4695.666666666667,"cosmic":1110,"dbsnp":8731},"known":{"1000genomes":4239,"average":4630.333333333333,"cosmic":921,"dbsnp":8731},"novel":{"1000genomes":7,"average":65.33333333333333,"cosmic":189,"dbsnp":0}},"value":1110,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.39,"average":87.57,"cosmic":70.52,"dbsnp":92.8},"known":{"1000genomes":99.39,"average":94.10333333333334,"cosmic":90.12,"dbsnp":92.8},"novel":{"1000genomes":100.0,"average":44.74666666666667,"cosmic":34.24,"dbsnp":0.0}},"value":70.52,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":6.69e-06,"average":6.690000000000001e-06,"cosmic":6.69e-06,"dbsnp":6.69e-06},"known":{"1000genomes":3e-06,"average":3e-06,"cosmic":3e-06,"dbsnp":3e-06},"novel":{"1000genomes":3.69e-06,"average":3.69e-06,"cosmic":3.69e-06,"dbsnp":3.69e-06}},"value":6.69e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":149495.0,"average":149495.0,"cosmic":149495.0,"dbsnp":149495.0},"known":{"1000genomes":333456.0,"average":333456.0,"cosmic":333456.0,"dbsnp":333456.0},"novel":{"1000genomes":270982.0,"average":270982.0,"cosmic":270982.0,"dbsnp":270982.0}},"value":149495.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":7.1,"average":7.099999999999999,"cosmic":7.1,"dbsnp":7.1},"known":{"1000genomes":4.01,"average":4.01,"cosmic":4.01,"dbsnp":4.01},"novel":{"1000genomes":144.33,"average":144.33,"cosmic":144.33,"dbsnp":144.33}},"value":7.1,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.44,"average":0.44,"cosmic":0.44,"dbsnp":0.44},"known":{"1000genomes":2.08,"average":2.08,"cosmic":2.08,"dbsnp":2.08},"novel":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01}},"value":0.44,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-2 freebayes","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-2/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-2/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":2156,"average":2156.0,"cosmic":2156,"dbsnp":2156},"known":{"1000genomes":380,"average":380.0,"cosmic":380,"dbsnp":380},"novel":{"1000genomes":1776,"average":1776.0,"cosmic":1776,"dbsnp":1776}},"value":2156,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":2156,"average":2156.0,"cosmic":2156,"dbsnp":2156},"known":{"1000genomes":380,"average":380.0,"cosmic":380,"dbsnp":380},"novel":{"1000genomes":1776,"average":1776.0,"cosmic":1776,"dbsnp":1776}},"value":2156,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":15,"average":170.66666666666666,"cosmic":117,"dbsnp":380},"known":{"1000genomes":15,"average":154.33333333333334,"cosmic":68,"dbsnp":380},"novel":{"1000genomes":0,"average":16.333333333333332,"cosmic":49,"dbsnp":0}},"value":117,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":0.7,"average":7.919999999999999,"cosmic":5.43,"dbsnp":17.63},"known":{"1000genomes":3.95,"average":40.61333333333334,"cosmic":17.89,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.9199999999999999,"cosmic":2.76,"dbsnp":0.0}},"value":5.43,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":14,"average":137.33333333333334,"cosmic":83,"dbsnp":315},"known":{"1000genomes":14,"average":130.0,"cosmic":61,"dbsnp":315},"novel":{"1000genomes":0,"average":7.333333333333333,"cosmic":22,"dbsnp":0}},"value":83,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":93.33,"average":82.38666666666667,"cosmic":70.94,"dbsnp":82.89},"known":{"1000genomes":93.33,"average":88.64333333333333,"cosmic":89.71,"dbsnp":82.89},"novel":{"1000genomes":0.0,"average":14.966666666666667,"cosmic":44.9,"dbsnp":0.0}},"value":70.94,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":6.9e-07,"average":6.899999999999999e-07,"cosmic":6.9e-07,"dbsnp":6.9e-07},"known":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07},"novel":{"1000genomes":5.7e-07,"average":5.7e-07,"cosmic":5.7e-07,"dbsnp":5.7e-07}},"value":6.9e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":1455084.0,"average":1455084.0,"cosmic":1455084.0,"dbsnp":1455084.0},"known":{"1000genomes":8255687.0,"average":8255687.0,"cosmic":8255687.0,"dbsnp":8255687.0},"novel":{"1000genomes":1766419.0,"average":1766419.0,"cosmic":1766419.0,"dbsnp":1766419.0}},"value":1455084.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":34.55,"average":34.55,"cosmic":34.55,"dbsnp":34.55},"known":{"1000genomes":7.02,"average":7.02,"cosmic":7.02,"dbsnp":7.02},"novel":{"1000genomes":196.67,"average":196.67,"cosmic":196.67,"dbsnp":196.67}},"value":34.55,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.16,"average":0.16,"cosmic":0.16,"dbsnp":0.16},"known":{"1000genomes":1.04,"average":1.04,"cosmic":1.04,"dbsnp":1.04},"novel":{"1000genomes":0.06,"average":0.06,"cosmic":0.06,"dbsnp":0.06}},"value":0.16,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-2 mutect","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-2/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-2/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":113193,"average":113193.0,"cosmic":113193,"dbsnp":113193},"known":{"1000genomes":6511,"average":6511.0,"cosmic":6511,"dbsnp":6511},"novel":{"1000genomes":106682,"average":106682.0,"cosmic":106682,"dbsnp":106682}},"value":113193,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":113193,"average":113193.0,"cosmic":113193,"dbsnp":113193},"known":{"1000genomes":6511,"average":6511.0,"cosmic":6511,"dbsnp":6511},"novel":{"1000genomes":106682,"average":106682.0,"cosmic":106682,"dbsnp":106682}},"value":113193,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":64,"average":3903.0,"cosmic":5134,"dbsnp":6511},"known":{"1000genomes":64,"average":2404.6666666666665,"cosmic":639,"dbsnp":6511},"novel":{"1000genomes":0,"average":1498.3333333333333,"cosmic":4495,"dbsnp":0}},"value":5134,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":0.06,"average":3.4499999999999997,"cosmic":4.54,"dbsnp":5.75},"known":{"1000genomes":0.98,"average":36.93,"cosmic":9.81,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4033333333333333,"cosmic":4.21,"dbsnp":0.0}},"value":4.54,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":8,"average":780.6666666666666,"cosmic":1604,"dbsnp":730},"known":{"1000genomes":8,"average":288.3333333333333,"cosmic":127,"dbsnp":730},"novel":{"1000genomes":0,"average":492.3333333333333,"cosmic":1477,"dbsnp":0}},"value":1604,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":12.5,"average":18.316666666666666,"cosmic":31.24,"dbsnp":11.21},"known":{"1000genomes":12.5,"average":14.526666666666666,"cosmic":19.87,"dbsnp":11.21},"novel":{"1000genomes":0.0,"average":10.953333333333333,"cosmic":32.86,"dbsnp":0.0}},"value":31.24,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.608e-05,"average":3.608e-05,"cosmic":3.608e-05,"dbsnp":3.608e-05},"known":{"1000genomes":2.08e-06,"average":2.08e-06,"cosmic":2.08e-06,"dbsnp":2.08e-06},"novel":{"1000genomes":3.401e-05,"average":3.401e-05,"cosmic":3.401e-05,"dbsnp":3.401e-05}},"value":3.608e-05,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":27715.0,"average":27715.0,"cosmic":27715.0,"dbsnp":27715.0},"known":{"1000genomes":481824.0,"average":481824.0,"cosmic":481824.0,"dbsnp":481824.0},"novel":{"1000genomes":29406.0,"average":29406.0,"cosmic":29406.0,"dbsnp":29406.0}},"value":27715.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":113193.0,"average":113193.0,"cosmic":113193.0,"dbsnp":113193.0},"known":{"1000genomes":6511.0,"average":6511.0,"cosmic":6511.0,"dbsnp":6511.0},"novel":{"1000genomes":106682.0,"average":106682.0,"cosmic":106682.0,"dbsnp":106682.0}},"value":113193.0,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"known":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-2 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-2/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-2/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":18848,"average":18848.0,"cosmic":18848,"dbsnp":18848},"known":{"1000genomes":3778,"average":3778.0,"cosmic":3778,"dbsnp":3778},"novel":{"1000genomes":15070,"average":15070.0,"cosmic":15070,"dbsnp":15070}},"value":18848,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":18651,"average":18651.0,"cosmic":18651,"dbsnp":18651},"known":{"1000genomes":3630,"average":3630.0,"cosmic":3630,"dbsnp":3630},"novel":{"1000genomes":15021,"average":15021.0,"cosmic":15021,"dbsnp":15021}},"value":18651,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":83,"average":83.0,"cosmic":83,"dbsnp":83},"known":{"1000genomes":71,"average":71.0,"cosmic":71,"dbsnp":71},"novel":{"1000genomes":12,"average":12.0,"cosmic":12,"dbsnp":12}},"value":83,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":93,"average":93.0,"cosmic":93,"dbsnp":93},"known":{"1000genomes":73,"average":73.0,"cosmic":73,"dbsnp":73},"novel":{"1000genomes":20,"average":20.0,"cosmic":20,"dbsnp":20}},"value":93,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":1402,"average":2095.3333333333335,"cosmic":1106,"dbsnp":3778},"known":{"1000genomes":1402,"average":1861.3333333333333,"cosmic":404,"dbsnp":3778},"novel":{"1000genomes":0,"average":234.0,"cosmic":702,"dbsnp":0}},"value":1106,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":7.44,"average":11.116666666666667,"cosmic":5.87,"dbsnp":20.04},"known":{"1000genomes":37.11,"average":49.26666666666667,"cosmic":10.69,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.5533333333333335,"cosmic":4.66,"dbsnp":0.0}},"value":5.87,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":1379,"average":1641.6666666666667,"cosmic":592,"dbsnp":2954},"known":{"1000genomes":1379,"average":1553.0,"cosmic":326,"dbsnp":2954},"novel":{"1000genomes":0,"average":88.66666666666667,"cosmic":266,"dbsnp":0}},"value":592,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":98.36,"average":76.69333333333334,"cosmic":53.53,"dbsnp":78.19},"known":{"1000genomes":98.36,"average":85.74666666666667,"cosmic":80.69,"dbsnp":78.19},"novel":{"1000genomes":0.0,"average":12.63,"cosmic":37.89,"dbsnp":0.0}},"value":53.53,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":6.01e-06,"average":6.010000000000001e-06,"cosmic":6.01e-06,"dbsnp":6.01e-06},"known":{"1000genomes":1.2e-06,"average":1.2e-06,"cosmic":1.2e-06,"dbsnp":1.2e-06},"novel":{"1000genomes":4.8e-06,"average":4.8e-06,"cosmic":4.8e-06,"dbsnp":4.8e-06}},"value":6.01e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":166445.0,"average":166445.0,"cosmic":166445.0,"dbsnp":166445.0},"known":{"1000genomes":830376.0,"average":830376.0,"cosmic":830376.0,"dbsnp":830376.0},"novel":{"1000genomes":208172.0,"average":208172.0,"cosmic":208172.0,"dbsnp":208172.0}},"value":166445.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":171.67,"average":171.67,"cosmic":171.67,"dbsnp":171.67},"known":{"1000genomes":56.95,"average":56.95000000000001,"cosmic":56.95,"dbsnp":56.95},"novel":{"1000genomes":1261.5,"average":1261.5,"cosmic":1261.5,"dbsnp":1261.5}},"value":171.67,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.13,"average":0.13,"cosmic":0.13,"dbsnp":0.13},"known":{"1000genomes":1.24,"average":1.24,"cosmic":1.24,"dbsnp":1.24},"novel":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01}},"value":0.13,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-4 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-4","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-4","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-4","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-4/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-4","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-4/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":20856,"average":20856.0,"cosmic":20856,"dbsnp":20856},"known":{"1000genomes":5960,"average":5960.0,"cosmic":5960,"dbsnp":5960},"novel":{"1000genomes":14896,"average":14896.0,"cosmic":14896,"dbsnp":14896}},"value":20856,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":20564,"average":20564.0,"cosmic":20564,"dbsnp":20564},"known":{"1000genomes":5739,"average":5739.0,"cosmic":5739,"dbsnp":5739},"novel":{"1000genomes":14825,"average":14825.0,"cosmic":14825,"dbsnp":14825}},"value":20564,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":105,"average":105.0,"cosmic":105,"dbsnp":105},"known":{"1000genomes":94,"average":94.0,"cosmic":94,"dbsnp":94},"novel":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11}},"value":105,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":147,"average":147.0,"cosmic":147,"dbsnp":147},"known":{"1000genomes":122,"average":122.0,"cosmic":122,"dbsnp":122},"novel":{"1000genomes":25,"average":25.0,"cosmic":25,"dbsnp":25}},"value":147,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":2505,"average":3247.3333333333335,"cosmic":1277,"dbsnp":5960},"known":{"1000genomes":2505,"average":3027.0,"cosmic":616,"dbsnp":5960},"novel":{"1000genomes":0,"average":220.33333333333334,"cosmic":661,"dbsnp":0}},"value":1277,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":12.01,"average":15.569999999999999,"cosmic":6.12,"dbsnp":28.58},"known":{"1000genomes":42.03,"average":50.79,"cosmic":10.34,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4800000000000002,"cosmic":4.44,"dbsnp":0.0}},"value":6.12,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":2486,"average":2773.3333333333335,"cosmic":736,"dbsnp":5098},"known":{"1000genomes":2486,"average":2705.3333333333335,"cosmic":532,"dbsnp":5098},"novel":{"1000genomes":0,"average":68.0,"cosmic":204,"dbsnp":0}},"value":736,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.24,"average":80.80666666666667,"cosmic":57.64,"dbsnp":85.54},"known":{"1000genomes":99.24,"average":90.38,"cosmic":86.36,"dbsnp":85.54},"novel":{"1000genomes":0.0,"average":10.286666666666667,"cosmic":30.86,"dbsnp":0.0}},"value":57.64,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":6.65e-06,"average":6.65e-06,"cosmic":6.65e-06,"dbsnp":6.65e-06},"known":{"1000genomes":1.9e-06,"average":1.8999999999999998e-06,"cosmic":1.9e-06,"dbsnp":1.9e-06},"novel":{"1000genomes":4.75e-06,"average":4.75e-06,"cosmic":4.75e-06,"dbsnp":4.75e-06}},"value":6.65e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":150420.0,"average":150420.0,"cosmic":150420.0,"dbsnp":150420.0},"known":{"1000genomes":526369.0,"average":526369.0,"cosmic":526369.0,"dbsnp":526369.0},"novel":{"1000genomes":210604.0,"average":210604.0,"cosmic":210604.0,"dbsnp":210604.0}},"value":150420.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":74.07,"average":74.07,"cosmic":74.07,"dbsnp":74.07},"known":{"1000genomes":32.48,"average":32.48,"cosmic":32.48,"dbsnp":32.48},"novel":{"1000genomes":751.9,"average":751.9,"cosmic":751.9,"dbsnp":751.9}},"value":74.07,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.22,"average":0.22,"cosmic":0.22,"dbsnp":0.22},"known":{"1000genomes":1.55,"average":1.55,"cosmic":1.55,"dbsnp":1.55},"novel":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01}},"value":0.22,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-5 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-5","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-5","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-5","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-5/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-5","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-5/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":7104,"average":7104.0,"cosmic":7104,"dbsnp":7104},"known":{"1000genomes":6849,"average":6849.0,"cosmic":6849,"dbsnp":6849},"novel":{"1000genomes":255,"average":255.0,"cosmic":255,"dbsnp":255}},"value":7104,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":6699,"average":6699.0,"cosmic":6699,"dbsnp":6699},"known":{"1000genomes":6504,"average":6504.0,"cosmic":6504,"dbsnp":6504},"novel":{"1000genomes":195,"average":195.0,"cosmic":195,"dbsnp":195}},"value":6699,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":151,"average":151.0,"cosmic":151,"dbsnp":151},"known":{"1000genomes":145,"average":145.0,"cosmic":145,"dbsnp":145},"novel":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6}},"value":151,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":216,"average":216.0,"cosmic":216,"dbsnp":216},"known":{"1000genomes":195,"average":195.0,"cosmic":195,"dbsnp":195},"novel":{"1000genomes":21,"average":21.0,"cosmic":21,"dbsnp":21}},"value":216,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":3356,"average":3635.3333333333335,"cosmic":701,"dbsnp":6849},"known":{"1000genomes":3355,"average":3629.3333333333335,"cosmic":684,"dbsnp":6849},"novel":{"1000genomes":1,"average":6.0,"cosmic":17,"dbsnp":0}},"value":701,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":47.24,"average":51.17333333333334,"cosmic":9.87,"dbsnp":96.41},"known":{"1000genomes":48.99,"average":52.99333333333334,"cosmic":9.99,"dbsnp":100.0},"novel":{"1000genomes":0.39,"average":2.353333333333333,"cosmic":6.67,"dbsnp":0.0}},"value":9.87,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":3353,"average":3624.6666666666665,"cosmic":683,"dbsnp":6838},"known":{"1000genomes":3352,"average":3619.6666666666665,"cosmic":669,"dbsnp":6838},"novel":{"1000genomes":1,"average":5.0,"cosmic":14,"dbsnp":0}},"value":683,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.91,"average":99.06,"cosmic":97.43,"dbsnp":99.84},"known":{"1000genomes":99.91,"average":99.18666666666667,"cosmic":97.81,"dbsnp":99.84},"novel":{"1000genomes":100.0,"average":60.78333333333333,"cosmic":82.35,"dbsnp":0.0}},"value":97.43,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":2.26e-06,"average":2.26e-06,"cosmic":2.26e-06,"dbsnp":2.26e-06},"known":{"1000genomes":2.18e-06,"average":2.18e-06,"cosmic":2.18e-06,"dbsnp":2.18e-06},"novel":{"1000genomes":8e-08,"average":8.000000000000001e-08,"cosmic":8e-08,"dbsnp":8e-08}},"value":2.26e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":441604.0,"average":441604.0,"cosmic":441604.0,"dbsnp":441604.0},"known":{"1000genomes":458046.0,"average":458046.0,"cosmic":458046.0,"dbsnp":458046.0},"novel":{"1000genomes":12302593.0,"average":12302593.0,"cosmic":12302593.0,"dbsnp":12302593.0}},"value":441604.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":6.97,"average":6.97,"cosmic":6.97,"dbsnp":6.97},"known":{"1000genomes":6.89,"average":6.89,"cosmic":6.89,"dbsnp":6.89},"novel":{"1000genomes":11.39,"average":11.39,"cosmic":11.39,"dbsnp":11.39}},"value":6.97,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.65,"average":2.65,"cosmic":2.65,"dbsnp":2.65},"known":{"1000genomes":2.73,"average":2.73,"cosmic":2.73,"dbsnp":2.73},"novel":{"1000genomes":1.17,"average":1.17,"cosmic":1.17,"dbsnp":1.17}},"value":2.65,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-6 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-6","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-6","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-6","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-6/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-6","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-6/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":11473,"average":11473.0,"cosmic":11473,"dbsnp":11473},"known":{"1000genomes":10890,"average":10890.0,"cosmic":10890,"dbsnp":10890},"novel":{"1000genomes":583,"average":583.0,"cosmic":583,"dbsnp":583}},"value":11473,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":10850,"average":10850.0,"cosmic":10850,"dbsnp":10850},"known":{"1000genomes":10372,"average":10372.0,"cosmic":10372,"dbsnp":10372},"novel":{"1000genomes":478,"average":478.0,"cosmic":478,"dbsnp":478}},"value":10850,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":236,"average":236.0,"cosmic":236,"dbsnp":236},"known":{"1000genomes":224,"average":224.0,"cosmic":224,"dbsnp":224},"novel":{"1000genomes":12,"average":12.0,"cosmic":12,"dbsnp":12}},"value":236,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":313,"average":313.0,"cosmic":313,"dbsnp":313},"known":{"1000genomes":284,"average":284.0,"cosmic":284,"dbsnp":284},"novel":{"1000genomes":29,"average":29.0,"cosmic":29,"dbsnp":29}},"value":313,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":5377,"average":5814.333333333333,"cosmic":1176,"dbsnp":10890},"known":{"1000genomes":5375,"average":5804.333333333333,"cosmic":1148,"dbsnp":10890},"novel":{"1000genomes":2,"average":10.0,"cosmic":28,"dbsnp":0}},"value":1176,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":46.87,"average":50.68,"cosmic":10.25,"dbsnp":94.92},"known":{"1000genomes":49.36,"average":53.300000000000004,"cosmic":10.54,"dbsnp":100.0},"novel":{"1000genomes":0.34,"average":1.7133333333333332,"cosmic":4.8,"dbsnp":0.0}},"value":10.25,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":5371,"average":5784.333333333333,"cosmic":1123,"dbsnp":10859},"known":{"1000genomes":5369,"average":5779.333333333333,"cosmic":1110,"dbsnp":10859},"novel":{"1000genomes":2,"average":5.0,"cosmic":13,"dbsnp":0}},"value":1123,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.89,"average":98.36666666666667,"cosmic":95.49,"dbsnp":99.72},"known":{"1000genomes":99.89,"average":98.76666666666667,"cosmic":96.69,"dbsnp":99.72},"novel":{"1000genomes":100.0,"average":48.81,"cosmic":46.43,"dbsnp":0.0}},"value":95.49,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.66e-06,"average":3.66e-06,"cosmic":3.66e-06,"dbsnp":3.66e-06},"known":{"1000genomes":3.47e-06,"average":3.4700000000000002e-06,"cosmic":3.47e-06,"dbsnp":3.47e-06},"novel":{"1000genomes":1.9e-07,"average":1.9e-07,"cosmic":1.9e-07,"dbsnp":1.9e-07}},"value":3.66e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":273438.0,"average":273438.0,"cosmic":273438.0,"dbsnp":273438.0},"known":{"1000genomes":288077.0,"average":288077.0,"cosmic":288077.0,"dbsnp":288077.0},"novel":{"1000genomes":5381065.0,"average":5381065.0,"cosmic":5381065.0,"dbsnp":5381065.0}},"value":273438.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":3.26,"average":3.26,"cosmic":3.26,"dbsnp":3.26},"known":{"1000genomes":3.17,"average":3.17,"cosmic":3.17,"dbsnp":3.17},"novel":{"1000genomes":9.64,"average":9.64,"cosmic":9.64,"dbsnp":9.64}},"value":3.26,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.48,"average":2.48,"cosmic":2.48,"dbsnp":2.48},"known":{"1000genomes":2.7,"average":2.7000000000000006,"cosmic":2.7,"dbsnp":2.7},"novel":{"1000genomes":0.5,"average":0.5,"cosmic":0.5,"dbsnp":0.5}},"value":2.48,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-7 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-7","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-7","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-7","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-7/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-7","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-7/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":4346,"average":4346.0,"cosmic":4346,"dbsnp":4346},"known":{"1000genomes":2857,"average":2857.0,"cosmic":2857,"dbsnp":2857},"novel":{"1000genomes":1489,"average":1489.0,"cosmic":1489,"dbsnp":1489}},"value":4346,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":3985,"average":3985.0,"cosmic":3985,"dbsnp":3985},"known":{"1000genomes":2693,"average":2693.0,"cosmic":2693,"dbsnp":2693},"novel":{"1000genomes":1292,"average":1292.0,"cosmic":1292,"dbsnp":1292}},"value":3985,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":153,"average":153.0,"cosmic":153,"dbsnp":153},"known":{"1000genomes":73,"average":73.0,"cosmic":73,"dbsnp":73},"novel":{"1000genomes":80,"average":80.0,"cosmic":80,"dbsnp":80}},"value":153,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":156,"average":156.0,"cosmic":156,"dbsnp":156},"known":{"1000genomes":87,"average":87.0,"cosmic":87,"dbsnp":87},"novel":{"1000genomes":69,"average":69.0,"cosmic":69,"dbsnp":69}},"value":156,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":1245,"average":1507.0,"cosmic":419,"dbsnp":2857},"known":{"1000genomes":1245,"average":1474.6666666666667,"cosmic":322,"dbsnp":2857},"novel":{"1000genomes":0,"average":32.333333333333336,"cosmic":97,"dbsnp":0}},"value":419,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":28.65,"average":34.67666666666666,"cosmic":9.64,"dbsnp":65.74},"known":{"1000genomes":43.58,"average":51.61666666666667,"cosmic":11.27,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":2.17,"cosmic":6.51,"dbsnp":0.0}},"value":9.64,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":1244,"average":1476.0,"cosmic":382,"dbsnp":2802},"known":{"1000genomes":1244,"average":1452.3333333333333,"cosmic":311,"dbsnp":2802},"novel":{"1000genomes":0,"average":23.666666666666668,"cosmic":71,"dbsnp":0}},"value":382,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.92,"average":96.38666666666667,"cosmic":91.17,"dbsnp":98.07},"known":{"1000genomes":99.92,"average":98.19,"cosmic":96.58,"dbsnp":98.07},"novel":{"1000genomes":0.0,"average":24.400000000000002,"cosmic":73.2,"dbsnp":0.0}},"value":91.17,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":1.39e-06,"average":1.39e-06,"cosmic":1.39e-06,"dbsnp":1.39e-06},"known":{"1000genomes":9.1e-07,"average":9.100000000000001e-07,"cosmic":9.1e-07,"dbsnp":9.1e-07},"novel":{"1000genomes":4.7e-07,"average":4.7e-07,"cosmic":4.7e-07,"dbsnp":4.7e-07}},"value":1.39e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":721850.0,"average":721850.0,"cosmic":721850.0,"dbsnp":721850.0},"known":{"1000genomes":1098061.0,"average":1098061.0,"cosmic":1098061.0,"dbsnp":1098061.0},"novel":{"1000genomes":2106891.0,"average":2106891.0,"cosmic":2106891.0,"dbsnp":2106891.0}},"value":721850.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":12.26,"average":12.26,"cosmic":12.26,"dbsnp":12.26},"known":{"1000genomes":9.8,"average":9.8,"cosmic":9.8,"dbsnp":9.8},"novel":{"1000genomes":60.35,"average":60.35,"cosmic":60.35,"dbsnp":60.35}},"value":12.26,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":1.81,"average":1.8099999999999998,"cosmic":1.81,"dbsnp":1.81},"known":{"1000genomes":2.64,"average":2.64,"cosmic":2.64,"dbsnp":2.64},"novel":{"1000genomes":0.9,"average":0.9,"cosmic":0.9,"dbsnp":0.9}},"value":1.81,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-8 freebayes","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-8","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-8","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-8","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-8/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-8","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-8/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":1136,"average":1136.0,"cosmic":1136,"dbsnp":1136},"known":{"1000genomes":436,"average":436.0,"cosmic":436,"dbsnp":436},"novel":{"1000genomes":700,"average":700.0,"cosmic":700,"dbsnp":700}},"value":1136,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":1136,"average":1136.0,"cosmic":1136,"dbsnp":1136},"known":{"1000genomes":436,"average":436.0,"cosmic":436,"dbsnp":436},"novel":{"1000genomes":700,"average":700.0,"cosmic":700,"dbsnp":700}},"value":1136,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":19,"average":185.33333333333334,"cosmic":101,"dbsnp":436},"known":{"1000genomes":19,"average":179.33333333333334,"cosmic":83,"dbsnp":436},"novel":{"1000genomes":0,"average":6.0,"cosmic":18,"dbsnp":0}},"value":101,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":1.67,"average":16.313333333333336,"cosmic":8.89,"dbsnp":38.38},"known":{"1000genomes":4.36,"average":41.13333333333333,"cosmic":19.04,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.8566666666666666,"cosmic":2.57,"dbsnp":0.0}},"value":8.89,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":19,"average":176.0,"cosmic":90,"dbsnp":419},"known":{"1000genomes":19,"average":172.0,"cosmic":78,"dbsnp":419},"novel":{"1000genomes":0,"average":4.0,"cosmic":12,"dbsnp":0}},"value":90,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":100.0,"average":95.07,"cosmic":89.11,"dbsnp":96.1},"known":{"1000genomes":100.0,"average":96.69333333333333,"cosmic":93.98,"dbsnp":96.1},"novel":{"1000genomes":0.0,"average":22.223333333333333,"cosmic":66.67,"dbsnp":0.0}},"value":89.11,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.6e-07,"average":3.6e-07,"cosmic":3.6e-07,"dbsnp":3.6e-07},"known":{"1000genomes":1.4e-07,"average":1.4e-07,"cosmic":1.4e-07,"dbsnp":1.4e-07},"novel":{"1000genomes":2.2e-07,"average":2.2e-07,"cosmic":2.2e-07,"dbsnp":2.2e-07}},"value":3.6e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":2761585.0,"average":2761585.0,"cosmic":2761585.0,"dbsnp":2761585.0},"known":{"1000genomes":7195324.0,"average":7195324.0,"cosmic":7195324.0,"dbsnp":7195324.0},"novel":{"1000genomes":4481658.0,"average":4481658.0,"cosmic":4481658.0,"dbsnp":4481658.0}},"value":2761585.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":28.85,"average":28.850000000000005,"cosmic":28.85,"dbsnp":28.85},"known":{"1000genomes":14.9,"average":14.9,"cosmic":14.9,"dbsnp":14.9},"novel":{"1000genomes":69.3,"average":69.3,"cosmic":69.3,"dbsnp":69.3}},"value":28.85,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.45,"average":0.45,"cosmic":0.45,"dbsnp":0.45},"known":{"1000genomes":1.58,"average":1.58,"cosmic":1.58,"dbsnp":1.58},"novel":{"1000genomes":0.14,"average":0.14,"cosmic":0.14,"dbsnp":0.14}},"value":0.45,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-8 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-8","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-8","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-8","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-8/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-8","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-8/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":3219,"average":3219.0,"cosmic":3219,"dbsnp":3219},"known":{"1000genomes":1777,"average":1777.0,"cosmic":1777,"dbsnp":1777},"novel":{"1000genomes":1442,"average":1442.0,"cosmic":1442,"dbsnp":1442}},"value":3219,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":3082,"average":3082.0,"cosmic":3082,"dbsnp":3082},"known":{"1000genomes":1675,"average":1675.0,"cosmic":1675,"dbsnp":1675},"novel":{"1000genomes":1407,"average":1407.0,"cosmic":1407,"dbsnp":1407}},"value":3082,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":50,"average":50.0,"cosmic":50,"dbsnp":50},"known":{"1000genomes":46,"average":46.0,"cosmic":46,"dbsnp":46},"novel":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4}},"value":50,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":69,"average":69.0,"cosmic":69,"dbsnp":69},"known":{"1000genomes":54,"average":54.0,"cosmic":54,"dbsnp":54},"novel":{"1000genomes":15,"average":15.0,"cosmic":15,"dbsnp":15}},"value":69,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":735,"average":918.0,"cosmic":242,"dbsnp":1777},"known":{"1000genomes":735,"average":896.3333333333334,"cosmic":177,"dbsnp":1777},"novel":{"1000genomes":0,"average":21.666666666666668,"cosmic":65,"dbsnp":0}},"value":242,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":22.83,"average":28.516666666666666,"cosmic":7.52,"dbsnp":55.2},"known":{"1000genomes":41.36,"average":50.440000000000005,"cosmic":9.96,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.5033333333333332,"cosmic":4.51,"dbsnp":0.0}},"value":7.52,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":734,"average":879.3333333333334,"cosmic":198,"dbsnp":1706},"known":{"1000genomes":734,"average":870.6666666666666,"cosmic":172,"dbsnp":1706},"novel":{"1000genomes":0,"average":8.666666666666666,"cosmic":26,"dbsnp":0}},"value":198,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.86,"average":92.56,"cosmic":81.82,"dbsnp":96.0},"known":{"1000genomes":99.86,"average":97.68,"cosmic":97.18,"dbsnp":96.0},"novel":{"1000genomes":0.0,"average":13.333333333333334,"cosmic":40.0,"dbsnp":0.0}},"value":81.82,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":1.03e-06,"average":1.03e-06,"cosmic":1.03e-06,"dbsnp":1.03e-06},"known":{"1000genomes":5.7e-07,"average":5.7e-07,"cosmic":5.7e-07,"dbsnp":5.7e-07},"novel":{"1000genomes":4.6e-07,"average":4.6e-07,"cosmic":4.6e-07,"dbsnp":4.6e-07}},"value":1.03e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":974576.0,"average":974576.0,"cosmic":974576.0,"dbsnp":974576.0},"known":{"1000genomes":1765425.0,"average":1765425.0,"cosmic":1765425.0,"dbsnp":1765425.0},"novel":{"1000genomes":2175562.0,"average":2175562.0,"cosmic":2175562.0,"dbsnp":2175562.0}},"value":974576.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":53.74,"average":53.74,"cosmic":53.74,"dbsnp":53.74},"known":{"1000genomes":40.96,"average":40.96,"cosmic":40.96,"dbsnp":40.96},"novel":{"1000genomes":184.75,"average":184.75,"cosmic":184.75,"dbsnp":184.75}},"value":53.74,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.63,"average":0.63,"cosmic":0.63,"dbsnp":0.63},"known":{"1000genomes":2.13,"average":2.13,"cosmic":2.13,"dbsnp":2.13},"novel":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04}},"value":0.63,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-9 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-9","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-9","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-9","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-9/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-9","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-9/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":19494,"average":19494.0,"cosmic":19494,"dbsnp":19494},"known":{"1000genomes":10342,"average":10342.0,"cosmic":10342,"dbsnp":10342},"novel":{"1000genomes":9152,"average":9152.0,"cosmic":9152,"dbsnp":9152}},"value":19494,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":18948,"average":18948.0,"cosmic":18948,"dbsnp":18948},"known":{"1000genomes":9910,"average":9910.0,"cosmic":9910,"dbsnp":9910},"novel":{"1000genomes":9038,"average":9038.0,"cosmic":9038,"dbsnp":9038}},"value":18948,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":209,"average":209.0,"cosmic":209,"dbsnp":209},"known":{"1000genomes":188,"average":188.0,"cosmic":188,"dbsnp":188},"novel":{"1000genomes":21,"average":21.0,"cosmic":21,"dbsnp":21}},"value":209,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":269,"average":269.0,"cosmic":269,"dbsnp":269},"known":{"1000genomes":234,"average":234.0,"cosmic":234,"dbsnp":234},"novel":{"1000genomes":35,"average":35.0,"cosmic":35,"dbsnp":35}},"value":269,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":4920,"average":5585.666666666667,"cosmic":1495,"dbsnp":10342},"known":{"1000genomes":4920,"average":5469.333333333333,"cosmic":1146,"dbsnp":10342},"novel":{"1000genomes":0,"average":116.33333333333333,"cosmic":349,"dbsnp":0}},"value":1495,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":25.24,"average":28.653333333333332,"cosmic":7.67,"dbsnp":53.05},"known":{"1000genomes":47.57,"average":52.88333333333333,"cosmic":11.08,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.27,"cosmic":3.81,"dbsnp":0.0}},"value":7.67,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":4903,"average":5326.666666666667,"cosmic":1222,"dbsnp":9855},"known":{"1000genomes":4903,"average":5281.333333333333,"cosmic":1086,"dbsnp":9855},"novel":{"1000genomes":0,"average":45.333333333333336,"cosmic":136,"dbsnp":0}},"value":1222,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.65,"average":92.22666666666667,"cosmic":81.74,"dbsnp":95.29},"known":{"1000genomes":99.65,"average":96.56666666666666,"cosmic":94.76,"dbsnp":95.29},"novel":{"1000genomes":0.0,"average":12.99,"cosmic":38.97,"dbsnp":0.0}},"value":81.74,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":6.21e-06,"average":6.21e-06,"cosmic":6.21e-06,"dbsnp":6.21e-06},"known":{"1000genomes":3.3e-06,"average":3.3e-06,"cosmic":3.3e-06,"dbsnp":3.3e-06},"novel":{"1000genomes":2.92e-06,"average":2.9200000000000004e-06,"cosmic":2.92e-06,"dbsnp":2.92e-06}},"value":6.21e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":160929.0,"average":160929.0,"cosmic":160929.0,"dbsnp":160929.0},"known":{"1000genomes":303341.0,"average":303341.0,"cosmic":303341.0,"dbsnp":303341.0},"novel":{"1000genomes":342784.0,"average":342784.0,"cosmic":342784.0,"dbsnp":342784.0}},"value":160929.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":9.41,"average":9.41,"cosmic":9.41,"dbsnp":9.41},"known":{"1000genomes":6.23,"average":6.23,"cosmic":6.23,"dbsnp":6.23},"novel":{"1000genomes":169.76,"average":169.76,"cosmic":169.76,"dbsnp":169.76}},"value":9.41,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":0.6,"average":0.6,"cosmic":0.6,"dbsnp":0.6},"known":{"1000genomes":2.22,"average":2.22,"cosmic":2.22,"dbsnp":2.22},"novel":{"1000genomes":0.03,"average":0.03,"cosmic":0.03,"dbsnp":0.03}},"value":0.6,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Resistant-10 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Resistant-10","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Resistant-10","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-10","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Resistant-10/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-10","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Resistant-10/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":10108,"average":10108.0,"cosmic":10108,"dbsnp":10108},"known":{"1000genomes":8729,"average":8729.0,"cosmic":8729,"dbsnp":8729},"novel":{"1000genomes":1379,"average":1379.0,"cosmic":1379,"dbsnp":1379}},"value":10108,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":9599,"average":9599.0,"cosmic":9599,"dbsnp":9599},"known":{"1000genomes":8307,"average":8307.0,"cosmic":8307,"dbsnp":8307},"novel":{"1000genomes":1292,"average":1292.0,"cosmic":1292,"dbsnp":1292}},"value":9599,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":196,"average":196.0,"cosmic":196,"dbsnp":196},"known":{"1000genomes":182,"average":182.0,"cosmic":182,"dbsnp":182},"novel":{"1000genomes":14,"average":14.0,"cosmic":14,"dbsnp":14}},"value":196,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":263,"average":263.0,"cosmic":263,"dbsnp":263},"known":{"1000genomes":230,"average":230.0,"cosmic":230,"dbsnp":230},"novel":{"1000genomes":33,"average":33.0,"cosmic":33,"dbsnp":33}},"value":263,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":4381,"average":4713.333333333333,"cosmic":1030,"dbsnp":8729},"known":{"1000genomes":4379,"average":4690.0,"cosmic":962,"dbsnp":8729},"novel":{"1000genomes":2,"average":23.333333333333332,"cosmic":68,"dbsnp":0}},"value":1030,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":43.34,"average":46.629999999999995,"cosmic":10.19,"dbsnp":86.36},"known":{"1000genomes":50.17,"average":53.73000000000001,"cosmic":11.02,"dbsnp":100.0},"novel":{"1000genomes":0.15,"average":1.6933333333333334,"cosmic":4.93,"dbsnp":0.0}},"value":10.19,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":4376,"average":4666.0,"cosmic":958,"dbsnp":8664},"known":{"1000genomes":4374,"average":4654.0,"cosmic":924,"dbsnp":8664},"novel":{"1000genomes":2,"average":12.0,"cosmic":34,"dbsnp":0}},"value":958,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.89,"average":97.38666666666667,"cosmic":93.01,"dbsnp":99.26},"known":{"1000genomes":99.89,"average":98.39999999999999,"cosmic":96.05,"dbsnp":99.26},"novel":{"1000genomes":100.0,"average":50.0,"cosmic":50.0,"dbsnp":0.0}},"value":93.01,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.22e-06,"average":3.22e-06,"cosmic":3.22e-06,"dbsnp":3.22e-06},"known":{"1000genomes":2.78e-06,"average":2.78e-06,"cosmic":2.78e-06,"dbsnp":2.78e-06},"novel":{"1000genomes":4.4e-07,"average":4.4e-07,"cosmic":4.4e-07,"dbsnp":4.4e-07}},"value":3.22e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":310364.0,"average":310364.0,"cosmic":310364.0,"dbsnp":310364.0},"known":{"1000genomes":359395.0,"average":359395.0,"cosmic":359395.0,"dbsnp":359395.0},"novel":{"1000genomes":2274953.0,"average":2274953.0,"cosmic":2274953.0,"dbsnp":2274953.0}},"value":310364.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":5.2,"average":5.2,"cosmic":5.2,"dbsnp":5.2},"known":{"1000genomes":4.82,"average":4.82,"cosmic":4.82,"dbsnp":4.82},"novel":{"1000genomes":21.51,"average":21.51,"cosmic":21.51,"dbsnp":21.51}},"value":5.2,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":1.86,"average":1.86,"cosmic":1.86,"dbsnp":1.86},"known":{"1000genomes":2.73,"average":2.73,"cosmic":2.73,"dbsnp":2.73},"novel":{"1000genomes":0.14,"average":0.14,"cosmic":0.14,"dbsnp":0.14}},"value":1.86,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-1 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-1","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-1","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-1","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-1/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-1","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-1/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":2235,"average":2235.0,"cosmic":2235,"dbsnp":2235},"known":{"1000genomes":2015,"average":2015.0,"cosmic":2015,"dbsnp":2015},"novel":{"1000genomes":220,"average":220.0,"cosmic":220,"dbsnp":220}},"value":2235,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":2098,"average":2098.0,"cosmic":2098,"dbsnp":2098},"known":{"1000genomes":1909,"average":1909.0,"cosmic":1909,"dbsnp":1909},"novel":{"1000genomes":189,"average":189.0,"cosmic":189,"dbsnp":189}},"value":2098,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":50,"average":50.0,"cosmic":50,"dbsnp":50},"known":{"1000genomes":44,"average":44.0,"cosmic":44,"dbsnp":44},"novel":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6}},"value":50,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":72,"average":72.0,"cosmic":72,"dbsnp":72},"known":{"1000genomes":60,"average":60.0,"cosmic":60,"dbsnp":60},"novel":{"1000genomes":12,"average":12.0,"cosmic":12,"dbsnp":12}},"value":72,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":923,"average":1056.0,"cosmic":230,"dbsnp":2015},"known":{"1000genomes":923,"average":1052.6666666666667,"cosmic":220,"dbsnp":2015},"novel":{"1000genomes":0,"average":3.3333333333333335,"cosmic":10,"dbsnp":0}},"value":230,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":41.3,"average":47.24999999999999,"cosmic":10.29,"dbsnp":90.16},"known":{"1000genomes":45.81,"average":52.24333333333333,"cosmic":10.92,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.5166666666666666,"cosmic":4.55,"dbsnp":0.0}},"value":10.29,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":922,"average":1046.3333333333333,"cosmic":214,"dbsnp":2003},"known":{"1000genomes":922,"average":1045.0,"cosmic":210,"dbsnp":2003},"novel":{"1000genomes":0,"average":1.3333333333333333,"cosmic":4,"dbsnp":0}},"value":214,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.89,"average":97.44333333333334,"cosmic":93.04,"dbsnp":99.4},"known":{"1000genomes":99.89,"average":98.24666666666667,"cosmic":95.45,"dbsnp":99.4},"novel":{"1000genomes":0.0,"average":13.333333333333334,"cosmic":40.0,"dbsnp":0.0}},"value":93.04,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":7.1e-07,"average":7.1e-07,"cosmic":7.1e-07,"dbsnp":7.1e-07},"known":{"1000genomes":6.4e-07,"average":6.400000000000001e-07,"cosmic":6.4e-07,"dbsnp":6.4e-07},"novel":{"1000genomes":7e-08,"average":7e-08,"cosmic":7e-08,"dbsnp":7e-08}},"value":7.1e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":1403651.0,"average":1403651.0,"cosmic":1403651.0,"dbsnp":1403651.0},"known":{"1000genomes":1556903.0,"average":1556903.0,"cosmic":1556903.0,"dbsnp":1556903.0},"novel":{"1000genomes":14259823.0,"average":14259823.0,"cosmic":14259823.0,"dbsnp":14259823.0}},"value":1403651.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":51.04,"average":51.04,"cosmic":51.04,"dbsnp":51.04},"known":{"1000genomes":51.04,"average":51.04,"cosmic":51.04,"dbsnp":51.04},"novel":{"1000genomes":51.0,"average":51.0,"cosmic":51.0,"dbsnp":51.0}},"value":51.04,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.55,"average":2.55,"cosmic":2.55,"dbsnp":2.55},"known":{"1000genomes":3.09,"average":3.09,"cosmic":3.09,"dbsnp":3.09},"novel":{"1000genomes":0.52,"average":0.52,"cosmic":0.52,"dbsnp":0.52}},"value":2.55,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-2 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-2","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-2/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-2","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-2/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":9937,"average":9937.0,"cosmic":9937,"dbsnp":9937},"known":{"1000genomes":9319,"average":9319.0,"cosmic":9319,"dbsnp":9319},"novel":{"1000genomes":618,"average":618.0,"cosmic":618,"dbsnp":618}},"value":9937,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":9372,"average":9372.0,"cosmic":9372,"dbsnp":9372},"known":{"1000genomes":8860,"average":8860.0,"cosmic":8860,"dbsnp":8860},"novel":{"1000genomes":512,"average":512.0,"cosmic":512,"dbsnp":512}},"value":9372,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":204,"average":204.0,"cosmic":204,"dbsnp":204},"known":{"1000genomes":191,"average":191.0,"cosmic":191,"dbsnp":191},"novel":{"1000genomes":13,"average":13.0,"cosmic":13,"dbsnp":13}},"value":204,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":299,"average":299.0,"cosmic":299,"dbsnp":299},"known":{"1000genomes":258,"average":258.0,"cosmic":258,"dbsnp":258},"novel":{"1000genomes":41,"average":41.0,"cosmic":41,"dbsnp":41}},"value":299,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":4604,"average":4982.333333333333,"cosmic":1024,"dbsnp":9319},"known":{"1000genomes":4604,"average":4976.0,"cosmic":1005,"dbsnp":9319},"novel":{"1000genomes":0,"average":6.333333333333333,"cosmic":19,"dbsnp":0}},"value":1024,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":46.33,"average":50.13666666666668,"cosmic":10.3,"dbsnp":93.78},"known":{"1000genomes":49.4,"average":53.39333333333334,"cosmic":10.78,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.0233333333333332,"cosmic":3.07,"dbsnp":0.0}},"value":10.3,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":4601,"average":4954.0,"cosmic":977,"dbsnp":9284},"known":{"1000genomes":4601,"average":4950.0,"cosmic":965,"dbsnp":9284},"novel":{"1000genomes":0,"average":4.0,"cosmic":12,"dbsnp":0}},"value":977,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.93,"average":98.32000000000001,"cosmic":95.41,"dbsnp":99.62},"known":{"1000genomes":99.93,"average":98.52333333333333,"cosmic":96.02,"dbsnp":99.62},"novel":{"1000genomes":0.0,"average":21.05333333333333,"cosmic":63.16,"dbsnp":0.0}},"value":95.41,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.17e-06,"average":3.17e-06,"cosmic":3.17e-06,"dbsnp":3.17e-06},"known":{"1000genomes":2.97e-06,"average":2.97e-06,"cosmic":2.97e-06,"dbsnp":2.97e-06},"novel":{"1000genomes":2e-07,"average":2e-07,"cosmic":2e-07,"dbsnp":2e-07}},"value":3.17e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":315705.0,"average":315705.0,"cosmic":315705.0,"dbsnp":315705.0},"known":{"1000genomes":336641.0,"average":336641.0,"cosmic":336641.0,"dbsnp":336641.0},"novel":{"1000genomes":5076312.0,"average":5076312.0,"cosmic":5076312.0,"dbsnp":5076312.0}},"value":315705.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":3.59,"average":3.59,"cosmic":3.59,"dbsnp":3.59},"known":{"1000genomes":3.49,"average":3.49,"cosmic":3.49,"dbsnp":3.49},"novel":{"1000genomes":7.88,"average":7.88,"cosmic":7.88,"dbsnp":7.88}},"value":3.59,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.56,"average":2.56,"cosmic":2.56,"dbsnp":2.56},"known":{"1000genomes":2.82,"average":2.82,"cosmic":2.82,"dbsnp":2.82},"novel":{"1000genomes":0.62,"average":0.62,"cosmic":0.62,"dbsnp":0.62}},"value":2.56,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-3 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-3","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-3","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-3","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-3/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-3","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-3/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":1120,"average":1120.0,"cosmic":1120,"dbsnp":1120},"known":{"1000genomes":997,"average":997.0,"cosmic":997,"dbsnp":997},"novel":{"1000genomes":123,"average":123.0,"cosmic":123,"dbsnp":123}},"value":1120,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":1024,"average":1024.0,"cosmic":1024,"dbsnp":1024},"known":{"1000genomes":918,"average":918.0,"cosmic":918,"dbsnp":918},"novel":{"1000genomes":106,"average":106.0,"cosmic":106,"dbsnp":106}},"value":1024,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":47,"average":47.0,"cosmic":47,"dbsnp":47},"known":{"1000genomes":44,"average":44.0,"cosmic":44,"dbsnp":44},"novel":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3}},"value":47,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":44,"average":44.0,"cosmic":44,"dbsnp":44},"known":{"1000genomes":34,"average":34.0,"cosmic":34,"dbsnp":34},"novel":{"1000genomes":10,"average":10.0,"cosmic":10,"dbsnp":10}},"value":44,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":409,"average":505.3333333333333,"cosmic":110,"dbsnp":997},"known":{"1000genomes":409,"average":502.3333333333333,"cosmic":101,"dbsnp":997},"novel":{"1000genomes":0,"average":3.0,"cosmic":9,"dbsnp":0}},"value":110,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":36.52,"average":45.12,"cosmic":9.82,"dbsnp":89.02},"known":{"1000genomes":41.02,"average":50.38333333333333,"cosmic":10.13,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":2.44,"cosmic":7.32,"dbsnp":0.0}},"value":9.82,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":409,"average":499.3333333333333,"cosmic":97,"dbsnp":992},"known":{"1000genomes":409,"average":498.0,"cosmic":93,"dbsnp":992},"novel":{"1000genomes":0,"average":1.3333333333333333,"cosmic":4,"dbsnp":0}},"value":97,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":100.0,"average":95.89333333333333,"cosmic":88.18,"dbsnp":99.5},"known":{"1000genomes":100.0,"average":97.19333333333333,"cosmic":92.08,"dbsnp":99.5},"novel":{"1000genomes":0.0,"average":14.813333333333333,"cosmic":44.44,"dbsnp":0.0}},"value":88.18,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.6e-07,"average":3.6e-07,"cosmic":3.6e-07,"dbsnp":3.6e-07},"known":{"1000genomes":3.2e-07,"average":3.2000000000000006e-07,"cosmic":3.2e-07,"dbsnp":3.2e-07},"novel":{"1000genomes":4e-08,"average":4.000000000000001e-08,"cosmic":4e-08,"dbsnp":4e-08}},"value":3.6e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":2801036.0,"average":2801036.0,"cosmic":2801036.0,"dbsnp":2801036.0},"known":{"1000genomes":3146601.0,"average":3146601.0,"cosmic":3146601.0,"dbsnp":3146601.0},"novel":{"1000genomes":25505376.0,"average":25505376.0,"cosmic":25505376.0,"dbsnp":25505376.0}},"value":2801036.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":39.44,"average":39.44,"cosmic":39.44,"dbsnp":39.44},"known":{"1000genomes":37.29,"average":37.29,"cosmic":37.29,"dbsnp":37.29},"novel":{"1000genomes":149.0,"average":149.0,"cosmic":149.0,"dbsnp":149.0}},"value":39.44,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":1.98,"average":1.9799999999999998,"cosmic":1.98,"dbsnp":1.98},"known":{"1000genomes":2.6,"average":2.6,"cosmic":2.6,"dbsnp":2.6},"novel":{"1000genomes":0.19,"average":0.19000000000000003,"cosmic":0.19,"dbsnp":0.19}},"value":1.98,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-5 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-5","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-5","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-5","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-5/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-5","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-5/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":11109,"average":11109.0,"cosmic":11109,"dbsnp":11109},"known":{"1000genomes":10303,"average":10303.0,"cosmic":10303,"dbsnp":10303},"novel":{"1000genomes":806,"average":806.0,"cosmic":806,"dbsnp":806}},"value":11109,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":10536,"average":10536.0,"cosmic":10536,"dbsnp":10536},"known":{"1000genomes":9826,"average":9826.0,"cosmic":9826,"dbsnp":9826},"novel":{"1000genomes":710,"average":710.0,"cosmic":710,"dbsnp":710}},"value":10536,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":239,"average":239.0,"cosmic":239,"dbsnp":239},"known":{"1000genomes":224,"average":224.0,"cosmic":224,"dbsnp":224},"novel":{"1000genomes":15,"average":15.0,"cosmic":15,"dbsnp":15}},"value":239,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":271,"average":271.0,"cosmic":271,"dbsnp":271},"known":{"1000genomes":245,"average":245.0,"cosmic":245,"dbsnp":245},"novel":{"1000genomes":26,"average":26.0,"cosmic":26,"dbsnp":26}},"value":271,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":5164,"average":5551.0,"cosmic":1186,"dbsnp":10303},"known":{"1000genomes":5162,"average":5534.666666666667,"cosmic":1139,"dbsnp":10303},"novel":{"1000genomes":2,"average":16.333333333333332,"cosmic":47,"dbsnp":0}},"value":1186,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":46.48,"average":49.96666666666667,"cosmic":10.68,"dbsnp":92.74},"known":{"1000genomes":50.1,"average":53.72,"cosmic":11.06,"dbsnp":100.0},"novel":{"1000genomes":0.25,"average":2.026666666666667,"cosmic":5.83,"dbsnp":0.0}},"value":10.68,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":5160,"average":5513.666666666667,"cosmic":1127,"dbsnp":10254},"known":{"1000genomes":5158,"average":5505.666666666667,"cosmic":1105,"dbsnp":10254},"novel":{"1000genomes":2,"average":8.0,"cosmic":22,"dbsnp":0}},"value":1127,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.92,"average":98.15666666666668,"cosmic":95.03,"dbsnp":99.52},"known":{"1000genomes":99.92,"average":98.81666666666666,"cosmic":97.01,"dbsnp":99.52},"novel":{"1000genomes":100.0,"average":48.93666666666667,"cosmic":46.81,"dbsnp":0.0}},"value":95.03,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.54e-06,"average":3.54e-06,"cosmic":3.54e-06,"dbsnp":3.54e-06},"known":{"1000genomes":3.28e-06,"average":3.2800000000000004e-06,"cosmic":3.28e-06,"dbsnp":3.28e-06},"novel":{"1000genomes":2.6e-07,"average":2.6e-07,"cosmic":2.6e-07,"dbsnp":2.6e-07}},"value":3.54e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":282398.0,"average":282398.0,"cosmic":282398.0,"dbsnp":282398.0},"known":{"1000genomes":304490.0,"average":304490.0,"cosmic":304490.0,"dbsnp":304490.0},"novel":{"1000genomes":3892259.0,"average":3892259.0,"cosmic":3892259.0,"dbsnp":3892259.0}},"value":282398.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":4.81,"average":4.81,"cosmic":4.81,"dbsnp":4.81},"known":{"1000genomes":4.63,"average":4.63,"cosmic":4.63,"dbsnp":4.63},"novel":{"1000genomes":13.75,"average":13.75,"cosmic":13.75,"dbsnp":13.75}},"value":4.81,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.27,"average":2.27,"cosmic":2.27,"dbsnp":2.27},"known":{"1000genomes":2.66,"average":2.66,"cosmic":2.66,"dbsnp":2.66},"novel":{"1000genomes":0.32,"average":0.32,"cosmic":0.32,"dbsnp":0.32}},"value":2.27,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-6 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-6","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-6","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-6","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-6/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-6","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-6/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":14608,"average":14608.0,"cosmic":14608,"dbsnp":14608},"known":{"1000genomes":13138,"average":13138.0,"cosmic":13138,"dbsnp":13138},"novel":{"1000genomes":1470,"average":1470.0,"cosmic":1470,"dbsnp":1470}},"value":14608,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":13851,"average":13851.0,"cosmic":13851,"dbsnp":13851},"known":{"1000genomes":12500,"average":12500.0,"cosmic":12500,"dbsnp":12500},"novel":{"1000genomes":1351,"average":1351.0,"cosmic":1351,"dbsnp":1351}},"value":13851,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":281,"average":281.0,"cosmic":281,"dbsnp":281},"known":{"1000genomes":266,"average":266.0,"cosmic":266,"dbsnp":266},"novel":{"1000genomes":15,"average":15.0,"cosmic":15,"dbsnp":15}},"value":281,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":400,"average":400.0,"cosmic":400,"dbsnp":400},"known":{"1000genomes":359,"average":359.0,"cosmic":359,"dbsnp":359},"novel":{"1000genomes":41,"average":41.0,"cosmic":41,"dbsnp":41}},"value":400,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":6447,"average":7006.666666666667,"cosmic":1435,"dbsnp":13138},"known":{"1000genomes":6444,"average":6990.0,"cosmic":1388,"dbsnp":13138},"novel":{"1000genomes":3,"average":16.666666666666668,"cosmic":47,"dbsnp":0}},"value":1435,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":44.13,"average":47.96333333333333,"cosmic":9.82,"dbsnp":89.94},"known":{"1000genomes":49.05,"average":53.20333333333334,"cosmic":10.56,"dbsnp":100.0},"novel":{"1000genomes":0.2,"average":1.1333333333333335,"cosmic":3.2,"dbsnp":0.0}},"value":9.82,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":6434,"average":6956.666666666667,"cosmic":1369,"dbsnp":13067},"known":{"1000genomes":6431,"average":6947.666666666667,"cosmic":1345,"dbsnp":13067},"novel":{"1000genomes":3,"average":9.0,"cosmic":24,"dbsnp":0}},"value":1369,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.8,"average":98.21999999999998,"cosmic":95.4,"dbsnp":99.46},"known":{"1000genomes":99.8,"average":98.71999999999998,"cosmic":96.9,"dbsnp":99.46},"novel":{"1000genomes":100.0,"average":50.35333333333333,"cosmic":51.06,"dbsnp":0.0}},"value":95.4,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":4.66e-06,"average":4.66e-06,"cosmic":4.66e-06,"dbsnp":4.66e-06},"known":{"1000genomes":4.19e-06,"average":4.19e-06,"cosmic":4.19e-06,"dbsnp":4.19e-06},"novel":{"1000genomes":4.7e-07,"average":4.7e-07,"cosmic":4.7e-07,"dbsnp":4.7e-07}},"value":4.66e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":214756.0,"average":214756.0,"cosmic":214756.0,"dbsnp":214756.0},"known":{"1000genomes":238785.0,"average":238785.0,"cosmic":238785.0,"dbsnp":238785.0},"novel":{"1000genomes":2134123.0,"average":2134123.0,"cosmic":2134123.0,"dbsnp":2134123.0}},"value":214756.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":5.7,"average":5.7,"cosmic":5.7,"dbsnp":5.7},"known":{"1000genomes":5.38,"average":5.38,"cosmic":5.38,"dbsnp":5.38},"novel":{"1000genomes":22.69,"average":22.69,"cosmic":22.69,"dbsnp":22.69}},"value":5.7,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.11,"average":2.11,"cosmic":2.11,"dbsnp":2.11},"known":{"1000genomes":2.71,"average":2.7099999999999995,"cosmic":2.71,"dbsnp":2.71},"novel":{"1000genomes":0.24,"average":0.24,"cosmic":0.24,"dbsnp":0.24}},"value":2.11,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-7 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-7","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-7","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-7","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-7/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-7","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-7/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":1384,"average":1384.0,"cosmic":1384,"dbsnp":1384},"known":{"1000genomes":1174,"average":1174.0,"cosmic":1174,"dbsnp":1174},"novel":{"1000genomes":210,"average":210.0,"cosmic":210,"dbsnp":210}},"value":1384,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":1255,"average":1255.0,"cosmic":1255,"dbsnp":1255},"known":{"1000genomes":1079,"average":1079.0,"cosmic":1079,"dbsnp":1079},"novel":{"1000genomes":176,"average":176.0,"cosmic":176,"dbsnp":176}},"value":1255,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":50,"average":50.0,"cosmic":50,"dbsnp":50},"known":{"1000genomes":44,"average":44.0,"cosmic":44,"dbsnp":44},"novel":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6}},"value":50,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":63,"average":63.0,"cosmic":63,"dbsnp":63},"known":{"1000genomes":49,"average":49.0,"cosmic":49,"dbsnp":49},"novel":{"1000genomes":14,"average":14.0,"cosmic":14,"dbsnp":14}},"value":63,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":462,"average":591.3333333333334,"cosmic":138,"dbsnp":1174},"known":{"1000genomes":462,"average":587.6666666666666,"cosmic":127,"dbsnp":1174},"novel":{"1000genomes":0,"average":3.6666666666666665,"cosmic":11,"dbsnp":0}},"value":138,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":33.38,"average":42.72666666666667,"cosmic":9.97,"dbsnp":84.83},"known":{"1000genomes":39.35,"average":50.056666666666665,"cosmic":10.82,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.7466666666666668,"cosmic":5.24,"dbsnp":0.0}},"value":9.97,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":462,"average":585.0,"cosmic":130,"dbsnp":1163},"known":{"1000genomes":462,"average":583.3333333333334,"cosmic":125,"dbsnp":1163},"novel":{"1000genomes":0,"average":1.6666666666666667,"cosmic":5,"dbsnp":0}},"value":130,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":100.0,"average":97.75333333333333,"cosmic":94.2,"dbsnp":99.06},"known":{"1000genomes":100.0,"average":99.16333333333334,"cosmic":98.43,"dbsnp":99.06},"novel":{"1000genomes":0.0,"average":15.15,"cosmic":45.45,"dbsnp":0.0}},"value":94.2,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":4.4e-07,"average":4.4e-07,"cosmic":4.4e-07,"dbsnp":4.4e-07},"known":{"1000genomes":3.7e-07,"average":3.7e-07,"cosmic":3.7e-07,"dbsnp":3.7e-07},"novel":{"1000genomes":7e-08,"average":7e-08,"cosmic":7e-08,"dbsnp":7e-08}},"value":4.4e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":2266735.0,"average":2266735.0,"cosmic":2266735.0,"dbsnp":2266735.0},"known":{"1000genomes":2672198.0,"average":2672198.0,"cosmic":2672198.0,"dbsnp":2672198.0},"novel":{"1000genomes":14938863.0,"average":14938863.0,"cosmic":14938863.0,"dbsnp":14938863.0}},"value":2266735.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":35.96,"average":35.96,"cosmic":35.96,"dbsnp":35.96},"known":{"1000genomes":36.1,"average":36.1,"cosmic":36.1,"dbsnp":36.1},"novel":{"1000genomes":34.71,"average":34.71,"cosmic":34.71,"dbsnp":34.71}},"value":35.96,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":1.72,"average":1.72,"cosmic":1.72,"dbsnp":1.72},"known":{"1000genomes":2.38,"average":2.38,"cosmic":2.38,"dbsnp":2.38},"novel":{"1000genomes":0.23,"average":0.23,"cosmic":0.23,"dbsnp":0.23}},"value":1.72,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-8 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-8","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-8","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-8","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-8/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-8","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-8/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":9743,"average":9743.0,"cosmic":9743,"dbsnp":9743},"known":{"1000genomes":9190,"average":9190.0,"cosmic":9190,"dbsnp":9190},"novel":{"1000genomes":553,"average":553.0,"cosmic":553,"dbsnp":553}},"value":9743,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":9192,"average":9192.0,"cosmic":9192,"dbsnp":9192},"known":{"1000genomes":8724,"average":8724.0,"cosmic":8724,"dbsnp":8724},"novel":{"1000genomes":468,"average":468.0,"cosmic":468,"dbsnp":468}},"value":9192,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":209,"average":209.0,"cosmic":209,"dbsnp":209},"known":{"1000genomes":199,"average":199.0,"cosmic":199,"dbsnp":199},"novel":{"1000genomes":10,"average":10.0,"cosmic":10,"dbsnp":10}},"value":209,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":288,"average":288.0,"cosmic":288,"dbsnp":288},"known":{"1000genomes":259,"average":259.0,"cosmic":259,"dbsnp":259},"novel":{"1000genomes":29,"average":29.0,"cosmic":29,"dbsnp":29}},"value":288,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":4536,"average":4914.666666666667,"cosmic":1018,"dbsnp":9190},"known":{"1000genomes":4536,"average":4904.0,"cosmic":986,"dbsnp":9190},"novel":{"1000genomes":0,"average":10.666666666666666,"cosmic":32,"dbsnp":0}},"value":1018,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":46.56,"average":50.44333333333333,"cosmic":10.45,"dbsnp":94.32},"known":{"1000genomes":49.36,"average":53.36333333333334,"cosmic":10.73,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.93,"cosmic":5.79,"dbsnp":0.0}},"value":10.45,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":4531,"average":4890.0,"cosmic":980,"dbsnp":9159},"known":{"1000genomes":4531,"average":4883.0,"cosmic":959,"dbsnp":9159},"novel":{"1000genomes":0,"average":7.0,"cosmic":21,"dbsnp":0}},"value":980,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.89,"average":98.60666666666667,"cosmic":96.27,"dbsnp":99.66},"known":{"1000genomes":99.89,"average":98.93666666666667,"cosmic":97.26,"dbsnp":99.66},"novel":{"1000genomes":0.0,"average":21.876666666666665,"cosmic":65.63,"dbsnp":0.0}},"value":96.27,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.11e-06,"average":3.1100000000000003e-06,"cosmic":3.11e-06,"dbsnp":3.11e-06},"known":{"1000genomes":2.93e-06,"average":2.9299999999999995e-06,"cosmic":2.93e-06,"dbsnp":2.93e-06},"novel":{"1000genomes":1.8e-07,"average":1.8e-07,"cosmic":1.8e-07,"dbsnp":1.8e-07}},"value":3.11e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":321991.0,"average":321991.0,"cosmic":321991.0,"dbsnp":321991.0},"known":{"1000genomes":341366.0,"average":341366.0,"cosmic":341366.0,"dbsnp":341366.0},"novel":{"1000genomes":5672986.0,"average":5672986.0,"cosmic":5672986.0,"dbsnp":5672986.0}},"value":321991.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":5.67,"average":5.669999999999999,"cosmic":5.67,"dbsnp":5.67},"known":{"1000genomes":5.52,"average":5.52,"cosmic":5.52,"dbsnp":5.52},"novel":{"1000genomes":15.68,"average":15.68,"cosmic":15.68,"dbsnp":15.68}},"value":5.67,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.5,"average":2.5,"cosmic":2.5,"dbsnp":2.5},"known":{"1000genomes":2.78,"average":2.78,"cosmic":2.78,"dbsnp":2.78},"novel":{"1000genomes":0.46,"average":0.46,"cosmic":0.46,"dbsnp":0.46}},"value":2.5,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-9 mutect","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-9","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-9","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-9","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-9/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-9","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-9/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":9828,"average":9828.0,"cosmic":9828,"dbsnp":9828},"known":{"1000genomes":9350,"average":9350.0,"cosmic":9350,"dbsnp":9350},"novel":{"1000genomes":478,"average":478.0,"cosmic":478,"dbsnp":478}},"value":9828,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":9285,"average":9285.0,"cosmic":9285,"dbsnp":9285},"known":{"1000genomes":8890,"average":8890.0,"cosmic":8890,"dbsnp":8890},"novel":{"1000genomes":395,"average":395.0,"cosmic":395,"dbsnp":395}},"value":9285,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":192,"average":192.0,"cosmic":192,"dbsnp":192},"known":{"1000genomes":185,"average":185.0,"cosmic":185,"dbsnp":185},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":192,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":302,"average":302.0,"cosmic":302,"dbsnp":302},"known":{"1000genomes":269,"average":269.0,"cosmic":269,"dbsnp":269},"novel":{"1000genomes":33,"average":33.0,"cosmic":33,"dbsnp":33}},"value":302,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":4603,"average":5007.666666666667,"cosmic":1070,"dbsnp":9350},"known":{"1000genomes":4598,"average":4999.666666666667,"cosmic":1051,"dbsnp":9350},"novel":{"1000genomes":5,"average":8.0,"cosmic":19,"dbsnp":0}},"value":1070,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":46.84,"average":50.95666666666667,"cosmic":10.89,"dbsnp":95.14},"known":{"1000genomes":49.18,"average":53.473333333333336,"cosmic":11.24,"dbsnp":100.0},"novel":{"1000genomes":1.05,"average":1.6733333333333336,"cosmic":3.97,"dbsnp":0.0}},"value":10.89,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":4600,"average":4984.333333333333,"cosmic":1035,"dbsnp":9318},"known":{"1000genomes":4595,"average":4978.333333333333,"cosmic":1022,"dbsnp":9318},"novel":{"1000genomes":5,"average":6.0,"cosmic":13,"dbsnp":0}},"value":1035,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.93,"average":98.77333333333333,"cosmic":96.73,"dbsnp":99.66},"known":{"1000genomes":99.93,"average":98.94333333333333,"cosmic":97.24,"dbsnp":99.66},"novel":{"1000genomes":100.0,"average":56.14000000000001,"cosmic":68.42,"dbsnp":0.0}},"value":96.73,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.13e-06,"average":3.13e-06,"cosmic":3.13e-06,"dbsnp":3.13e-06},"known":{"1000genomes":2.98e-06,"average":2.98e-06,"cosmic":2.98e-06,"dbsnp":2.98e-06},"novel":{"1000genomes":1.5e-07,"average":1.5e-07,"cosmic":1.5e-07,"dbsnp":1.5e-07}},"value":3.13e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":319206.0,"average":319206.0,"cosmic":319206.0,"dbsnp":319206.0},"known":{"1000genomes":335525.0,"average":335525.0,"cosmic":335525.0,"dbsnp":335525.0},"novel":{"1000genomes":6563098.0,"average":6563098.0,"cosmic":6563098.0,"dbsnp":6563098.0}},"value":319206.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":4.85,"average":4.85,"cosmic":4.85,"dbsnp":4.85},"known":{"1000genomes":4.74,"average":4.74,"cosmic":4.74,"dbsnp":4.74},"novel":{"1000genomes":11.54,"average":11.54,"cosmic":11.54,"dbsnp":11.54}},"value":4.85,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.48,"average":2.48,"cosmic":2.48,"dbsnp":2.48},"known":{"1000genomes":2.69,"average":2.69,"cosmic":2.69,"dbsnp":2.69},"novel":{"1000genomes":0.54,"average":0.54,"cosmic":0.54,"dbsnp":0.54}},"value":2.48,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-9 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-9","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-9","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-9","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-9/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-9","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-9/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":9828,"average":9828.0,"cosmic":9828,"dbsnp":9828},"known":{"1000genomes":9350,"average":9350.0,"cosmic":9350,"dbsnp":9350},"novel":{"1000genomes":478,"average":478.0,"cosmic":478,"dbsnp":478}},"value":9828,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":9285,"average":9285.0,"cosmic":9285,"dbsnp":9285},"known":{"1000genomes":8890,"average":8890.0,"cosmic":8890,"dbsnp":8890},"novel":{"1000genomes":395,"average":395.0,"cosmic":395,"dbsnp":395}},"value":9285,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":192,"average":192.0,"cosmic":192,"dbsnp":192},"known":{"1000genomes":185,"average":185.0,"cosmic":185,"dbsnp":185},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":192,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":302,"average":302.0,"cosmic":302,"dbsnp":302},"known":{"1000genomes":269,"average":269.0,"cosmic":269,"dbsnp":269},"novel":{"1000genomes":33,"average":33.0,"cosmic":33,"dbsnp":33}},"value":302,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":4603,"average":5007.666666666667,"cosmic":1070,"dbsnp":9350},"known":{"1000genomes":4598,"average":4999.666666666667,"cosmic":1051,"dbsnp":9350},"novel":{"1000genomes":5,"average":8.0,"cosmic":19,"dbsnp":0}},"value":1070,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":46.84,"average":50.95666666666667,"cosmic":10.89,"dbsnp":95.14},"known":{"1000genomes":49.18,"average":53.473333333333336,"cosmic":11.24,"dbsnp":100.0},"novel":{"1000genomes":1.05,"average":1.6733333333333336,"cosmic":3.97,"dbsnp":0.0}},"value":10.89,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":4600,"average":4984.333333333333,"cosmic":1035,"dbsnp":9318},"known":{"1000genomes":4595,"average":4978.333333333333,"cosmic":1022,"dbsnp":9318},"novel":{"1000genomes":5,"average":6.0,"cosmic":13,"dbsnp":0}},"value":1035,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.93,"average":98.77333333333333,"cosmic":96.73,"dbsnp":99.66},"known":{"1000genomes":99.93,"average":98.94333333333333,"cosmic":97.24,"dbsnp":99.66},"novel":{"1000genomes":100.0,"average":56.14000000000001,"cosmic":68.42,"dbsnp":0.0}},"value":96.73,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":3.13e-06,"average":3.13e-06,"cosmic":3.13e-06,"dbsnp":3.13e-06},"known":{"1000genomes":2.98e-06,"average":2.98e-06,"cosmic":2.98e-06,"dbsnp":2.98e-06},"novel":{"1000genomes":1.5e-07,"average":1.5e-07,"cosmic":1.5e-07,"dbsnp":1.5e-07}},"value":3.13e-06,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":319206.0,"average":319206.0,"cosmic":319206.0,"dbsnp":319206.0},"known":{"1000genomes":335525.0,"average":335525.0,"cosmic":335525.0,"dbsnp":335525.0},"novel":{"1000genomes":6563098.0,"average":6563098.0,"cosmic":6563098.0,"dbsnp":6563098.0}},"value":319206.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":4.85,"average":4.85,"cosmic":4.85,"dbsnp":4.85},"known":{"1000genomes":4.74,"average":4.74,"cosmic":4.74,"dbsnp":4.74},"novel":{"1000genomes":11.54,"average":11.54,"cosmic":11.54,"dbsnp":11.54}},"value":4.85,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":2.48,"average":2.48,"cosmic":2.48,"dbsnp":2.48},"known":{"1000genomes":2.69,"average":2.69,"cosmic":2.69,"dbsnp":2.69},"novel":{"1000genomes":0.54,"average":0.54,"cosmic":0.54,"dbsnp":0.54}},"value":2.48,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-10 freebayes","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-10","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-10","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-10","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-10/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-10","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-10/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":1564,"average":1564.0,"cosmic":1564,"dbsnp":1564},"known":{"1000genomes":1203,"average":1203.0,"cosmic":1203,"dbsnp":1203},"novel":{"1000genomes":361,"average":361.0,"cosmic":361,"dbsnp":361}},"value":1564,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":1433,"average":1433.0,"cosmic":1433,"dbsnp":1433},"known":{"1000genomes":1119,"average":1119.0,"cosmic":1119,"dbsnp":1119},"novel":{"1000genomes":314,"average":314.0,"cosmic":314,"dbsnp":314}},"value":1433,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":52,"average":52.0,"cosmic":52,"dbsnp":52},"known":{"1000genomes":45,"average":45.0,"cosmic":45,"dbsnp":45},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":52,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":60,"average":60.0,"cosmic":60,"dbsnp":60},"known":{"1000genomes":39,"average":39.0,"cosmic":39,"dbsnp":39},"novel":{"1000genomes":21,"average":21.0,"cosmic":21,"dbsnp":21}},"value":60,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":514,"average":624.3333333333334,"cosmic":156,"dbsnp":1203},"known":{"1000genomes":514,"average":619.3333333333334,"cosmic":141,"dbsnp":1203},"novel":{"1000genomes":0,"average":5.0,"cosmic":15,"dbsnp":0}},"value":156,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":32.86,"average":39.916666666666664,"cosmic":9.97,"dbsnp":76.92},"known":{"1000genomes":42.73,"average":51.48333333333333,"cosmic":11.72,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.3866666666666667,"cosmic":4.16,"dbsnp":0.0}},"value":9.97,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":513,"average":614.6666666666666,"cosmic":147,"dbsnp":1184},"known":{"1000genomes":513,"average":611.3333333333334,"cosmic":137,"dbsnp":1184},"novel":{"1000genomes":0,"average":3.3333333333333335,"cosmic":10,"dbsnp":0}},"value":147,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.81,"average":97.48666666666668,"cosmic":94.23,"dbsnp":98.42},"known":{"1000genomes":99.81,"average":98.46333333333332,"cosmic":97.16,"dbsnp":98.42},"novel":{"1000genomes":0.0,"average":22.223333333333333,"cosmic":66.67,"dbsnp":0.0}},"value":94.23,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":5e-07,"average":5e-07,"cosmic":5e-07,"dbsnp":5e-07},"known":{"1000genomes":3.8e-07,"average":3.8e-07,"cosmic":3.8e-07,"dbsnp":3.8e-07},"novel":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07}},"value":5e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":2005857.0,"average":2005857.0,"cosmic":2005857.0,"dbsnp":2005857.0},"known":{"1000genomes":2607781.0,"average":2607781.0,"cosmic":2607781.0,"dbsnp":2607781.0},"novel":{"1000genomes":8690197.0,"average":8690197.0,"cosmic":8690197.0,"dbsnp":8690197.0}},"value":2005857.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":44.93,"average":44.93,"cosmic":44.93,"dbsnp":44.93},"known":{"1000genomes":42.48,"average":42.48,"cosmic":42.48,"dbsnp":42.48},"novel":{"1000genomes":67.0,"average":67.0,"cosmic":67.0,"dbsnp":67.0}},"value":44.93,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":1.55,"average":1.55,"cosmic":1.55,"dbsnp":1.55},"known":{"1000genomes":2.45,"average":2.45,"cosmic":2.45,"dbsnp":2.45},"novel":{"1000genomes":0.31,"average":0.31,"cosmic":0.31,"dbsnp":0.31}},"value":1.55,"html_fpath":null}],"plots":[],"html_fpath":null},{"metric_storage":null,"display_name":"Tissue-Sensitive-10 vardict","json_fpath":null,"report_name":"","sample":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"Tissue-Sensitive-10","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":{"vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"name":"PBMC-Sensitive-10","vcf":null,"min_af":0.01,"bed":"/Users/vladsaveliev/Dropbox/az/UCSC_Gene_Exon_Canonical.chr21.bed","normal_match":null,"phenotype":"normal","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-10","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/PBMC-Sensitive-10/var"},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-10","bam":null,"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99-update__only_reports/final/Tissue-Sensitive-10/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"all":{"1000genomes":1564,"average":1564.0,"cosmic":1564,"dbsnp":1564},"known":{"1000genomes":1203,"average":1203.0,"cosmic":1203,"dbsnp":1203},"novel":{"1000genomes":361,"average":361.0,"cosmic":361,"dbsnp":361}},"value":1564,"html_fpath":null},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"all":{"1000genomes":1433,"average":1433.0,"cosmic":1433,"dbsnp":1433},"known":{"1000genomes":1119,"average":1119.0,"cosmic":1119,"dbsnp":1119},"novel":{"1000genomes":314,"average":314.0,"cosmic":314,"dbsnp":314}},"value":1433,"html_fpath":null},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"all":{"1000genomes":52,"average":52.0,"cosmic":52,"dbsnp":52},"known":{"1000genomes":45,"average":45.0,"cosmic":45,"dbsnp":45},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":52,"html_fpath":null},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"all":{"1000genomes":60,"average":60.0,"cosmic":60,"dbsnp":60},"known":{"1000genomes":39,"average":39.0,"cosmic":39,"dbsnp":39},"novel":{"1000genomes":21,"average":21.0,"cosmic":21,"dbsnp":21}},"value":60,"html_fpath":null},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"all":{"1000genomes":514,"average":624.3333333333334,"cosmic":156,"dbsnp":1203},"known":{"1000genomes":514,"average":619.3333333333334,"cosmic":141,"dbsnp":1203},"novel":{"1000genomes":0,"average":5.0,"cosmic":15,"dbsnp":0}},"value":156,"html_fpath":null},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"all":{"1000genomes":32.86,"average":39.916666666666664,"cosmic":9.97,"dbsnp":76.92},"known":{"1000genomes":42.73,"average":51.48333333333333,"cosmic":11.72,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.3866666666666667,"cosmic":4.16,"dbsnp":0.0}},"value":9.97,"html_fpath":null},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"all":{"1000genomes":513,"average":614.6666666666666,"cosmic":147,"dbsnp":1184},"known":{"1000genomes":513,"average":611.3333333333334,"cosmic":137,"dbsnp":1184},"novel":{"1000genomes":0,"average":3.3333333333333335,"cosmic":10,"dbsnp":0}},"value":147,"html_fpath":null},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"all":{"1000genomes":99.81,"average":97.48666666666668,"cosmic":94.23,"dbsnp":98.42},"known":{"1000genomes":99.81,"average":98.46333333333332,"cosmic":97.16,"dbsnp":98.42},"novel":{"1000genomes":0.0,"average":22.223333333333333,"cosmic":66.67,"dbsnp":0.0}},"value":94.23,"html_fpath":null},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"all":{"1000genomes":5e-07,"average":5e-07,"cosmic":5e-07,"dbsnp":5e-07},"known":{"1000genomes":3.8e-07,"average":3.8e-07,"cosmic":3.8e-07,"dbsnp":3.8e-07},"novel":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07}},"value":5e-07,"html_fpath":null},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"all":{"1000genomes":2005857.0,"average":2005857.0,"cosmic":2005857.0,"dbsnp":2005857.0},"known":{"1000genomes":2607781.0,"average":2607781.0,"cosmic":2607781.0,"dbsnp":2607781.0},"novel":{"1000genomes":8690197.0,"average":8690197.0,"cosmic":8690197.0,"dbsnp":8690197.0}},"value":2005857.0,"html_fpath":null},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"all":{"1000genomes":44.93,"average":44.93,"cosmic":44.93,"dbsnp":44.93},"known":{"1000genomes":42.48,"average":42.48,"cosmic":42.48,"dbsnp":42.48},"novel":{"1000genomes":67.0,"average":67.0,"cosmic":67.0,"dbsnp":67.0}},"value":44.93,"html_fpath":null},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"all":{"1000genomes":1.55,"average":1.55,"cosmic":1.55,"dbsnp":1.55},"known":{"1000genomes":2.45,"average":2.45,"cosmic":2.45,"dbsnp":2.45},"novel":{"1000genomes":0.31,"average":0.31,"cosmic":0.31,"dbsnp":0.31}},"value":1.55,"html_fpath":null}],"plots":[],"html_fpath":null}],"name":""},"data_outside_reports":{},"type_":null,"date":"14 November 2014, Friday, 11:37:44"}
            </div>
        </div>
    </div>
</div>
</body>
</html>
