<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>AstraZeneca NGS Reporting Suite by AstraZeneca-NGS</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <script src="javascripts/scale.fix.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1 class="header">AstraZeneca NGS Reporting Suite</h1>
        <p class="header">Target sequencing coverage analysis, Variant annotation and quality control, Post-BCBio-nextgen processing</p>

        <ul>
          <li class="download"><a class="buttons" href="https://github.com/AstraZeneca-NGS/Reporting_Suite/zipball/master">Download ZIP</a></li>
          <li class="download"><a class="buttons" href="https://github.com/AstraZeneca-NGS/Reporting_Suite/tarball/master">Download TAR</a></li>
          <li><a class="buttons github" href="https://github.com/AstraZeneca-NGS/Reporting_Suite">View On GitHub</a></li>
        </ul>

        <p class="header">This project is maintained by <a class="header name" href="https://github.com/AstraZeneca-NGS">AstraZeneca-NGS</a></p>


      </header>
      <section>
        <h1>
<a name="ngs-reporting-suite" class="anchor" href="#ngs-reporting-suite"><span class="octicon octicon-link"></span></a>NGS Reporting Suite</h1>

<p>Variant annotation, filtering and quality control, targeted sequencing coverage analysis.</p>

<h2>
<a name="common-options-for-all-scripts" class="anchor" href="#common-options-for-all-scripts"><span class="octicon octicon-link"></span></a>Common options for all scripts</h2>

<ul>
<li><p><code>--sys-cnf SYSTEM_INFO.yaml</code>. Configuration file with paths to external tools and genome resources. The default is system_info_Waltham.yaml</p></li>
<li><p><code>--run-cnf RUN_INFO.yaml</code>. Customised run details: list of annotations/QC metrics/databases/filtering criteria. The default is run_info.yaml</p></li>
<li><p><code>-o DIR</code> Output directory to save the results. Intermediate files will also be stored in DIR/work_SAMPLE. </p></li>
<li><p><code>-s NAME</code> Sample name. Default is the base name of the input VCF/BAM before the first . or - (i.e. sample-mutect.vcf will give the name "sample").</p></li>
<li><p><code>-t N</code> Number of threads.</p></li>
<li><p><code>--reuse</code> Reuse intermediate results.</p></li>
<li><p><code>--clean</code> Remove DIR/work_SAMPLE directory.</p></li>
<li><p><code>--tmp-base DIR</code> Base directory to store tmp files. Default is /ngs/tmp.</p></li>
</ul><p>Note that all options have defaults in run_info.yaml and can be overwritten by inputting a different configuration file with the option --run-cnf.</p>

<h2>
<a name="variant-annotation" class="anchor" href="#variant-annotation"><span class="octicon octicon-link"></span></a>Variant annotation</h2>

<p><code>varannotate.py --vcf SAMPLE.vcf[.gz] [--bam SAMPLE.bam] -o OUT_DIR</code></p>

<p><code>OUT_DIR</code> will contain:</p>

<ul>
<li><code>SAMPLE.anno.vcf</code></li>
<li><code>SAMPLE.anno.tsv</code></li>
<li><code>SAMPLE.anno.maf</code></li>
</ul><p>The script integrates annotations from GATK, DB NSFP, dbSNP, COSMIC, SNPEff, custom set of VCF and BED files. See default configuration file run_info.yaml. Custom file can be provided by the --run-cnf options.</p>

<p>SAMPLE.anno.tsv file contains the following fields:</p>

<ul>
<li>Sample<br>
</li>
<li>Chromosome<br>
</li>
<li>Position<br>
</li>
<li>REF<br>
</li>
<li>ALT<br>
</li>
<li>ID<br>
</li>
<li>QualityScore<br>
</li>
<li>Depth<br>
</li>
<li>Filter<br>
</li>
<li>dbSNPBuildID<br>
</li>
<li>Effect<br>
</li>
<li>Impact<br>
</li>
<li>FunctionalClass<br>
</li>
<li>CodingRegion<br>
</li>
<li>Gene<br>
</li>
<li>EnsemblGeneID<br>
</li>
<li>EnsemblTranscriptID<br>
</li>
<li>UniportAccNumber<br>
</li>
<li>Codon<br>
</li>
<li>AminoAcid<br>
</li>
<li>AminoAcidLength<br>
</li>
<li>CountInCosmic<br>
</li>
<li>AminoAcidInCosmic<br>
</li>
<li>CDSInCosmic<br>
</li>
<li>OncomineCancer<br>
</li>
<li>OncominePatient<br>
</li>
<li>OncomineMutClass<br>
</li>
<li>OncomineMutClassPC<br>
</li>
<li>Rank<br>
</li>
<li>BaseQailityRankSum<br>
</li>
<li>FischerScore<br>
</li>
<li>GC<br>
</li>
<li>HaplotypeScore<br>
</li>
<li>HomopolymerRunSize<br>
</li>
<li>MappingQuality<br>
</li>
<li>MQualityRankSum<br>
</li>
<li>TotalMappingQualityZeroReads<br>
</li>
<li>QualityByDepth<br>
</li>
<li>ReadPosRankSum<br>
</li>
<li>Genotype<br>
</li>
<li>GenotypeLikelihoods_PhredScaled<br>
</li>
<li>AllelicDepthsForRefAndAlt<br>
</li>
<li>VariantAlleleOrigin<br>
</li>
<li>AlleleBalanceAtHeterozygousSites  # from GATK<br>
</li>
<li>VariationClass<br>
</li>
<li>AlleleType<br>
</li>
<li>VariantIsPrecious<br>
</li>
<li>PresentInDBGap<br>
</li>
<li>Has1000GenomeSubmission<br>
</li>
<li>HapMapValidated<br>
</li>
<li>InOMIM<br>
</li>
<li>CommonVariant<br>
</li>
<li>5%MAFinPopulationsCAF<br>
</li>
<li>SiftScore<br>
</li>
<li>PolyPhen2Score<br>
</li>
<li>PolyPhen2Prediction<br>
</li>
<li>LrtScore<br>
</li>
<li>LrtPrediction<br>
</li>
<li>MutationTasterScore<br>
</li>
<li>MutationTasterPrediction<br>
</li>
<li>MutationAssessorScore<br>
</li>
<li>MutationAssessorPrediction<br>
</li>
<li>FATHMMScore<br>
</li>
<li>1000GenomesAlleleCount<br>
</li>
<li>1000GenomesAlleleFrequency<br>
</li>
<li>ESP6500AfricanAmericanAlleleFrequency<br>
</li>
<li>ESP6500EuropeanAmericanAlleleFrequency<br>
</li>
<li>CPGIsland<br>
</li>
<li>GenomicDuplication<br>
</li>
<li>KnownRepeat<br>
</li>
<li>KnownCopyNumber<br>
</li>
<li>SourceVCF_forMergedRecInCombineVariants<br>
</li>
</ul><h2>
<a name="variant-qc" class="anchor" href="#variant-qc"><span class="octicon octicon-link"></span></a>Variant QC</h2>

<p><code>varqc.py --vcf VARIANTS.vcf[.gz] -o OUT_DIR</code></p>

<p><code>OUT_DIR</code> will contain:</p>

<ul>
<li>SAMPLE.varqc.txt</li>
<li>SAMPLE.variant_distribution.png</li>
<li>SAMPLE.indels.png</li>
<li>SAMPLE.substitutions.png</li>
</ul><p>SAMPLE.vcfqc.txt calculates the following values:</p>

<ul>
<li>nEvalVariants</li>
<li>nSNPs</li>
<li>nInsertions</li>
<li>nDeletions</li>
<li>nVariantsAtComp</li>
<li>compRate</li>
<li>nConcordant</li>
<li>concordantRate</li>
<li>variantRate</li>
<li>variantRatePerBp</li>
<li>hetHomRatio</li>
<li>tiTvRatio</li>
</ul><p>The values also calculated separately for novel and known variants.
For comparison, the values separately calculated for dbSNP, COSMIC, 1000genomes, and also average number is taken.</p>

<h3>
<a name="summary" class="anchor" href="#summary"><span class="octicon octicon-link"></span></a>Summary</h3>

<p><code>varqc_summary.py -d BCBIO_FINAL_DIR [-s samples.txt]</code></p>

<p>Generates project-level reports for each sample:</p>

<ul>
<li>varqc.summary.txt</li>
</ul><h2>
<a name="variant-filtering" class="anchor" href="#variant-filtering"><span class="octicon octicon-link"></span></a>Variant filtering</h2>

<p><code>varfilter.py --vcf VARIANTS.vcf[.gz] &gt; VARIANTS.vcf</code></p>

<p>The program filters an annotated VCF file by SNPEff and using dbSNP and COSMIC. It adds REJECT into the FILTER column if variant </p>

<p>The program can be configurated in run_info.yaml, or by command line options. The defaults are:</p>

<ul>
<li>
<code>count_undetermined</code>: True<br>
Undeteremined won't be counted for the sample count.</li>
<li><p><code>bias</code>:               0.4<br>
Novel or dbSNP variants with strand bias "2;1" or "2;0" and AF &lt; 0.3 will be considered as false positive.</p></li>
<li><p><code>fraction</code>:           1<br>
When a novel variant is present in more than [fraction] of samples and mean allele frequency is less than [freq], it's considered as likely false positive.</p></li>
<li>
<code>freq</code>:               10<br>
When the average allele frequency is also below the [freq], the variant is considered likely false positive.</li>
<li><p><code>sample_cnt</code>:         0.15<br>
When the variant is detected in greater or equal [sample_cnt] samples, the variant is considered likely false positive.</p></li>
<li><p><code>max_ratio</code>:          1<br>
When a variant is present in more than [fraction] of samples, and AF &lt; 0.3, it's considered as likely false positive, even if it's in COSMIC.</p></li>
<li>
<code>min_freq</code>:           0.05<br>
When individual allele frequency &lt; feq for variants, it was considered likely false poitives.</li>
<li>
<code>mean_mq</code>:            20<br>
The filtering mean mapping quality score for variants. The raw variant will be filtered if the mean mapping quality score is less then specified.</li>
<li>
<code>mean_vd</code>:            2<br>
The filtering variant depth. Variants with depth &lt; [mean_vd] will be considered false positive.
<code>maf</code>:                0.0025<br>
If there's MAF with frequency, it'll be considered dbSNP regardless of COSMIC.</li>
<li>
<code>signal_noise</code>:       4<br>
Signal/noise value.</li>
<li><p><code>filt_depth</code>:         0<br>
The filtering total depth. The raw variant will be filtered on first place if the total depth is less then [filt_depth].</p></li>
<li><p><code>min_p_mean</code>:         5<br>
The minimum mean position in reads for variants.</p></li>
<li>
<code>min_q_mean</code>:         25<br>
The minimum mean base quality phred score for variant.</li>
<li>
<code>filt_p_mean</code>:        0<br>
The filtering mean position in reads for variants. The raw variant will be filtered on first place if the mean posititon is less then [filt_p_mean].</li>
<li>
<code>filt_q_mean</code>:        0<br>
The filtering mean base quality phred score for variants. The raw variant will be filtered on first place if the mean quality is less then [filt_q_mean].</li>
</ul><p>A novel variant (non-dbSNP, non-COSMIC) is considered false positive if all three conditions ([fraction], [freq], [sample_cnt]) are met. Any variant meeting the [min_p_mean] or [min_q_mean] conditions are also considered likely false positive. False positive variants are annotated REJECT in the FILTER column, PASS otherwise.</p>

<h2>
<a name="target-sequencing-coverage-statistics" class="anchor" href="#target-sequencing-coverage-statistics"><span class="octicon octicon-link"></span></a>Target sequencing coverage statistics</h2>

<p><code>targetcov.py --bam SAMPLE.bam --bed PANEL.bed -o OUT_DIR</code></p>

<p><code>OUT_DIR</code> will contain:</p>

<ul>
<li>SAMPLE.targetseq.details.gene.txt</li>
<li>SAMPLE.targetseq.summary.txt</li>
</ul><h3>
<a name="summary-1" class="anchor" href="#summary-1"><span class="octicon octicon-link"></span></a>Summary</h3>

<p><code>targetcov_summary.py -d BCBIO_FINAL_DIR [-s samples.txt]</code></p>

<p>Generates project-level reports for each sample:</p>

<ul>
<li>targetcov.summary.txt</li>
</ul><h2>
<a name="ngscat" class="anchor" href="#ngscat"><span class="octicon octicon-link"></span></a>NGScat</h2>

<p><code>ngscat.py --bam SAMPLE.bam --extra-bam SAMPLE2.bam --bed PANEL.bed -o OUT_DIR --saturation y</code></p>

<h2>
<a name="post-bcbio-results-processing" class="anchor" href="#post-bcbio-results-processing"><span class="octicon octicon-link"></span></a>Post-BCBio results processing.</h2>

<p><code>post_for_bcbio.py -d BCB_FINAL_DIR [-s samples.txt] -b PANEL.bed --vcf-suffix mutect [-q]</code></p>

<p>For each sample in the list, runs the following steps:</p>

<ul>
<li>VarAnnotate</li>
<li>VarQC</li>
<li>VarFilter</li>
<li>TargetCov</li>
<li>NGScat</li>
</ul><p><code>-q</code> options adds QualiMap step.</p>

<p>The steps are customisable at run_info.yaml.</p>

<p>Uses qsub to submit jobs. The qsub command line template is the following:</p>

<p><code>{qsub} -pe smp {threads} -S /bin/bash -q batch.q -j n -o {out_fpath} -e {log_fpath} {hold_jid_line} -N {job_name} {runner_script} "{cmdline}"</code></p>

<p>Finally generates summaries for TargetCov and VarQC.</p>

<hr><p>All scripts take their configuration from <code>run_info.yaml</code> by default. Custom configuration file can be provided using the<code>--run-cnf</code> option.</p>

<h1></h1>

<h3>
<a name="authors-and-contributors" class="anchor" href="#authors-and-contributors"><span class="octicon octicon-link"></span></a>Authors and Contributors</h3>

<p>Justin Johnson (<a href="https://github.com/BioInfo" class="user-mention">@BioInfo</a>), Miika Ahdesmaki (<a href="https://github.com/mjafin" class="user-mention">@mjafin</a>), Alexey Gurevich (<a href="https://github.com/alexeigurevich" class="user-mention">@alexeigurevich</a>), Alla Bushoy (<a href="https://github.com/abushoy" class="user-mention">@abushoy</a>), and Vlad Saveliev (<a href="https://github.com/vladsaveliev" class="user-mention">@vladsaveliev</a>).</p>
      </section>
      <footer>
        <p><small>Hosted on <a href="http://pages.github.com">GitHub Pages</a> using the Dinky theme</small></p>
      </footer>
    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
		          <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-51814523-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
