<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/table_sorter/style.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="html_aux/jquery.align-column.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>

    <script type="text/javascript" src="html_aux/dragtable.js"></script>
    <script type="text/javascript" src="html_aux/table_sorter/tsort.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });

            $('table.tableSorter').tableSort();
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for freebayes</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"report":{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"sample_reports":[{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"GPP16015","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP16015","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16015","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16015/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":540,"dbsnp":540,"cosmic":540,"average":540.0},"all":{"1000genomes":654,"dbsnp":654,"cosmic":654,"average":654.0},"novel":{"1000genomes":114,"dbsnp":114,"cosmic":114,"average":114.0}},"value":654},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":509,"dbsnp":509,"cosmic":509,"average":509.0},"all":{"1000genomes":618,"dbsnp":618,"cosmic":618,"average":618.0},"novel":{"1000genomes":109,"dbsnp":109,"cosmic":109,"average":109.0}},"value":618},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":9,"dbsnp":9,"cosmic":9,"average":9.0},"all":{"1000genomes":9,"dbsnp":9,"cosmic":9,"average":9.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":9},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":19,"dbsnp":19,"cosmic":19,"average":19.0},"all":{"1000genomes":24,"dbsnp":24,"cosmic":24,"average":24.0},"novel":{"1000genomes":5,"dbsnp":5,"cosmic":5,"average":5.0}},"value":24},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":246,"dbsnp":540,"cosmic":62,"average":282.6666666666667},"all":{"1000genomes":246,"dbsnp":540,"cosmic":70,"average":285.3333333333333},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":8,"average":2.6666666666666665}},"value":70},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":45.56,"dbsnp":100.0,"cosmic":11.48,"average":52.346666666666664},"all":{"1000genomes":37.61,"dbsnp":82.57,"cosmic":10.7,"average":43.626666666666665},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":7.02,"average":2.34}},"value":10.7},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":246,"dbsnp":530,"cosmic":58,"average":278.0},"all":{"1000genomes":246,"dbsnp":530,"cosmic":60,"average":278.6666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":2,"average":0.6666666666666666}},"value":60},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":98.15,"cosmic":93.55,"average":97.23333333333333},"all":{"1000genomes":100.0,"dbsnp":98.15,"cosmic":85.71,"average":94.62},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":25.0,"average":8.333333333333334}},"value":85.71},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.7e-07,"dbsnp":1.7e-07,"cosmic":1.7e-07,"average":1.7e-07},"all":{"1000genomes":2.1e-07,"dbsnp":2.1e-07,"cosmic":2.1e-07,"average":2.1e-07},"novel":{"1000genomes":4e-08,"dbsnp":4e-08,"cosmic":4e-08,"average":4.000000000000001e-08}},"value":2.1e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":5809557.0,"dbsnp":5809557.0,"cosmic":5809557.0,"average":5809557.0},"all":{"1000genomes":4796882.0,"dbsnp":4796882.0,"cosmic":4796882.0,"average":4796882.0},"novel":{"1000genomes":27518958.0,"dbsnp":27518958.0,"cosmic":27518958.0,"average":27518958.0}},"value":4796882.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.5,"dbsnp":1.5,"cosmic":1.5,"average":1.5},"all":{"1000genomes":1.6,"dbsnp":1.6,"cosmic":1.6,"average":1.6000000000000003},"novel":{"1000genomes":2.17,"dbsnp":2.17,"cosmic":2.17,"average":2.17}},"value":1.6},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.34,"dbsnp":2.34,"cosmic":2.34,"average":2.34},"all":{"1000genomes":1.68,"dbsnp":1.68,"cosmic":1.68,"average":1.68},"novel":{"1000genomes":0.4,"dbsnp":0.4,"cosmic":0.4,"average":0.4000000000000001}},"value":1.68}],"plots":[],"html_fpath":null},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"GPP16188","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP16188","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16188","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16188/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":478,"dbsnp":478,"cosmic":478,"average":478.0},"all":{"1000genomes":504,"dbsnp":504,"cosmic":504,"average":504.0},"novel":{"1000genomes":26,"dbsnp":26,"cosmic":26,"average":26.0}},"value":504},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":442,"dbsnp":442,"cosmic":442,"average":442.0},"all":{"1000genomes":461,"dbsnp":461,"cosmic":461,"average":461.0},"novel":{"1000genomes":19,"dbsnp":19,"cosmic":19,"average":19.0}},"value":461},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":15,"dbsnp":15,"cosmic":15,"average":15.0},"all":{"1000genomes":15,"dbsnp":15,"cosmic":15,"average":15.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":15},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":20,"dbsnp":20,"cosmic":20,"average":20.0},"all":{"1000genomes":27,"dbsnp":27,"cosmic":27,"average":27.0},"novel":{"1000genomes":7,"dbsnp":7,"cosmic":7,"average":7.0}},"value":27},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":200,"dbsnp":478,"cosmic":51,"average":243.0},"all":{"1000genomes":200,"dbsnp":478,"cosmic":55,"average":244.33333333333334},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":4,"average":1.3333333333333333}},"value":55},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":41.84,"dbsnp":100.0,"cosmic":10.67,"average":50.836666666666666},"all":{"1000genomes":39.68,"dbsnp":94.84,"cosmic":10.91,"average":48.47666666666667},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":15.38,"average":5.126666666666667}},"value":10.91},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":200,"dbsnp":477,"cosmic":47,"average":241.33333333333334},"all":{"1000genomes":200,"dbsnp":477,"cosmic":51,"average":242.66666666666666},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":4,"average":1.3333333333333333}},"value":51},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":99.79,"cosmic":92.16,"average":97.31666666666668},"all":{"1000genomes":100.0,"dbsnp":99.79,"cosmic":92.73,"average":97.50666666666667},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":100.0,"average":33.333333333333336}},"value":92.73},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.5e-07,"dbsnp":1.5e-07,"cosmic":1.5e-07,"average":1.5e-07},"all":{"1000genomes":1.6e-07,"dbsnp":1.6e-07,"cosmic":1.6e-07,"average":1.6000000000000003e-07},"novel":{"1000genomes":1e-08,"dbsnp":1e-08,"cosmic":1e-08,"average":1.0000000000000002e-08}},"value":1.6e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":6563098.0,"dbsnp":6563098.0,"cosmic":6563098.0,"average":6563098.0},"all":{"1000genomes":6224526.0,"dbsnp":6224526.0,"cosmic":6224526.0,"average":6224526.0},"novel":{"1000genomes":120660048.0,"dbsnp":120660048.0,"cosmic":120660048.0,"average":120660048.0}},"value":6224526.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":0.76,"dbsnp":0.76,"cosmic":0.76,"average":0.7600000000000001},"all":{"1000genomes":0.77,"dbsnp":0.77,"cosmic":0.77,"average":0.77},"novel":{"1000genomes":1.0,"dbsnp":1.0,"cosmic":1.0,"average":1.0}},"value":0.77},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.4,"dbsnp":2.4,"cosmic":2.4,"average":2.4},"all":{"1000genomes":2.29,"dbsnp":2.29,"cosmic":2.29,"average":2.29},"novel":{"1000genomes":0.9,"dbsnp":0.9,"cosmic":0.9,"average":0.9}},"value":2.29}],"plots":[],"html_fpath":null},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"GPP17600","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP17600","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17600","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17600/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":460,"dbsnp":460,"cosmic":460,"average":460.0},"all":{"1000genomes":484,"dbsnp":484,"cosmic":484,"average":484.0},"novel":{"1000genomes":24,"dbsnp":24,"cosmic":24,"average":24.0}},"value":484},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":428,"dbsnp":428,"cosmic":428,"average":428.0},"all":{"1000genomes":447,"dbsnp":447,"cosmic":447,"average":447.0},"novel":{"1000genomes":19,"dbsnp":19,"cosmic":19,"average":19.0}},"value":447},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":12,"dbsnp":12,"cosmic":12,"average":12.0},"all":{"1000genomes":13,"dbsnp":13,"cosmic":13,"average":13.0},"novel":{"1000genomes":1,"dbsnp":1,"cosmic":1,"average":1.0}},"value":13},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":20,"dbsnp":20,"cosmic":20,"average":20.0},"all":{"1000genomes":24,"dbsnp":24,"cosmic":24,"average":24.0},"novel":{"1000genomes":4,"dbsnp":4,"cosmic":4,"average":4.0}},"value":24},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":193,"dbsnp":460,"cosmic":45,"average":232.66666666666666},"all":{"1000genomes":193,"dbsnp":460,"cosmic":48,"average":233.66666666666666},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":48},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":41.96,"dbsnp":100.0,"cosmic":9.78,"average":50.580000000000005},"all":{"1000genomes":39.88,"dbsnp":95.04,"cosmic":9.92,"average":48.28},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":12.5,"average":4.166666666666667}},"value":9.92},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":193,"dbsnp":460,"cosmic":43,"average":232.0},"all":{"1000genomes":193,"dbsnp":460,"cosmic":46,"average":233.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":46},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":100.0,"cosmic":95.56,"average":98.52},"all":{"1000genomes":100.0,"dbsnp":100.0,"cosmic":95.83,"average":98.61},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":100.0,"average":33.333333333333336}},"value":95.83},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.5e-07,"dbsnp":1.5e-07,"cosmic":1.5e-07,"average":1.5e-07},"all":{"1000genomes":1.5e-07,"dbsnp":1.5e-07,"cosmic":1.5e-07,"average":1.5e-07},"novel":{"1000genomes":1e-08,"dbsnp":1e-08,"cosmic":1e-08,"average":1.0000000000000002e-08}},"value":1.5e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":6819915.0,"dbsnp":6819915.0,"cosmic":6819915.0,"average":6819915.0},"all":{"1000genomes":6481738.0,"dbsnp":6481738.0,"cosmic":6481738.0,"average":6481738.0},"novel":{"1000genomes":130715052.0,"dbsnp":130715052.0,"cosmic":130715052.0,"average":130715052.0}},"value":6481738.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":0.41,"dbsnp":0.41,"cosmic":0.41,"average":0.41},"all":{"1000genomes":0.45,"dbsnp":0.45,"cosmic":0.45,"average":0.45},"novel":{"1000genomes":2.43,"dbsnp":2.43,"cosmic":2.43,"average":2.43}},"value":0.45},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.51,"dbsnp":2.51,"cosmic":2.51,"average":2.51},"all":{"1000genomes":2.26,"dbsnp":2.26,"cosmic":2.26,"average":2.26},"novel":{"1000genomes":0.27,"dbsnp":0.27,"cosmic":0.27,"average":0.27}},"value":2.26}],"plots":[],"html_fpath":null},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"GPP17601","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP17601","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17601","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17601/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":513,"dbsnp":513,"cosmic":513,"average":513.0},"all":{"1000genomes":536,"dbsnp":536,"cosmic":536,"average":536.0},"novel":{"1000genomes":23,"dbsnp":23,"cosmic":23,"average":23.0}},"value":536},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":484,"dbsnp":484,"cosmic":484,"average":484.0},"all":{"1000genomes":500,"dbsnp":500,"cosmic":500,"average":500.0},"novel":{"1000genomes":16,"dbsnp":16,"cosmic":16,"average":16.0}},"value":500},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":11,"dbsnp":11,"cosmic":11,"average":11.0},"all":{"1000genomes":12,"dbsnp":12,"cosmic":12,"average":12.0},"novel":{"1000genomes":1,"dbsnp":1,"cosmic":1,"average":1.0}},"value":12},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":17,"dbsnp":17,"cosmic":17,"average":17.0},"all":{"1000genomes":23,"dbsnp":23,"cosmic":23,"average":23.0},"novel":{"1000genomes":6,"dbsnp":6,"cosmic":6,"average":6.0}},"value":23},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":227,"dbsnp":513,"cosmic":50,"average":263.3333333333333},"all":{"1000genomes":227,"dbsnp":513,"cosmic":53,"average":264.3333333333333},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":53},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":44.25,"dbsnp":100.0,"cosmic":9.75,"average":51.333333333333336},"all":{"1000genomes":42.35,"dbsnp":95.71,"cosmic":9.89,"average":49.31666666666666},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":13.04,"average":4.346666666666667}},"value":9.89},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":227,"dbsnp":511,"cosmic":48,"average":262.0},"all":{"1000genomes":227,"dbsnp":511,"cosmic":51,"average":263.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":51},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":99.61,"cosmic":96.0,"average":98.53666666666668},"all":{"1000genomes":100.0,"dbsnp":99.61,"cosmic":96.23,"average":98.61333333333334},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":100.0,"average":33.333333333333336}},"value":96.23},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.6e-07,"dbsnp":1.6e-07,"cosmic":1.6e-07,"average":1.6000000000000003e-07},"all":{"1000genomes":1.7e-07,"dbsnp":1.7e-07,"cosmic":1.7e-07,"average":1.7e-07},"novel":{"1000genomes":1e-08,"dbsnp":1e-08,"cosmic":1e-08,"average":1.0000000000000002e-08}},"value":1.7e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":6115324.0,"dbsnp":6115324.0,"cosmic":6115324.0,"average":6115324.0},"all":{"1000genomes":5852912.0,"dbsnp":5852912.0,"cosmic":5852912.0,"average":5852912.0},"novel":{"1000genomes":136398315.0,"dbsnp":136398315.0,"cosmic":136398315.0,"average":136398315.0}},"value":5852912.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.44,"dbsnp":1.44,"cosmic":1.44,"average":1.4400000000000002},"all":{"1000genomes":1.54,"dbsnp":1.54,"cosmic":1.54,"average":1.54},"novel":{"1000genomes":22.0,"dbsnp":22.0,"cosmic":22.0,"average":22.0}},"value":1.54},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.41,"dbsnp":2.41,"cosmic":2.41,"average":2.41},"all":{"1000genomes":2.31,"dbsnp":2.31,"cosmic":2.31,"average":2.31},"novel":{"1000genomes":0.78,"dbsnp":0.78,"cosmic":0.78,"average":0.7799999999999999}},"value":2.31}],"plots":[],"html_fpath":null}],"name":""},"data_outside_reports":{},"date":"22 September 2014, Monday, 20:27:13"}
            </div>
        </div>
    </div>
</div>
</body>
</html>
