<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/table_sorter/style.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="html_aux/jquery.align-column.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>

    <script type="text/javascript" src="html_aux/dragtable.js"></script>
    <script type="text/javascript" src="html_aux/table_sorter/tsort.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });

            $('table.tableSorter').tableSort();
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"report":{"metric_storage":{"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"Less is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"Less is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"general_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"Less is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"Less is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"sample_reports":[{"metric_storage":null,"display_name":"GPP16015 freebayes","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP16015","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16015","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16015/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":540,"dbsnp":540,"cosmic":540,"average":540.0},"all":{"1000genomes":654,"dbsnp":654,"cosmic":654,"average":654.0},"novel":{"1000genomes":114,"dbsnp":114,"cosmic":114,"average":114.0}},"value":654},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":509,"dbsnp":509,"cosmic":509,"average":509.0},"all":{"1000genomes":618,"dbsnp":618,"cosmic":618,"average":618.0},"novel":{"1000genomes":109,"dbsnp":109,"cosmic":109,"average":109.0}},"value":618},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":9,"dbsnp":9,"cosmic":9,"average":9.0},"all":{"1000genomes":9,"dbsnp":9,"cosmic":9,"average":9.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":9},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":19,"dbsnp":19,"cosmic":19,"average":19.0},"all":{"1000genomes":24,"dbsnp":24,"cosmic":24,"average":24.0},"novel":{"1000genomes":5,"dbsnp":5,"cosmic":5,"average":5.0}},"value":24},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":246,"dbsnp":540,"cosmic":62,"average":282.6666666666667},"all":{"1000genomes":246,"dbsnp":540,"cosmic":70,"average":285.3333333333333},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":8,"average":2.6666666666666665}},"value":70},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":45.56,"dbsnp":100.0,"cosmic":11.48,"average":52.346666666666664},"all":{"1000genomes":37.61,"dbsnp":82.57,"cosmic":10.7,"average":43.626666666666665},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":7.02,"average":2.34}},"value":10.7},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":246,"dbsnp":530,"cosmic":58,"average":278.0},"all":{"1000genomes":246,"dbsnp":530,"cosmic":60,"average":278.6666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":2,"average":0.6666666666666666}},"value":60},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":98.15,"cosmic":93.55,"average":97.23333333333333},"all":{"1000genomes":100.0,"dbsnp":98.15,"cosmic":85.71,"average":94.62},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":25.0,"average":8.333333333333334}},"value":85.71},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.7e-07,"dbsnp":1.7e-07,"cosmic":1.7e-07,"average":1.7e-07},"all":{"1000genomes":2.1e-07,"dbsnp":2.1e-07,"cosmic":2.1e-07,"average":2.1e-07},"novel":{"1000genomes":4e-08,"dbsnp":4e-08,"cosmic":4e-08,"average":4.000000000000001e-08}},"value":2.1e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":5809557.0,"dbsnp":5809557.0,"cosmic":5809557.0,"average":5809557.0},"all":{"1000genomes":4796882.0,"dbsnp":4796882.0,"cosmic":4796882.0,"average":4796882.0},"novel":{"1000genomes":27518958.0,"dbsnp":27518958.0,"cosmic":27518958.0,"average":27518958.0}},"value":4796882.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.5,"dbsnp":1.5,"cosmic":1.5,"average":1.5},"all":{"1000genomes":1.6,"dbsnp":1.6,"cosmic":1.6,"average":1.6000000000000003},"novel":{"1000genomes":2.17,"dbsnp":2.17,"cosmic":2.17,"average":2.17}},"value":1.6},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.34,"dbsnp":2.34,"cosmic":2.34,"average":2.34},"all":{"1000genomes":1.68,"dbsnp":1.68,"cosmic":1.68,"average":1.68},"novel":{"1000genomes":0.4,"dbsnp":0.4,"cosmic":0.4,"average":0.4000000000000001}},"value":1.68}],"plots":[],"html_fpath":"../../../GPP16015/qc/varQC/GPP16015-freebayes.varQC.json"},{"metric_storage":null,"display_name":"GPP16015 mutect","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP16015","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16015","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16015/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":170,"dbsnp":170,"cosmic":170,"average":170.0},"all":{"1000genomes":517,"dbsnp":517,"cosmic":517,"average":517.0},"novel":{"1000genomes":347,"dbsnp":347,"cosmic":347,"average":347.0}},"value":517},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":163,"dbsnp":163,"cosmic":163,"average":163.0},"all":{"1000genomes":504,"dbsnp":504,"cosmic":504,"average":504.0},"novel":{"1000genomes":341,"dbsnp":341,"cosmic":341,"average":341.0}},"value":504},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":4,"dbsnp":4,"cosmic":4,"average":4.0},"all":{"1000genomes":8,"dbsnp":8,"cosmic":8,"average":8.0},"novel":{"1000genomes":4,"dbsnp":4,"cosmic":4,"average":4.0}},"value":8},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":3,"dbsnp":3,"cosmic":3,"average":3.0},"all":{"1000genomes":5,"dbsnp":5,"cosmic":5,"average":5.0},"novel":{"1000genomes":2,"dbsnp":2,"cosmic":2,"average":2.0}},"value":5},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":50,"dbsnp":170,"cosmic":23,"average":81.0},"all":{"1000genomes":50,"dbsnp":170,"cosmic":40,"average":86.66666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":17,"average":5.666666666666667}},"value":40},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":29.41,"dbsnp":100.0,"cosmic":13.53,"average":47.64666666666667},"all":{"1000genomes":9.67,"dbsnp":32.88,"cosmic":7.74,"average":16.763333333333335},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":4.9,"average":1.6333333333333335}},"value":7.74},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":49,"dbsnp":149,"cosmic":22,"average":73.33333333333333},"all":{"1000genomes":49,"dbsnp":149,"cosmic":32,"average":76.66666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":10,"average":3.3333333333333335}},"value":32},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":98.0,"dbsnp":87.65,"cosmic":95.65,"average":93.76666666666667},"all":{"1000genomes":98.0,"dbsnp":87.65,"cosmic":80.0,"average":88.55},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":58.82,"average":19.606666666666666}},"value":80.0},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":5e-08,"dbsnp":5e-08,"cosmic":5e-08,"average":5e-08},"all":{"1000genomes":1.6e-07,"dbsnp":1.6e-07,"cosmic":1.6e-07,"average":1.6000000000000003e-07},"novel":{"1000genomes":1.1e-07,"dbsnp":1.1e-07,"cosmic":1.1e-07,"average":1.1e-07}},"value":1.6e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":18453889.0,"dbsnp":18453889.0,"cosmic":18453889.0,"average":18453889.0},"all":{"1000genomes":6068010.0,"dbsnp":6068010.0,"cosmic":6068010.0,"average":6068010.0},"novel":{"1000genomes":9040810.0,"dbsnp":9040810.0,"cosmic":9040810.0,"average":9040810.0}},"value":6068010.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":84.0,"dbsnp":84.0,"cosmic":84.0,"average":84.0},"all":{"1000genomes":72.86,"dbsnp":72.86,"cosmic":72.86,"average":72.86},"novel":{"1000genomes":68.4,"dbsnp":68.4,"cosmic":68.4,"average":68.4}},"value":72.86},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.62,"dbsnp":2.62,"cosmic":2.62,"average":2.62},"all":{"1000genomes":0.63,"dbsnp":0.63,"cosmic":0.63,"average":0.63},"novel":{"1000genomes":0.29,"dbsnp":0.29,"cosmic":0.29,"average":0.29}},"value":0.63}],"plots":[],"html_fpath":"../../../GPP16015/qc/varQC/GPP16015-mutect.varQC.json"},{"metric_storage":null,"display_name":"GPP16015 vardict","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP16015","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16015","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16015/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":700,"dbsnp":700,"cosmic":700,"average":700.0},"all":{"1000genomes":2474,"dbsnp":2474,"cosmic":2474,"average":2474.0},"novel":{"1000genomes":1774,"dbsnp":1774,"cosmic":1774,"average":1774.0}},"value":2474},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":673,"dbsnp":673,"cosmic":673,"average":673.0},"all":{"1000genomes":2168,"dbsnp":2168,"cosmic":2168,"average":2168.0},"novel":{"1000genomes":1495,"dbsnp":1495,"cosmic":1495,"average":1495.0}},"value":2168},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":9,"dbsnp":9,"cosmic":9,"average":9.0},"all":{"1000genomes":258,"dbsnp":258,"cosmic":258,"average":258.0},"novel":{"1000genomes":249,"dbsnp":249,"cosmic":249,"average":249.0}},"value":258},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":18,"dbsnp":18,"cosmic":18,"average":18.0},"all":{"1000genomes":42,"dbsnp":42,"cosmic":42,"average":42.0},"novel":{"1000genomes":24,"dbsnp":24,"cosmic":24,"average":24.0}},"value":42},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":239,"dbsnp":700,"cosmic":76,"average":338.3333333333333},"all":{"1000genomes":239,"dbsnp":700,"cosmic":188,"average":375.6666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":112,"average":37.333333333333336}},"value":188},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":34.14,"dbsnp":100.0,"cosmic":10.86,"average":48.333333333333336},"all":{"1000genomes":9.66,"dbsnp":28.29,"cosmic":7.6,"average":15.183333333333335},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":6.31,"average":2.103333333333333}},"value":7.6},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":237,"dbsnp":593,"cosmic":65,"average":298.3333333333333},"all":{"1000genomes":237,"dbsnp":593,"cosmic":130,"average":320.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":65,"average":21.666666666666668}},"value":130},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":99.16,"dbsnp":84.71,"cosmic":85.53,"average":89.8},"all":{"1000genomes":99.16,"dbsnp":84.71,"cosmic":69.15,"average":84.34},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":58.04,"average":19.346666666666668}},"value":69.15},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.2e-07,"dbsnp":2.2e-07,"cosmic":2.2e-07,"average":2.2e-07},"all":{"1000genomes":7.9e-07,"dbsnp":7.9e-07,"cosmic":7.9e-07,"average":7.9e-07},"novel":{"1000genomes":5.7e-07,"dbsnp":5.7e-07,"cosmic":5.7e-07,"average":5.7e-07}},"value":7.9e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":4481658.0,"dbsnp":4481658.0,"cosmic":4481658.0,"average":4481658.0},"all":{"1000genomes":1268052.0,"dbsnp":1268052.0,"cosmic":1268052.0,"average":1268052.0},"novel":{"1000genomes":1768411.0,"dbsnp":1768411.0,"cosmic":1768411.0,"average":1768411.0}},"value":1268052.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":2.78,"dbsnp":2.78,"cosmic":2.78,"average":2.78},"all":{"1000genomes":10.56,"dbsnp":10.56,"cosmic":10.56,"average":10.56},"novel":{"1000genomes":60.17,"dbsnp":60.17,"cosmic":60.17,"average":60.169999999999995}},"value":10.56},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":1.83,"dbsnp":1.83,"cosmic":1.83,"average":1.83},"all":{"1000genomes":0.53,"dbsnp":0.53,"cosmic":0.53,"average":0.53},"novel":{"1000genomes":0.27,"dbsnp":0.27,"cosmic":0.27,"average":0.27}},"value":0.53}],"plots":[],"html_fpath":"../../../GPP16015/qc/varQC/GPP16015-vardict.varQC.json"},{"metric_storage":null,"display_name":"GPP16188 freebayes","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP16188","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16188","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16188/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":478,"dbsnp":478,"cosmic":478,"average":478.0},"all":{"1000genomes":504,"dbsnp":504,"cosmic":504,"average":504.0},"novel":{"1000genomes":26,"dbsnp":26,"cosmic":26,"average":26.0}},"value":504},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":442,"dbsnp":442,"cosmic":442,"average":442.0},"all":{"1000genomes":461,"dbsnp":461,"cosmic":461,"average":461.0},"novel":{"1000genomes":19,"dbsnp":19,"cosmic":19,"average":19.0}},"value":461},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":15,"dbsnp":15,"cosmic":15,"average":15.0},"all":{"1000genomes":15,"dbsnp":15,"cosmic":15,"average":15.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":15},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":20,"dbsnp":20,"cosmic":20,"average":20.0},"all":{"1000genomes":27,"dbsnp":27,"cosmic":27,"average":27.0},"novel":{"1000genomes":7,"dbsnp":7,"cosmic":7,"average":7.0}},"value":27},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":200,"dbsnp":478,"cosmic":51,"average":243.0},"all":{"1000genomes":200,"dbsnp":478,"cosmic":55,"average":244.33333333333334},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":4,"average":1.3333333333333333}},"value":55},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":41.84,"dbsnp":100.0,"cosmic":10.67,"average":50.836666666666666},"all":{"1000genomes":39.68,"dbsnp":94.84,"cosmic":10.91,"average":48.47666666666667},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":15.38,"average":5.126666666666667}},"value":10.91},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":200,"dbsnp":477,"cosmic":47,"average":241.33333333333334},"all":{"1000genomes":200,"dbsnp":477,"cosmic":51,"average":242.66666666666666},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":4,"average":1.3333333333333333}},"value":51},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":99.79,"cosmic":92.16,"average":97.31666666666668},"all":{"1000genomes":100.0,"dbsnp":99.79,"cosmic":92.73,"average":97.50666666666667},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":100.0,"average":33.333333333333336}},"value":92.73},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.5e-07,"dbsnp":1.5e-07,"cosmic":1.5e-07,"average":1.5e-07},"all":{"1000genomes":1.6e-07,"dbsnp":1.6e-07,"cosmic":1.6e-07,"average":1.6000000000000003e-07},"novel":{"1000genomes":1e-08,"dbsnp":1e-08,"cosmic":1e-08,"average":1.0000000000000002e-08}},"value":1.6e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":6563098.0,"dbsnp":6563098.0,"cosmic":6563098.0,"average":6563098.0},"all":{"1000genomes":6224526.0,"dbsnp":6224526.0,"cosmic":6224526.0,"average":6224526.0},"novel":{"1000genomes":120660048.0,"dbsnp":120660048.0,"cosmic":120660048.0,"average":120660048.0}},"value":6224526.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":0.76,"dbsnp":0.76,"cosmic":0.76,"average":0.7600000000000001},"all":{"1000genomes":0.77,"dbsnp":0.77,"cosmic":0.77,"average":0.77},"novel":{"1000genomes":1.0,"dbsnp":1.0,"cosmic":1.0,"average":1.0}},"value":0.77},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.4,"dbsnp":2.4,"cosmic":2.4,"average":2.4},"all":{"1000genomes":2.29,"dbsnp":2.29,"cosmic":2.29,"average":2.29},"novel":{"1000genomes":0.9,"dbsnp":0.9,"cosmic":0.9,"average":0.9}},"value":2.29}],"plots":[],"html_fpath":"../../../GPP16188/qc/varQC/GPP16188-freebayes.varQC.json"},{"metric_storage":null,"display_name":"GPP16188 mutect","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP16188","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16188","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16188/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":234,"dbsnp":234,"cosmic":234,"average":234.0},"all":{"1000genomes":279,"dbsnp":279,"cosmic":279,"average":279.0},"novel":{"1000genomes":45,"dbsnp":45,"cosmic":45,"average":45.0}},"value":279},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":222,"dbsnp":222,"cosmic":222,"average":222.0},"all":{"1000genomes":267,"dbsnp":267,"cosmic":267,"average":267.0},"novel":{"1000genomes":45,"dbsnp":45,"cosmic":45,"average":45.0}},"value":267},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":5,"dbsnp":5,"cosmic":5,"average":5.0},"all":{"1000genomes":5,"dbsnp":5,"cosmic":5,"average":5.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":5},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":7,"dbsnp":7,"cosmic":7,"average":7.0},"all":{"1000genomes":7,"dbsnp":7,"cosmic":7,"average":7.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":7},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":81,"dbsnp":234,"cosmic":36,"average":117.0},"all":{"1000genomes":81,"dbsnp":234,"cosmic":42,"average":119.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":6,"average":2.0}},"value":42},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":34.62,"dbsnp":100.0,"cosmic":15.38,"average":50.0},"all":{"1000genomes":29.03,"dbsnp":83.87,"cosmic":15.05,"average":42.65},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":13.33,"average":4.443333333333333}},"value":15.05},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":81,"dbsnp":233,"cosmic":31,"average":115.0},"all":{"1000genomes":81,"dbsnp":233,"cosmic":36,"average":116.66666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":5,"average":1.6666666666666667}},"value":36},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":99.57,"cosmic":86.11,"average":95.22666666666667},"all":{"1000genomes":100.0,"dbsnp":99.57,"cosmic":85.71,"average":95.09333333333332},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":83.33,"average":27.776666666666667}},"value":85.71},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":7e-08,"dbsnp":7e-08,"cosmic":7e-08,"average":7e-08},"all":{"1000genomes":9e-08,"dbsnp":9e-08,"cosmic":9e-08,"average":9e-08},"novel":{"1000genomes":1e-08,"dbsnp":1e-08,"cosmic":1e-08,"average":1.0000000000000002e-08}},"value":9e-08},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":13406672.0,"dbsnp":13406672.0,"cosmic":13406672.0,"average":13406672.0},"all":{"1000genomes":11244305.0,"dbsnp":11244305.0,"cosmic":11244305.0,"average":11244305.0},"novel":{"1000genomes":69714694.0,"dbsnp":69714694.0,"cosmic":69714694.0,"average":69714694.0}},"value":11244305.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":57.5,"dbsnp":57.5,"cosmic":57.5,"average":57.5},"all":{"1000genomes":68.75,"dbsnp":68.75,"cosmic":68.75,"average":68.75},"novel":{"1000genomes":45.0,"dbsnp":45.0,"cosmic":45.0,"average":45.0}},"value":68.75},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.36,"dbsnp":2.36,"cosmic":2.36,"average":2.36},"all":{"1000genomes":2.34,"dbsnp":2.34,"cosmic":2.34,"average":2.34},"novel":{"1000genomes":2.21,"dbsnp":2.21,"cosmic":2.21,"average":2.21}},"value":2.34}],"plots":[],"html_fpath":"../../../GPP16188/qc/varQC/GPP16188-mutect.varQC.json"},{"metric_storage":null,"display_name":"GPP16188 vardict","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP16188","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16188","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP16188/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":489,"dbsnp":489,"cosmic":489,"average":489.0},"all":{"1000genomes":705,"dbsnp":705,"cosmic":705,"average":705.0},"novel":{"1000genomes":216,"dbsnp":216,"cosmic":216,"average":216.0}},"value":705},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":446,"dbsnp":446,"cosmic":446,"average":446.0},"all":{"1000genomes":600,"dbsnp":600,"cosmic":600,"average":600.0},"novel":{"1000genomes":154,"dbsnp":154,"cosmic":154,"average":154.0}},"value":600},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":14,"dbsnp":14,"cosmic":14,"average":14.0},"all":{"1000genomes":31,"dbsnp":31,"cosmic":31,"average":31.0},"novel":{"1000genomes":17,"dbsnp":17,"cosmic":17,"average":17.0}},"value":31},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":29,"dbsnp":29,"cosmic":29,"average":29.0},"all":{"1000genomes":74,"dbsnp":74,"cosmic":74,"average":74.0},"novel":{"1000genomes":45,"dbsnp":45,"cosmic":45,"average":45.0}},"value":74},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":193,"dbsnp":489,"cosmic":62,"average":248.0},"all":{"1000genomes":193,"dbsnp":489,"cosmic":81,"average":254.33333333333334},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":19,"average":6.333333333333333}},"value":81},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":39.47,"dbsnp":100.0,"cosmic":12.68,"average":50.71666666666667},"all":{"1000genomes":27.38,"dbsnp":69.36,"cosmic":11.49,"average":36.07666666666666},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":8.8,"average":2.9333333333333336}},"value":11.49},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":193,"dbsnp":485,"cosmic":56,"average":244.66666666666666},"all":{"1000genomes":193,"dbsnp":485,"cosmic":73,"average":250.33333333333334},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":17,"average":5.666666666666667}},"value":73},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":99.18,"cosmic":90.32,"average":96.5},"all":{"1000genomes":100.0,"dbsnp":99.18,"cosmic":90.12,"average":96.43333333333334},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":89.47,"average":29.823333333333334}},"value":90.12},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.6e-07,"dbsnp":1.6e-07,"cosmic":1.6e-07,"average":1.6000000000000003e-07},"all":{"1000genomes":2.2e-07,"dbsnp":2.2e-07,"cosmic":2.2e-07,"average":2.2e-07},"novel":{"1000genomes":7e-08,"dbsnp":7e-08,"cosmic":7e-08,"average":7e-08}},"value":2.2e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":6415462.0,"dbsnp":6415462.0,"cosmic":6415462.0,"average":6415462.0},"all":{"1000genomes":4449874.0,"dbsnp":4449874.0,"cosmic":4449874.0,"average":4449874.0},"novel":{"1000genomes":14523894.0,"dbsnp":14523894.0,"cosmic":14523894.0,"average":14523894.0}},"value":4449874.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":0.93,"dbsnp":0.93,"cosmic":0.93,"average":0.93},"all":{"1000genomes":1.63,"dbsnp":1.63,"cosmic":1.63,"average":1.63},"novel":{"1000genomes":14.43,"dbsnp":14.43,"cosmic":14.43,"average":14.43}},"value":1.63},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.6,"dbsnp":2.6,"cosmic":2.6,"average":2.6},"all":{"1000genomes":2.23,"dbsnp":2.23,"cosmic":2.23,"average":2.23},"novel":{"1000genomes":1.48,"dbsnp":1.48,"cosmic":1.48,"average":1.4799999999999998}},"value":2.23}],"plots":[],"html_fpath":"../../../GPP16188/qc/varQC/GPP16188-vardict.varQC.json"},{"metric_storage":null,"display_name":"GPP17600 freebayes","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP17600","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17600","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17600/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":460,"dbsnp":460,"cosmic":460,"average":460.0},"all":{"1000genomes":484,"dbsnp":484,"cosmic":484,"average":484.0},"novel":{"1000genomes":24,"dbsnp":24,"cosmic":24,"average":24.0}},"value":484},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":428,"dbsnp":428,"cosmic":428,"average":428.0},"all":{"1000genomes":447,"dbsnp":447,"cosmic":447,"average":447.0},"novel":{"1000genomes":19,"dbsnp":19,"cosmic":19,"average":19.0}},"value":447},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":12,"dbsnp":12,"cosmic":12,"average":12.0},"all":{"1000genomes":13,"dbsnp":13,"cosmic":13,"average":13.0},"novel":{"1000genomes":1,"dbsnp":1,"cosmic":1,"average":1.0}},"value":13},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":20,"dbsnp":20,"cosmic":20,"average":20.0},"all":{"1000genomes":24,"dbsnp":24,"cosmic":24,"average":24.0},"novel":{"1000genomes":4,"dbsnp":4,"cosmic":4,"average":4.0}},"value":24},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":193,"dbsnp":460,"cosmic":45,"average":232.66666666666666},"all":{"1000genomes":193,"dbsnp":460,"cosmic":48,"average":233.66666666666666},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":48},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":41.96,"dbsnp":100.0,"cosmic":9.78,"average":50.580000000000005},"all":{"1000genomes":39.88,"dbsnp":95.04,"cosmic":9.92,"average":48.28},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":12.5,"average":4.166666666666667}},"value":9.92},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":193,"dbsnp":460,"cosmic":43,"average":232.0},"all":{"1000genomes":193,"dbsnp":460,"cosmic":46,"average":233.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":46},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":100.0,"cosmic":95.56,"average":98.52},"all":{"1000genomes":100.0,"dbsnp":100.0,"cosmic":95.83,"average":98.61},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":100.0,"average":33.333333333333336}},"value":95.83},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.5e-07,"dbsnp":1.5e-07,"cosmic":1.5e-07,"average":1.5e-07},"all":{"1000genomes":1.5e-07,"dbsnp":1.5e-07,"cosmic":1.5e-07,"average":1.5e-07},"novel":{"1000genomes":1e-08,"dbsnp":1e-08,"cosmic":1e-08,"average":1.0000000000000002e-08}},"value":1.5e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":6819915.0,"dbsnp":6819915.0,"cosmic":6819915.0,"average":6819915.0},"all":{"1000genomes":6481738.0,"dbsnp":6481738.0,"cosmic":6481738.0,"average":6481738.0},"novel":{"1000genomes":130715052.0,"dbsnp":130715052.0,"cosmic":130715052.0,"average":130715052.0}},"value":6481738.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":0.41,"dbsnp":0.41,"cosmic":0.41,"average":0.41},"all":{"1000genomes":0.45,"dbsnp":0.45,"cosmic":0.45,"average":0.45},"novel":{"1000genomes":2.43,"dbsnp":2.43,"cosmic":2.43,"average":2.43}},"value":0.45},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.51,"dbsnp":2.51,"cosmic":2.51,"average":2.51},"all":{"1000genomes":2.26,"dbsnp":2.26,"cosmic":2.26,"average":2.26},"novel":{"1000genomes":0.27,"dbsnp":0.27,"cosmic":0.27,"average":0.27}},"value":2.26}],"plots":[],"html_fpath":"../../../GPP17600/qc/varQC/GPP17600-freebayes.varQC.json"},{"metric_storage":null,"display_name":"GPP17600 mutect","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP17600","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17600","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17600/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":250,"dbsnp":250,"cosmic":250,"average":250.0},"all":{"1000genomes":410,"dbsnp":410,"cosmic":410,"average":410.0},"novel":{"1000genomes":160,"dbsnp":160,"cosmic":160,"average":160.0}},"value":410},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":239,"dbsnp":239,"cosmic":239,"average":239.0},"all":{"1000genomes":398,"dbsnp":398,"cosmic":398,"average":398.0},"novel":{"1000genomes":159,"dbsnp":159,"cosmic":159,"average":159.0}},"value":398},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":3,"dbsnp":3,"cosmic":3,"average":3.0},"all":{"1000genomes":4,"dbsnp":4,"cosmic":4,"average":4.0},"novel":{"1000genomes":1,"dbsnp":1,"cosmic":1,"average":1.0}},"value":4},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":8,"dbsnp":8,"cosmic":8,"average":8.0},"all":{"1000genomes":8,"dbsnp":8,"cosmic":8,"average":8.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":8},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":83,"dbsnp":250,"cosmic":33,"average":122.0},"all":{"1000genomes":83,"dbsnp":250,"cosmic":45,"average":126.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":12,"average":4.0}},"value":45},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":33.2,"dbsnp":100.0,"cosmic":13.2,"average":48.79999999999999},"all":{"1000genomes":20.24,"dbsnp":60.98,"cosmic":10.98,"average":30.733333333333334},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":7.5,"average":2.5}},"value":10.98},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":82,"dbsnp":240,"cosmic":31,"average":117.66666666666667},"all":{"1000genomes":82,"dbsnp":240,"cosmic":40,"average":120.66666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":9,"average":3.0}},"value":40},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":98.8,"dbsnp":96.0,"cosmic":93.94,"average":96.24666666666667},"all":{"1000genomes":98.8,"dbsnp":96.0,"cosmic":88.89,"average":94.56333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":75.0,"average":25.0}},"value":88.89},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":8e-08,"dbsnp":8e-08,"cosmic":8e-08,"average":8.000000000000001e-08},"all":{"1000genomes":1.3e-07,"dbsnp":1.3e-07,"cosmic":1.3e-07,"average":1.3e-07},"novel":{"1000genomes":5e-08,"dbsnp":5e-08,"cosmic":5e-08,"average":5e-08}},"value":1.3e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":12548645.0,"dbsnp":12548645.0,"cosmic":12548645.0,"average":12548645.0},"all":{"1000genomes":7651612.0,"dbsnp":7651612.0,"cosmic":7651612.0,"average":7651612.0},"novel":{"1000genomes":19607257.0,"dbsnp":19607257.0,"cosmic":19607257.0,"average":19607257.0}},"value":7651612.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":61.5,"dbsnp":61.5,"cosmic":61.5,"average":61.5},"all":{"1000genomes":81.0,"dbsnp":81.0,"cosmic":81.0,"average":81.0},"novel":{"1000genomes":159.0,"dbsnp":159.0,"cosmic":159.0,"average":159.0}},"value":81.0},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.37,"dbsnp":2.37,"cosmic":2.37,"average":2.37},"all":{"1000genomes":1.29,"dbsnp":1.29,"cosmic":1.29,"average":1.29},"novel":{"1000genomes":0.54,"dbsnp":0.54,"cosmic":0.54,"average":0.54}},"value":1.29}],"plots":[],"html_fpath":"../../../GPP17600/qc/varQC/GPP17600-mutect.varQC.json"},{"metric_storage":null,"display_name":"GPP17600 vardict","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP17600","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17600","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17600/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":488,"dbsnp":488,"cosmic":488,"average":488.0},"all":{"1000genomes":659,"dbsnp":659,"cosmic":659,"average":659.0},"novel":{"1000genomes":171,"dbsnp":171,"cosmic":171,"average":171.0}},"value":659},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":445,"dbsnp":445,"cosmic":445,"average":445.0},"all":{"1000genomes":561,"dbsnp":561,"cosmic":561,"average":561.0},"novel":{"1000genomes":116,"dbsnp":116,"cosmic":116,"average":116.0}},"value":561},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":14,"dbsnp":14,"cosmic":14,"average":14.0},"all":{"1000genomes":28,"dbsnp":28,"cosmic":28,"average":28.0},"novel":{"1000genomes":14,"dbsnp":14,"cosmic":14,"average":14.0}},"value":28},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":29,"dbsnp":29,"cosmic":29,"average":29.0},"all":{"1000genomes":70,"dbsnp":70,"cosmic":70,"average":70.0},"novel":{"1000genomes":41,"dbsnp":41,"cosmic":41,"average":41.0}},"value":70},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":196,"dbsnp":488,"cosmic":55,"average":246.33333333333334},"all":{"1000genomes":196,"dbsnp":488,"cosmic":76,"average":253.33333333333334},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":21,"average":7.0}},"value":76},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":40.16,"dbsnp":100.0,"cosmic":11.27,"average":50.47666666666667},"all":{"1000genomes":29.74,"dbsnp":74.05,"cosmic":11.53,"average":38.44},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":12.28,"average":4.093333333333333}},"value":11.53},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":196,"dbsnp":482,"cosmic":52,"average":243.33333333333334},"all":{"1000genomes":196,"dbsnp":482,"cosmic":71,"average":249.66666666666666},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":19,"average":6.333333333333333}},"value":71},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":98.77,"cosmic":94.55,"average":97.77333333333333},"all":{"1000genomes":100.0,"dbsnp":98.77,"cosmic":93.42,"average":97.39666666666666},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":90.48,"average":30.16}},"value":93.42},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.6e-07,"dbsnp":1.6e-07,"cosmic":1.6e-07,"average":1.6000000000000003e-07},"all":{"1000genomes":2.1e-07,"dbsnp":2.1e-07,"cosmic":2.1e-07,"average":2.1e-07},"novel":{"1000genomes":5e-08,"dbsnp":5e-08,"cosmic":5e-08,"average":5e-08}},"value":2.1e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":6428609.0,"dbsnp":6428609.0,"cosmic":6428609.0,"average":6428609.0},"all":{"1000genomes":4760487.0,"dbsnp":4760487.0,"cosmic":4760487.0,"average":4760487.0},"novel":{"1000genomes":18345972.0,"dbsnp":18345972.0,"cosmic":18345972.0,"average":18345972.0}},"value":4760487.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":0.52,"dbsnp":0.52,"cosmic":0.52,"average":0.52},"all":{"1000genomes":0.99,"dbsnp":0.99,"cosmic":0.99,"average":0.9899999999999999},"novel":{"1000genomes":18.0,"dbsnp":18.0,"cosmic":18.0,"average":18.0}},"value":0.99},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.71,"dbsnp":2.71,"cosmic":2.71,"average":2.7099999999999995},"all":{"1000genomes":1.98,"dbsnp":1.98,"cosmic":1.98,"average":1.9799999999999998},"novel":{"1000genomes":0.71,"dbsnp":0.71,"cosmic":0.71,"average":0.71}},"value":1.98}],"plots":[],"html_fpath":"../../../GPP17600/qc/varQC/GPP17600-vardict.varQC.json"},{"metric_storage":null,"display_name":"GPP17601 freebayes","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP17601","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17601","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17601/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":513,"dbsnp":513,"cosmic":513,"average":513.0},"all":{"1000genomes":536,"dbsnp":536,"cosmic":536,"average":536.0},"novel":{"1000genomes":23,"dbsnp":23,"cosmic":23,"average":23.0}},"value":536},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":484,"dbsnp":484,"cosmic":484,"average":484.0},"all":{"1000genomes":500,"dbsnp":500,"cosmic":500,"average":500.0},"novel":{"1000genomes":16,"dbsnp":16,"cosmic":16,"average":16.0}},"value":500},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":11,"dbsnp":11,"cosmic":11,"average":11.0},"all":{"1000genomes":12,"dbsnp":12,"cosmic":12,"average":12.0},"novel":{"1000genomes":1,"dbsnp":1,"cosmic":1,"average":1.0}},"value":12},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":17,"dbsnp":17,"cosmic":17,"average":17.0},"all":{"1000genomes":23,"dbsnp":23,"cosmic":23,"average":23.0},"novel":{"1000genomes":6,"dbsnp":6,"cosmic":6,"average":6.0}},"value":23},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":227,"dbsnp":513,"cosmic":50,"average":263.3333333333333},"all":{"1000genomes":227,"dbsnp":513,"cosmic":53,"average":264.3333333333333},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":53},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":44.25,"dbsnp":100.0,"cosmic":9.75,"average":51.333333333333336},"all":{"1000genomes":42.35,"dbsnp":95.71,"cosmic":9.89,"average":49.31666666666666},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":13.04,"average":4.346666666666667}},"value":9.89},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":227,"dbsnp":511,"cosmic":48,"average":262.0},"all":{"1000genomes":227,"dbsnp":511,"cosmic":51,"average":263.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":51},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":99.61,"cosmic":96.0,"average":98.53666666666668},"all":{"1000genomes":100.0,"dbsnp":99.61,"cosmic":96.23,"average":98.61333333333334},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":100.0,"average":33.333333333333336}},"value":96.23},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.6e-07,"dbsnp":1.6e-07,"cosmic":1.6e-07,"average":1.6000000000000003e-07},"all":{"1000genomes":1.7e-07,"dbsnp":1.7e-07,"cosmic":1.7e-07,"average":1.7e-07},"novel":{"1000genomes":1e-08,"dbsnp":1e-08,"cosmic":1e-08,"average":1.0000000000000002e-08}},"value":1.7e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":6115324.0,"dbsnp":6115324.0,"cosmic":6115324.0,"average":6115324.0},"all":{"1000genomes":5852912.0,"dbsnp":5852912.0,"cosmic":5852912.0,"average":5852912.0},"novel":{"1000genomes":136398315.0,"dbsnp":136398315.0,"cosmic":136398315.0,"average":136398315.0}},"value":5852912.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.44,"dbsnp":1.44,"cosmic":1.44,"average":1.4400000000000002},"all":{"1000genomes":1.54,"dbsnp":1.54,"cosmic":1.54,"average":1.54},"novel":{"1000genomes":22.0,"dbsnp":22.0,"cosmic":22.0,"average":22.0}},"value":1.54},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.41,"dbsnp":2.41,"cosmic":2.41,"average":2.41},"all":{"1000genomes":2.31,"dbsnp":2.31,"cosmic":2.31,"average":2.31},"novel":{"1000genomes":0.78,"dbsnp":0.78,"cosmic":0.78,"average":0.7799999999999999}},"value":2.31}],"plots":[],"html_fpath":"../../../GPP17601/qc/varQC/GPP17601-freebayes.varQC.json"},{"metric_storage":null,"display_name":"GPP17601 mutect","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP17601","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17601","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17601/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":325,"dbsnp":325,"cosmic":325,"average":325.0},"all":{"1000genomes":810,"dbsnp":810,"cosmic":810,"average":810.0},"novel":{"1000genomes":485,"dbsnp":485,"cosmic":485,"average":485.0}},"value":810},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":321,"dbsnp":321,"cosmic":321,"average":321.0},"all":{"1000genomes":805,"dbsnp":805,"cosmic":805,"average":805.0},"novel":{"1000genomes":484,"dbsnp":484,"cosmic":484,"average":484.0}},"value":805},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":4,"dbsnp":4,"cosmic":4,"average":4.0},"all":{"1000genomes":4,"dbsnp":4,"cosmic":4,"average":4.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":4},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":1,"dbsnp":1,"cosmic":1,"average":1.0},"novel":{"1000genomes":1,"dbsnp":1,"cosmic":1,"average":1.0}},"value":1},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":90,"dbsnp":325,"cosmic":38,"average":151.0},"all":{"1000genomes":90,"dbsnp":325,"cosmic":89,"average":168.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":51,"average":17.0}},"value":89},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":27.69,"dbsnp":100.0,"cosmic":11.69,"average":46.46},"all":{"1000genomes":11.11,"dbsnp":40.12,"cosmic":10.99,"average":20.74},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":10.52,"average":3.5066666666666664}},"value":10.99},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":90,"dbsnp":301,"cosmic":36,"average":142.33333333333334},"all":{"1000genomes":90,"dbsnp":301,"cosmic":75,"average":155.33333333333334},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":39,"average":13.0}},"value":75},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":92.62,"cosmic":94.74,"average":95.78666666666668},"all":{"1000genomes":100.0,"dbsnp":92.62,"cosmic":84.27,"average":92.29666666666667},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":76.47,"average":25.49}},"value":84.27},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1e-07,"dbsnp":1e-07,"cosmic":1e-07,"average":1e-07},"all":{"1000genomes":2.6e-07,"dbsnp":2.6e-07,"cosmic":2.6e-07,"average":2.6e-07},"novel":{"1000genomes":1.5e-07,"dbsnp":1.5e-07,"cosmic":1.5e-07,"average":1.5e-07}},"value":2.6e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":9652803.0,"dbsnp":9652803.0,"cosmic":9652803.0,"average":9652803.0},"all":{"1000genomes":3873038.0,"dbsnp":3873038.0,"cosmic":3873038.0,"average":3873038.0},"novel":{"1000genomes":6468373.0,"dbsnp":6468373.0,"cosmic":6468373.0,"average":6468373.0}},"value":3873038.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":325.0,"dbsnp":325.0,"cosmic":325.0,"average":325.0},"all":{"1000genomes":810.0,"dbsnp":810.0,"cosmic":810.0,"average":810.0},"novel":{"1000genomes":485.0,"dbsnp":485.0,"cosmic":485.0,"average":485.0}},"value":810.0},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.69,"dbsnp":2.69,"cosmic":2.69,"average":2.69},"all":{"1000genomes":1.57,"dbsnp":1.57,"cosmic":1.57,"average":1.57},"novel":{"1000genomes":1.14,"dbsnp":1.14,"cosmic":1.14,"average":1.14}},"value":1.57}],"plots":[],"html_fpath":"../../../GPP17601/qc/varQC/GPP17601-mutect.varQC.json"},{"metric_storage":null,"display_name":"GPP17601 vardict","json_fpath":null,"report_name":"","sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"GPP17601","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17601","vcf_by_callername":{"mutect":null,"freebayes":null,"vardict":null},"var_dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/EXT_021_Courtagen_Faslodex_HaloPlex/final/GPP17601/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":659,"dbsnp":659,"cosmic":659,"average":659.0},"all":{"1000genomes":1352,"dbsnp":1352,"cosmic":1352,"average":1352.0},"novel":{"1000genomes":693,"dbsnp":693,"cosmic":693,"average":693.0}},"value":1352},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":616,"dbsnp":616,"cosmic":616,"average":616.0},"all":{"1000genomes":1228,"dbsnp":1228,"cosmic":1228,"average":1228.0},"novel":{"1000genomes":612,"dbsnp":612,"cosmic":612,"average":612.0}},"value":1228},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":12,"dbsnp":12,"cosmic":12,"average":12.0},"all":{"1000genomes":53,"dbsnp":53,"cosmic":53,"average":53.0},"novel":{"1000genomes":41,"dbsnp":41,"cosmic":41,"average":41.0}},"value":53},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":31,"dbsnp":31,"cosmic":31,"average":31.0},"all":{"1000genomes":70,"dbsnp":70,"cosmic":70,"average":70.0},"novel":{"1000genomes":39,"dbsnp":39,"cosmic":39,"average":39.0}},"value":70},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":241,"dbsnp":659,"cosmic":81,"average":327.0},"all":{"1000genomes":241,"dbsnp":659,"cosmic":159,"average":353.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":78,"average":26.0}},"value":159},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":36.57,"dbsnp":100.0,"cosmic":12.29,"average":49.62},"all":{"1000genomes":17.83,"dbsnp":48.74,"cosmic":11.76,"average":26.11},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":11.26,"average":3.7533333333333334}},"value":11.76},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":240,"dbsnp":633,"cosmic":74,"average":315.6666666666667},"all":{"1000genomes":240,"dbsnp":633,"cosmic":140,"average":337.6666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":66,"average":22.0}},"value":140},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":99.59,"dbsnp":96.05,"cosmic":91.36,"average":95.66666666666667},"all":{"1000genomes":99.59,"dbsnp":96.05,"cosmic":88.05,"average":94.56333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":84.62,"average":28.206666666666667}},"value":88.05},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.1e-07,"dbsnp":2.1e-07,"cosmic":2.1e-07,"average":2.1e-07},"all":{"1000genomes":4.3e-07,"dbsnp":4.3e-07,"cosmic":4.3e-07,"average":4.2999999999999996e-07},"novel":{"1000genomes":2.2e-07,"dbsnp":2.2e-07,"cosmic":2.2e-07,"average":2.2e-07}},"value":4.3e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":4760487.0,"dbsnp":4760487.0,"cosmic":4760487.0,"average":4760487.0},"all":{"1000genomes":2320385.0,"dbsnp":2320385.0,"cosmic":2320385.0,"average":2320385.0},"novel":{"1000genomes":4526928.0,"dbsnp":4526928.0,"cosmic":4526928.0,"average":4526928.0}},"value":2320385.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":2.17,"dbsnp":2.17,"cosmic":2.17,"average":2.17},"all":{"1000genomes":5.06,"dbsnp":5.06,"cosmic":5.06,"average":5.06},"novel":{"1000genomes":45.2,"dbsnp":45.2,"cosmic":45.2,"average":45.20000000000001}},"value":5.06},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.76,"dbsnp":2.76,"cosmic":2.76,"average":2.76},"all":{"1000genomes":1.67,"dbsnp":1.67,"cosmic":1.67,"average":1.67},"novel":{"1000genomes":1.07,"dbsnp":1.07,"cosmic":1.07,"average":1.07}},"value":1.67}],"plots":[],"html_fpath":"../../../GPP17601/qc/varQC/GPP17601-vardict.varQC.json"}],"name":""},"data_outside_reports":{},"date":"26 September 2014, Friday, 19:49:38"}
            </div>
        </div>
    </div>
</div>
</body>
</html>
