<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/table_sorter/style.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="html_aux/jquery.align-column.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>

    <script type="text/javascript" src="html_aux/dragtable.js"></script>
    <script type="text/javascript" src="html_aux/table_sorter/tsort.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });

            $('table.tableSorter').tableSort();
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for mutect</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"report":{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"sample_reports":[{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-1","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-ready.bam","filtered_vcf_by_callername":{},"name":"Tissue-Resistant-1","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-1/PBMC-Resistant-1-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Resistant-1","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-1","vcf_by_callername":{},"filtered_clean_vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-1/var"},"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/var/Tissue-Resistant-1-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/var/Tissue-Resistant-1-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/var/Tissue-Resistant-1-vardict.vcf"},"filtered_clean_vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":10649,"average":10649.0,"cosmic":10649,"dbsnp":10649},"all":{"1000genomes":183906,"average":183906.0,"cosmic":183906,"dbsnp":183906},"novel":{"1000genomes":173257,"average":173257.0,"cosmic":173257,"dbsnp":173257}},"value":183906},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":10649,"average":10649.0,"cosmic":10649,"dbsnp":10649},"all":{"1000genomes":183906,"average":183906.0,"cosmic":183906,"dbsnp":183906},"novel":{"1000genomes":173257,"average":173257.0,"cosmic":173257,"dbsnp":173257}},"value":183906},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":126,"average":3955.0,"cosmic":1090,"dbsnp":10649},"all":{"1000genomes":126,"average":6470.333333333333,"cosmic":8636,"dbsnp":10649},"novel":{"1000genomes":0,"average":2515.3333333333335,"cosmic":7546,"dbsnp":0}},"value":8636},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":1.18,"average":37.14,"cosmic":10.24,"dbsnp":100.0},"all":{"1000genomes":0.07,"average":3.52,"cosmic":4.7,"dbsnp":5.79},"novel":{"1000genomes":0.0,"average":1.4533333333333334,"cosmic":4.36,"dbsnp":0.0}},"value":4.7},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":13,"average":475.3333333333333,"cosmic":195,"dbsnp":1218},"all":{"1000genomes":13,"average":1296.3333333333333,"cosmic":2658,"dbsnp":1218},"novel":{"1000genomes":0,"average":821.0,"cosmic":2463,"dbsnp":0}},"value":2658},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":10.32,"average":13.216666666666667,"cosmic":17.89,"dbsnp":11.44},"all":{"1000genomes":10.32,"average":17.513333333333332,"cosmic":30.78,"dbsnp":11.44},"novel":{"1000genomes":0.0,"average":10.88,"cosmic":32.64,"dbsnp":0.0}},"value":30.78},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":3.39e-06,"average":3.39e-06,"cosmic":3.39e-06,"dbsnp":3.39e-06},"all":{"1000genomes":5.862e-05,"average":5.862e-05,"cosmic":5.862e-05,"dbsnp":5.862e-05},"novel":{"1000genomes":5.523e-05,"average":5.523e-05,"cosmic":5.523e-05,"dbsnp":5.523e-05}},"value":5.862e-05},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":294596.0,"average":294596.0,"cosmic":294596.0,"dbsnp":294596.0},"all":{"1000genomes":17058.0,"average":17058.0,"cosmic":17058.0,"dbsnp":17058.0},"novel":{"1000genomes":18106.0,"average":18106.0,"cosmic":18106.0,"dbsnp":18106.0}},"value":17058.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":10649.0,"average":10649.0,"cosmic":10649.0,"dbsnp":10649.0},"all":{"1000genomes":183906.0,"average":183906.0,"cosmic":183906.0,"dbsnp":183906.0},"novel":{"1000genomes":173257.0,"average":173257.0,"cosmic":173257.0,"dbsnp":173257.0}},"value":183906.0},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0}],"plots":[],"html_fpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/qc/varQC/Tissue-Resistant-1-mutect.varQC.html"},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-2","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-2/Tissue-Resistant-2-ready.bam","filtered_vcf_by_callername":{},"name":"Tissue-Resistant-2","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-2/PBMC-Resistant-2-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Resistant-2","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-2","vcf_by_callername":{},"filtered_clean_vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-2/var"},"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-2","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-2/var/Tissue-Resistant-2-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-2/var/Tissue-Resistant-2-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-2/var/Tissue-Resistant-2-vardict.vcf"},"filtered_clean_vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-2/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":6511,"average":6511.0,"cosmic":6511,"dbsnp":6511},"all":{"1000genomes":113193,"average":113193.0,"cosmic":113193,"dbsnp":113193},"novel":{"1000genomes":106682,"average":106682.0,"cosmic":106682,"dbsnp":106682}},"value":113193},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":6511,"average":6511.0,"cosmic":6511,"dbsnp":6511},"all":{"1000genomes":113193,"average":113193.0,"cosmic":113193,"dbsnp":113193},"novel":{"1000genomes":106682,"average":106682.0,"cosmic":106682,"dbsnp":106682}},"value":113193},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"all":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":64,"average":2404.6666666666665,"cosmic":639,"dbsnp":6511},"all":{"1000genomes":64,"average":3903.0,"cosmic":5134,"dbsnp":6511},"novel":{"1000genomes":0,"average":1498.3333333333333,"cosmic":4495,"dbsnp":0}},"value":5134},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":0.98,"average":36.93,"cosmic":9.81,"dbsnp":100.0},"all":{"1000genomes":0.06,"average":3.4499999999999997,"cosmic":4.54,"dbsnp":5.75},"novel":{"1000genomes":0.0,"average":1.4033333333333333,"cosmic":4.21,"dbsnp":0.0}},"value":4.54},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":8,"average":288.3333333333333,"cosmic":127,"dbsnp":730},"all":{"1000genomes":8,"average":780.6666666666666,"cosmic":1604,"dbsnp":730},"novel":{"1000genomes":0,"average":492.3333333333333,"cosmic":1477,"dbsnp":0}},"value":1604},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":12.5,"average":14.526666666666666,"cosmic":19.87,"dbsnp":11.21},"all":{"1000genomes":12.5,"average":18.316666666666666,"cosmic":31.24,"dbsnp":11.21},"novel":{"1000genomes":0.0,"average":10.953333333333333,"cosmic":32.86,"dbsnp":0.0}},"value":31.24},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.08e-06,"average":2.08e-06,"cosmic":2.08e-06,"dbsnp":2.08e-06},"all":{"1000genomes":3.608e-05,"average":3.608e-05,"cosmic":3.608e-05,"dbsnp":3.608e-05},"novel":{"1000genomes":3.401e-05,"average":3.401e-05,"cosmic":3.401e-05,"dbsnp":3.401e-05}},"value":3.608e-05},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":481824.0,"average":481824.0,"cosmic":481824.0,"dbsnp":481824.0},"all":{"1000genomes":27715.0,"average":27715.0,"cosmic":27715.0,"dbsnp":27715.0},"novel":{"1000genomes":29406.0,"average":29406.0,"cosmic":29406.0,"dbsnp":29406.0}},"value":27715.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":6511.0,"average":6511.0,"cosmic":6511.0,"dbsnp":6511.0},"all":{"1000genomes":113193.0,"average":113193.0,"cosmic":113193.0,"dbsnp":113193.0},"novel":{"1000genomes":106682.0,"average":106682.0,"cosmic":106682.0,"dbsnp":106682.0}},"value":113193.0},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0}],"plots":[],"html_fpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-2/qc/varQC/Tissue-Resistant-2-mutect.varQC.html"},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Sensitive-9","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-9/Tissue-Sensitive-9-ready.bam","filtered_vcf_by_callername":{},"name":"Tissue-Sensitive-9","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-9/PBMC-Sensitive-9-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Sensitive-9","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-9","vcf_by_callername":{},"filtered_clean_vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-9/var"},"phenotype":"tumor","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-9","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-9/var/Tissue-Sensitive-9-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-9/var/Tissue-Sensitive-9-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-9/var/Tissue-Sensitive-9-vardict.vcf"},"filtered_clean_vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-9/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":9350,"average":9350.0,"cosmic":9350,"dbsnp":9350},"all":{"1000genomes":9828,"average":9828.0,"cosmic":9828,"dbsnp":9828},"novel":{"1000genomes":478,"average":478.0,"cosmic":478,"dbsnp":478}},"value":9828},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":8890,"average":8890.0,"cosmic":8890,"dbsnp":8890},"all":{"1000genomes":9285,"average":9285.0,"cosmic":9285,"dbsnp":9285},"novel":{"1000genomes":395,"average":395.0,"cosmic":395,"dbsnp":395}},"value":9285},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":185,"average":185.0,"cosmic":185,"dbsnp":185},"all":{"1000genomes":192,"average":192.0,"cosmic":192,"dbsnp":192},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":192},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":269,"average":269.0,"cosmic":269,"dbsnp":269},"all":{"1000genomes":302,"average":302.0,"cosmic":302,"dbsnp":302},"novel":{"1000genomes":33,"average":33.0,"cosmic":33,"dbsnp":33}},"value":302},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":4598,"average":4999.666666666667,"cosmic":1051,"dbsnp":9350},"all":{"1000genomes":4603,"average":5007.666666666667,"cosmic":1070,"dbsnp":9350},"novel":{"1000genomes":5,"average":8.0,"cosmic":19,"dbsnp":0}},"value":1070},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":49.18,"average":53.473333333333336,"cosmic":11.24,"dbsnp":100.0},"all":{"1000genomes":46.84,"average":50.95666666666667,"cosmic":10.89,"dbsnp":95.14},"novel":{"1000genomes":1.05,"average":1.6733333333333336,"cosmic":3.97,"dbsnp":0.0}},"value":10.89},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":4595,"average":4978.333333333333,"cosmic":1022,"dbsnp":9318},"all":{"1000genomes":4600,"average":4984.333333333333,"cosmic":1035,"dbsnp":9318},"novel":{"1000genomes":5,"average":6.0,"cosmic":13,"dbsnp":0}},"value":1035},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":99.93,"average":98.94333333333333,"cosmic":97.24,"dbsnp":99.66},"all":{"1000genomes":99.93,"average":98.77333333333333,"cosmic":96.73,"dbsnp":99.66},"novel":{"1000genomes":100.0,"average":56.14000000000001,"cosmic":68.42,"dbsnp":0.0}},"value":96.73},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.98e-06,"average":2.98e-06,"cosmic":2.98e-06,"dbsnp":2.98e-06},"all":{"1000genomes":3.13e-06,"average":3.13e-06,"cosmic":3.13e-06,"dbsnp":3.13e-06},"novel":{"1000genomes":1.5e-07,"average":1.5e-07,"cosmic":1.5e-07,"dbsnp":1.5e-07}},"value":3.13e-06},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":335525.0,"average":335525.0,"cosmic":335525.0,"dbsnp":335525.0},"all":{"1000genomes":319206.0,"average":319206.0,"cosmic":319206.0,"dbsnp":319206.0},"novel":{"1000genomes":6563098.0,"average":6563098.0,"cosmic":6563098.0,"dbsnp":6563098.0}},"value":319206.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":4.74,"average":4.74,"cosmic":4.74,"dbsnp":4.74},"all":{"1000genomes":4.85,"average":4.85,"cosmic":4.85,"dbsnp":4.85},"novel":{"1000genomes":11.54,"average":11.54,"cosmic":11.54,"dbsnp":11.54}},"value":4.85},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.69,"average":2.69,"cosmic":2.69,"dbsnp":2.69},"all":{"1000genomes":2.48,"average":2.48,"cosmic":2.48,"dbsnp":2.48},"novel":{"1000genomes":0.54,"average":0.54,"cosmic":0.54,"dbsnp":0.54}},"value":2.48}],"plots":[],"html_fpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-9/qc/varQC/Tissue-Sensitive-9-mutect.varQC.html"}],"name":""},"data_outside_reports":{},"date":"02 September 2014, Tuesday, 09:51:49"}
            </div>
        </div>
    </div>
</div>
</body>
</html>
