<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>
    <script type="text/javascript" src="html_aux/dragtable.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for mutect</h1>
        <p id='report_date'></p>
        <p id="extended_link"></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"report":[{"sample":{"name":"Tissue-Resistant-10","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-10","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-10/Tissue-Resistant-10-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":12062,"average":12062.0,"cosmic":12062,"dbsnp":12062},"known":{"1000genomes":929,"average":929.0,"cosmic":929,"dbsnp":929},"novel":{"1000genomes":11133,"average":11133.0,"cosmic":11133,"dbsnp":11133}},"value":12062},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":11855,"average":11855.0,"cosmic":11855,"dbsnp":11855},"known":{"1000genomes":755,"average":755.0,"cosmic":755,"dbsnp":755},"novel":{"1000genomes":11100,"average":11100.0,"cosmic":11100,"dbsnp":11100}},"value":11855},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":78,"average":78.0,"cosmic":78,"dbsnp":78},"known":{"1000genomes":69,"average":69.0,"cosmic":69,"dbsnp":69},"novel":{"1000genomes":9,"average":9.0,"cosmic":9,"dbsnp":9}},"value":78},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":121,"average":121.0,"cosmic":121,"dbsnp":121},"known":{"1000genomes":97,"average":97.0,"cosmic":97,"dbsnp":97},"novel":{"1000genomes":24,"average":24.0,"cosmic":24,"dbsnp":24}},"value":121},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":25,"average":490.3333333333333,"cosmic":517,"dbsnp":929},"known":{"1000genomes":25,"average":343.6666666666667,"cosmic":77,"dbsnp":929},"novel":{"1000genomes":0,"average":146.66666666666666,"cosmic":440,"dbsnp":0}},"value":517},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.21,"average":4.066666666666666,"cosmic":4.29,"dbsnp":7.7},"known":{"1000genomes":2.69,"average":36.99333333333333,"cosmic":8.29,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.3166666666666667,"cosmic":3.95,"dbsnp":0.0}},"value":4.29},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":5,"average":156.33333333333334,"cosmic":175,"dbsnp":289},"known":{"1000genomes":5,"average":107.66666666666667,"cosmic":29,"dbsnp":289},"novel":{"1000genomes":0,"average":48.666666666666664,"cosmic":146,"dbsnp":0}},"value":175},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":20.0,"average":28.320000000000004,"cosmic":33.85,"dbsnp":31.11},"known":{"1000genomes":20.0,"average":29.59,"cosmic":37.66,"dbsnp":31.11},"novel":{"1000genomes":0.0,"average":11.06,"cosmic":33.18,"dbsnp":0.0}},"value":33.85},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":3.84e-06,"average":3.84e-06,"cosmic":3.84e-06,"dbsnp":3.84e-06},"known":{"1000genomes":3e-07,"average":3e-07,"cosmic":3e-07,"dbsnp":3e-07},"novel":{"1000genomes":3.55e-06,"average":3.55e-06,"cosmic":3.55e-06,"dbsnp":3.55e-06}},"value":3.84e-06},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":260086.0,"average":260086.0,"cosmic":260086.0,"dbsnp":260086.0},"known":{"1000genomes":3376922.0,"average":3376922.0,"cosmic":3376922.0,"dbsnp":3376922.0},"novel":{"1000genomes":281789.0,"average":281789.0,"cosmic":281789.0,"dbsnp":281789.0}},"value":260086.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":127.32,"average":127.32,"cosmic":127.32,"dbsnp":127.32},"known":{"1000genomes":9.8,"average":9.8,"cosmic":9.8,"dbsnp":9.8},"novel":{"1000genomes":1390.63,"average":1390.63,"cosmic":1390.63,"dbsnp":1390.63}},"value":127.32},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01},"known":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04},"novel":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01}},"value":0.01}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-10/qc/varQC/Tissue-Resistant-10-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-10/qc/varQC/Tissue-Resistant-10-mutect.varQC.json"},{"sample":{"name":"Tissue-Resistant-1","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-1","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":147740,"average":147740.0,"cosmic":147740,"dbsnp":147740},"known":{"1000genomes":9029,"average":9029.0,"cosmic":9029,"dbsnp":9029},"novel":{"1000genomes":138711,"average":138711.0,"cosmic":138711,"dbsnp":138711}},"value":147740},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":147492,"average":147492.0,"cosmic":147492,"dbsnp":147492},"known":{"1000genomes":8825,"average":8825.0,"cosmic":8825,"dbsnp":8825},"novel":{"1000genomes":138667,"average":138667.0,"cosmic":138667,"dbsnp":138667}},"value":147492},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":85,"average":85.0,"cosmic":85,"dbsnp":85},"known":{"1000genomes":74,"average":74.0,"cosmic":74,"dbsnp":74},"novel":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11}},"value":85},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":145,"average":145.0,"cosmic":145,"dbsnp":145},"known":{"1000genomes":113,"average":113.0,"cosmic":113,"dbsnp":113},"novel":{"1000genomes":32,"average":32.0,"cosmic":32,"dbsnp":32}},"value":145},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":256,"average":5241.666666666667,"cosmic":6440,"dbsnp":9029},"known":{"1000genomes":256,"average":3368.3333333333335,"cosmic":820,"dbsnp":9029},"novel":{"1000genomes":0,"average":1873.3333333333333,"cosmic":5620,"dbsnp":0}},"value":6440},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.17,"average":3.546666666666667,"cosmic":4.36,"dbsnp":6.11},"known":{"1000genomes":2.84,"average":37.306666666666665,"cosmic":9.08,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.3499999999999999,"cosmic":4.05,"dbsnp":0.0}},"value":4.36},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":20,"average":1058.3333333333333,"cosmic":1965,"dbsnp":1190},"known":{"1000genomes":20,"average":453.6666666666667,"cosmic":151,"dbsnp":1190},"novel":{"1000genomes":0,"average":604.6666666666666,"cosmic":1814,"dbsnp":0}},"value":1965},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":7.81,"average":17.166666666666668,"cosmic":30.51,"dbsnp":13.18},"known":{"1000genomes":7.81,"average":13.133333333333333,"cosmic":18.41,"dbsnp":13.18},"novel":{"1000genomes":0.0,"average":10.76,"cosmic":32.28,"dbsnp":0.0}},"value":30.51},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":4.709e-05,"average":4.708999999999999e-05,"cosmic":4.709e-05,"dbsnp":4.709e-05},"known":{"1000genomes":2.88e-06,"average":2.88e-06,"cosmic":2.88e-06,"dbsnp":2.88e-06},"novel":{"1000genomes":4.422e-05,"average":4.422e-05,"cosmic":4.422e-05,"dbsnp":4.422e-05}},"value":4.709e-05},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":21234.0,"average":21234.0,"cosmic":21234.0,"dbsnp":21234.0},"known":{"1000genomes":347453.0,"average":347453.0,"cosmic":347453.0,"dbsnp":347453.0},"novel":{"1000genomes":22616.0,"average":22616.0,"cosmic":22616.0,"dbsnp":22616.0}},"value":21234.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":1180.92,"average":1180.92,"cosmic":1180.92,"dbsnp":1180.92},"known":{"1000genomes":76.84,"average":76.84,"cosmic":76.84,"dbsnp":76.84},"novel":{"1000genomes":15411.33,"average":15411.33,"cosmic":15411.33,"dbsnp":15411.33}},"value":1180.92},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"known":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-1/qc/varQC/Tissue-Resistant-1-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-1/qc/varQC/Tissue-Resistant-1-mutect.varQC.json"},{"sample":{"name":"Tissue-Resistant-2","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-2","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-2/Tissue-Resistant-2-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":107133,"average":107133.0,"cosmic":107133,"dbsnp":107133},"known":{"1000genomes":6508,"average":6508.0,"cosmic":6508,"dbsnp":6508},"novel":{"1000genomes":100625,"average":100625.0,"cosmic":100625,"dbsnp":100625}},"value":107133},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":106954,"average":106954.0,"cosmic":106954,"dbsnp":106954},"known":{"1000genomes":6364,"average":6364.0,"cosmic":6364,"dbsnp":6364},"novel":{"1000genomes":100590,"average":100590.0,"cosmic":100590,"dbsnp":100590}},"value":106954},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":72,"average":72.0,"cosmic":72,"dbsnp":72},"known":{"1000genomes":58,"average":58.0,"cosmic":58,"dbsnp":58},"novel":{"1000genomes":14,"average":14.0,"cosmic":14,"dbsnp":14}},"value":72},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":95,"average":95.0,"cosmic":95,"dbsnp":95},"known":{"1000genomes":75,"average":75.0,"cosmic":75,"dbsnp":75},"novel":{"1000genomes":20,"average":20.0,"cosmic":20,"dbsnp":20}},"value":95},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":164,"average":3780.3333333333335,"cosmic":4669,"dbsnp":6508},"known":{"1000genomes":164,"average":2417.6666666666665,"cosmic":581,"dbsnp":6508},"novel":{"1000genomes":0,"average":1362.6666666666667,"cosmic":4088,"dbsnp":0}},"value":4669},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.15,"average":3.5266666666666673,"cosmic":4.36,"dbsnp":6.07},"known":{"1000genomes":2.52,"average":37.15,"cosmic":8.93,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.3533333333333333,"cosmic":4.06,"dbsnp":0.0}},"value":4.36},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":14,"average":762.0,"cosmic":1454,"dbsnp":818},"known":{"1000genomes":14,"average":314.6666666666667,"cosmic":112,"dbsnp":818},"novel":{"1000genomes":0,"average":447.3333333333333,"cosmic":1342,"dbsnp":0}},"value":1454},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":8.54,"average":17.416666666666668,"cosmic":31.14,"dbsnp":12.57},"known":{"1000genomes":8.54,"average":13.463333333333333,"cosmic":19.28,"dbsnp":12.57},"novel":{"1000genomes":0.0,"average":10.943333333333333,"cosmic":32.83,"dbsnp":0.0}},"value":31.14},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":3.415e-05,"average":3.415e-05,"cosmic":3.415e-05,"dbsnp":3.415e-05},"known":{"1000genomes":2.07e-06,"average":2.07e-06,"cosmic":2.07e-06,"dbsnp":2.07e-06},"novel":{"1000genomes":3.208e-05,"average":3.208e-05,"cosmic":3.208e-05,"dbsnp":3.208e-05}},"value":3.415e-05},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":29282.0,"average":29282.0,"cosmic":29282.0,"dbsnp":29282.0},"known":{"1000genomes":482046.0,"average":482046.0,"cosmic":482046.0,"dbsnp":482046.0},"novel":{"1000genomes":31176.0,"average":31176.0,"cosmic":31176.0,"dbsnp":31176.0}},"value":29282.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":1784.55,"average":1784.55,"cosmic":1784.55,"dbsnp":1784.55},"known":{"1000genomes":121.79,"average":121.79,"cosmic":121.79,"dbsnp":121.79},"novel":{"1000genomes":14374.0,"average":14374.0,"cosmic":14374.0,"dbsnp":14374.0}},"value":1784.55},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"known":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-2/qc/varQC/Tissue-Resistant-2-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-2/qc/varQC/Tissue-Resistant-2-mutect.varQC.json"},{"sample":{"name":"Tissue-Resistant-4","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-4","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":139552,"average":139552.0,"cosmic":139552,"dbsnp":139552},"known":{"1000genomes":8507,"average":8507.0,"cosmic":8507,"dbsnp":8507},"novel":{"1000genomes":131045,"average":131045.0,"cosmic":131045,"dbsnp":131045}},"value":139552},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":139375,"average":139375.0,"cosmic":139375,"dbsnp":139375},"known":{"1000genomes":8354,"average":8354.0,"cosmic":8354,"dbsnp":8354},"novel":{"1000genomes":131021,"average":131021.0,"cosmic":131021,"dbsnp":131021}},"value":139375},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":56,"average":56.0,"cosmic":56,"dbsnp":56},"known":{"1000genomes":51,"average":51.0,"cosmic":51,"dbsnp":51},"novel":{"1000genomes":5,"average":5.0,"cosmic":5,"dbsnp":5}},"value":56},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":111,"average":111.0,"cosmic":111,"dbsnp":111},"known":{"1000genomes":92,"average":92.0,"cosmic":92,"dbsnp":92},"novel":{"1000genomes":19,"average":19.0,"cosmic":19,"dbsnp":19}},"value":111},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":256,"average":5024.666666666667,"cosmic":6311,"dbsnp":8507},"known":{"1000genomes":256,"average":3186.0,"cosmic":795,"dbsnp":8507},"novel":{"1000genomes":0,"average":1838.6666666666667,"cosmic":5516,"dbsnp":0}},"value":6311},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.18,"average":3.5999999999999996,"cosmic":4.52,"dbsnp":6.1},"known":{"1000genomes":3.01,"average":37.45333333333333,"cosmic":9.35,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4033333333333333,"cosmic":4.21,"dbsnp":0.0}},"value":4.52},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":36,"average":1031.6666666666667,"cosmic":1881,"dbsnp":1178},"known":{"1000genomes":36,"average":457.3333333333333,"cosmic":158,"dbsnp":1178},"novel":{"1000genomes":0,"average":574.3333333333334,"cosmic":1723,"dbsnp":0}},"value":1881},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":14.06,"average":19.24,"cosmic":29.81,"dbsnp":13.85},"known":{"1000genomes":14.06,"average":15.926666666666668,"cosmic":19.87,"dbsnp":13.85},"novel":{"1000genomes":0.0,"average":10.413333333333332,"cosmic":31.24,"dbsnp":0.0}},"value":29.81},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":4.448e-05,"average":4.448e-05,"cosmic":4.448e-05,"dbsnp":4.448e-05},"known":{"1000genomes":2.71e-06,"average":2.71e-06,"cosmic":2.71e-06,"dbsnp":2.71e-06},"novel":{"1000genomes":4.177e-05,"average":4.177e-05,"cosmic":4.177e-05,"dbsnp":4.177e-05}},"value":4.448e-05},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":22480.0,"average":22480.0,"cosmic":22480.0,"dbsnp":22480.0},"known":{"1000genomes":368774.0,"average":368774.0,"cosmic":368774.0,"dbsnp":368774.0},"novel":{"1000genomes":23939.0,"average":23939.0,"cosmic":23939.0,"dbsnp":23939.0}},"value":22480.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":2113.42,"average":2113.42,"cosmic":2113.42,"dbsnp":2113.42},"known":{"1000genomes":140.78,"average":140.78,"cosmic":140.78,"dbsnp":140.78},"novel":{"1000genomes":21839.83,"average":21839.83,"cosmic":21839.83,"dbsnp":21839.83}},"value":2113.42},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"known":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-4/qc/varQC/Tissue-Resistant-4-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-4/qc/varQC/Tissue-Resistant-4-mutect.varQC.json"},{"sample":{"name":"Tissue-Resistant-5","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-5","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1102,"average":1102.0,"cosmic":1102,"dbsnp":1102},"known":{"1000genomes":214,"average":214.0,"cosmic":214,"dbsnp":214},"novel":{"1000genomes":888,"average":888.0,"cosmic":888,"dbsnp":888}},"value":1102},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":944,"average":944.0,"cosmic":944,"dbsnp":944},"known":{"1000genomes":83,"average":83.0,"cosmic":83,"dbsnp":83},"novel":{"1000genomes":861,"average":861.0,"cosmic":861,"dbsnp":861}},"value":944},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":64,"average":64.0,"cosmic":64,"dbsnp":64},"known":{"1000genomes":53,"average":53.0,"cosmic":53,"dbsnp":53},"novel":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11}},"value":64},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":85,"average":85.0,"cosmic":85,"dbsnp":85},"known":{"1000genomes":69,"average":69.0,"cosmic":69,"dbsnp":69},"novel":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16}},"value":85},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":5,"average":94.66666666666667,"cosmic":65,"dbsnp":214},"known":{"1000genomes":5,"average":78.33333333333333,"cosmic":16,"dbsnp":214},"novel":{"1000genomes":0,"average":16.333333333333332,"cosmic":49,"dbsnp":0}},"value":65},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.45,"average":8.590000000000002,"cosmic":5.9,"dbsnp":19.42},"known":{"1000genomes":2.34,"average":36.60666666666667,"cosmic":7.48,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.8399999999999999,"cosmic":5.52,"dbsnp":0.0}},"value":5.9},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":1,"average":59.333333333333336,"cosmic":32,"dbsnp":145},"known":{"1000genomes":1,"average":52.666666666666664,"cosmic":12,"dbsnp":145},"novel":{"1000genomes":0,"average":6.666666666666667,"cosmic":20,"dbsnp":0}},"value":32},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":20.0,"average":45.663333333333334,"cosmic":49.23,"dbsnp":67.76},"known":{"1000genomes":20.0,"average":54.25333333333333,"cosmic":75.0,"dbsnp":67.76},"novel":{"1000genomes":0.0,"average":13.606666666666667,"cosmic":40.82,"dbsnp":0.0}},"value":49.23},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":3.5e-07,"average":3.5e-07,"cosmic":3.5e-07,"dbsnp":3.5e-07},"known":{"1000genomes":7e-08,"average":7e-08,"cosmic":7e-08,"dbsnp":7e-08},"novel":{"1000genomes":2.8e-07,"average":2.8e-07,"cosmic":2.8e-07,"dbsnp":2.8e-07}},"value":3.5e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":2846788.0,"average":2846788.0,"cosmic":2846788.0,"dbsnp":2846788.0},"known":{"1000genomes":14659632.0,"average":14659632.0,"cosmic":14659632.0,"dbsnp":14659632.0},"novel":{"1000genomes":3532839.0,"average":3532839.0,"cosmic":3532839.0,"dbsnp":3532839.0}},"value":2846788.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":16.22,"average":16.22,"cosmic":16.22,"dbsnp":16.22},"known":{"1000genomes":2.69,"average":2.69,"cosmic":2.69,"dbsnp":2.69},"novel":{"1000genomes":147.0,"average":147.0,"cosmic":147.0,"dbsnp":147.0}},"value":16.22},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.08,"average":0.08,"cosmic":0.08,"dbsnp":0.08},"known":{"1000genomes":0.34,"average":0.34,"cosmic":0.34,"dbsnp":0.34},"novel":{"1000genomes":0.06,"average":0.06,"cosmic":0.06,"dbsnp":0.06}},"value":0.08}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-5/qc/varQC/Tissue-Resistant-5-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-5/qc/varQC/Tissue-Resistant-5-mutect.varQC.json"},{"sample":{"name":"Tissue-Resistant-6","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-6","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1761,"average":1761.0,"cosmic":1761,"dbsnp":1761},"known":{"1000genomes":354,"average":354.0,"cosmic":354,"dbsnp":354},"novel":{"1000genomes":1407,"average":1407.0,"cosmic":1407,"dbsnp":1407}},"value":1761},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1498,"average":1498.0,"cosmic":1498,"dbsnp":1498},"known":{"1000genomes":127,"average":127.0,"cosmic":127,"dbsnp":127},"novel":{"1000genomes":1371,"average":1371.0,"cosmic":1371,"dbsnp":1371}},"value":1498},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":118,"average":118.0,"cosmic":118,"dbsnp":118},"known":{"1000genomes":97,"average":97.0,"cosmic":97,"dbsnp":97},"novel":{"1000genomes":21,"average":21.0,"cosmic":21,"dbsnp":21}},"value":118},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":129,"average":129.0,"cosmic":129,"dbsnp":129},"known":{"1000genomes":114,"average":114.0,"cosmic":114,"dbsnp":114},"novel":{"1000genomes":15,"average":15.0,"cosmic":15,"dbsnp":15}},"value":129},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":6,"average":143.33333333333334,"cosmic":70,"dbsnp":354},"known":{"1000genomes":6,"average":129.66666666666666,"cosmic":29,"dbsnp":354},"novel":{"1000genomes":0,"average":13.666666666666666,"cosmic":41,"dbsnp":0}},"value":70},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.34,"average":8.14,"cosmic":3.98,"dbsnp":20.1},"known":{"1000genomes":1.69,"average":36.626666666666665,"cosmic":8.19,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.9700000000000001,"cosmic":2.91,"dbsnp":0.0}},"value":3.98},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":5,"average":97.66666666666667,"cosmic":40,"dbsnp":248},"known":{"1000genomes":5,"average":91.33333333333333,"cosmic":21,"dbsnp":248},"novel":{"1000genomes":0,"average":6.333333333333333,"cosmic":19,"dbsnp":0}},"value":40},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":83.33,"average":70.17666666666666,"cosmic":57.14,"dbsnp":70.06},"known":{"1000genomes":83.33,"average":75.26666666666667,"cosmic":72.41,"dbsnp":70.06},"novel":{"1000genomes":0.0,"average":15.446666666666667,"cosmic":46.34,"dbsnp":0.0}},"value":57.14},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":5.6e-07,"average":5.6e-07,"cosmic":5.6e-07,"dbsnp":5.6e-07},"known":{"1000genomes":1.1e-07,"average":1.1e-07,"cosmic":1.1e-07,"dbsnp":1.1e-07},"novel":{"1000genomes":4.5e-07,"average":4.5e-07,"cosmic":4.5e-07,"dbsnp":4.5e-07}},"value":5.6e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":1781465.0,"average":1781465.0,"cosmic":1781465.0,"dbsnp":1781465.0},"known":{"1000genomes":8862037.0,"average":8862037.0,"cosmic":8862037.0,"dbsnp":8862037.0},"novel":{"1000genomes":2229681.0,"average":2229681.0,"cosmic":2229681.0,"dbsnp":2229681.0}},"value":1781465.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":10.98,"average":10.979999999999999,"cosmic":10.98,"dbsnp":10.98},"known":{"1000genomes":1.6,"average":1.6000000000000003,"cosmic":1.6,"dbsnp":1.6},"novel":{"1000genomes":126.91,"average":126.91000000000001,"cosmic":126.91,"dbsnp":126.91}},"value":10.98},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.05,"average":0.05000000000000001,"cosmic":0.05,"dbsnp":0.05},"known":{"1000genomes":0.28,"average":0.28,"cosmic":0.28,"dbsnp":0.28},"novel":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04}},"value":0.05}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-6/qc/varQC/Tissue-Resistant-6-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-6/qc/varQC/Tissue-Resistant-6-mutect.varQC.json"},{"sample":{"name":"Tissue-Resistant-7","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-7","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-7/Tissue-Resistant-7-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1753,"average":1753.0,"cosmic":1753,"dbsnp":1753},"known":{"1000genomes":372,"average":372.0,"cosmic":372,"dbsnp":372},"novel":{"1000genomes":1381,"average":1381.0,"cosmic":1381,"dbsnp":1381}},"value":1753},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1584,"average":1584.0,"cosmic":1584,"dbsnp":1584},"known":{"1000genomes":221,"average":221.0,"cosmic":221,"dbsnp":221},"novel":{"1000genomes":1363,"average":1363.0,"cosmic":1363,"dbsnp":1363}},"value":1584},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":69,"average":69.0,"cosmic":69,"dbsnp":69},"known":{"1000genomes":62,"average":62.0,"cosmic":62,"dbsnp":62},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":69},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":95,"average":95.0,"cosmic":95,"dbsnp":95},"known":{"1000genomes":84,"average":84.0,"cosmic":84,"dbsnp":84},"novel":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11}},"value":95},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":2,"average":159.33333333333334,"cosmic":104,"dbsnp":372},"known":{"1000genomes":2,"average":135.33333333333334,"cosmic":32,"dbsnp":372},"novel":{"1000genomes":0,"average":24.0,"cosmic":72,"dbsnp":0}},"value":104},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.11,"average":9.086666666666666,"cosmic":5.93,"dbsnp":21.22},"known":{"1000genomes":0.54,"average":36.38,"cosmic":8.6,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.7366666666666666,"cosmic":5.21,"dbsnp":0.0}},"value":5.93},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":2,"average":119.0,"cosmic":65,"dbsnp":290},"known":{"1000genomes":2,"average":105.0,"cosmic":23,"dbsnp":290},"novel":{"1000genomes":0,"average":14.0,"cosmic":42,"dbsnp":0}},"value":65},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":100.0,"average":80.15333333333332,"cosmic":62.5,"dbsnp":77.96},"known":{"1000genomes":100.0,"average":83.27999999999999,"cosmic":71.88,"dbsnp":77.96},"novel":{"1000genomes":0.0,"average":19.44333333333333,"cosmic":58.33,"dbsnp":0.0}},"value":62.5},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":5.6e-07,"average":5.6e-07,"cosmic":5.6e-07,"dbsnp":5.6e-07},"known":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07},"novel":{"1000genomes":4.4e-07,"average":4.4e-07,"cosmic":4.4e-07,"dbsnp":4.4e-07}},"value":5.6e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":1789595.0,"average":1789595.0,"cosmic":1789595.0,"dbsnp":1789595.0},"known":{"1000genomes":8433229.0,"average":8433229.0,"cosmic":8433229.0,"dbsnp":8433229.0},"novel":{"1000genomes":2271659.0,"average":2271659.0,"cosmic":2271659.0,"dbsnp":2271659.0}},"value":1789595.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":19.62,"average":19.62,"cosmic":19.62,"dbsnp":19.62},"known":{"1000genomes":3.59,"average":3.59,"cosmic":3.59,"dbsnp":3.59},"novel":{"1000genomes":344.25,"average":344.25,"cosmic":344.25,"dbsnp":344.25}},"value":19.62},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.51,"average":0.51,"cosmic":0.51,"dbsnp":0.51},"known":{"1000genomes":1.08,"average":1.08,"cosmic":1.08,"dbsnp":1.08},"novel":{"1000genomes":0.45,"average":0.45,"cosmic":0.45,"dbsnp":0.45}},"value":0.51}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-7/qc/varQC/Tissue-Resistant-7-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-7/qc/varQC/Tissue-Resistant-7-mutect.varQC.json"},{"sample":{"name":"Tissue-Resistant-8","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-8","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-8/Tissue-Resistant-8-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":37096,"average":37096.0,"cosmic":37096,"dbsnp":37096},"known":{"1000genomes":2319,"average":2319.0,"cosmic":2319,"dbsnp":2319},"novel":{"1000genomes":34777,"average":34777.0,"cosmic":34777,"dbsnp":34777}},"value":37096},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":36950,"average":36950.0,"cosmic":36950,"dbsnp":36950},"known":{"1000genomes":2196,"average":2196.0,"cosmic":2196,"dbsnp":2196},"novel":{"1000genomes":34754,"average":34754.0,"cosmic":34754,"dbsnp":34754}},"value":36950},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":62,"average":62.0,"cosmic":62,"dbsnp":62},"known":{"1000genomes":55,"average":55.0,"cosmic":55,"dbsnp":55},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":62},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":72,"average":72.0,"cosmic":72,"dbsnp":72},"known":{"1000genomes":56,"average":56.0,"cosmic":56,"dbsnp":56},"novel":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16}},"value":72},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":62,"average":1258.6666666666667,"cosmic":1395,"dbsnp":2319},"known":{"1000genomes":62,"average":858.0,"cosmic":193,"dbsnp":2319},"novel":{"1000genomes":0,"average":400.6666666666667,"cosmic":1202,"dbsnp":0}},"value":1395},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.17,"average":3.393333333333333,"cosmic":3.76,"dbsnp":6.25},"known":{"1000genomes":2.67,"average":36.99666666666667,"cosmic":8.32,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.1533333333333333,"cosmic":3.46,"dbsnp":0.0}},"value":3.76},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":4,"average":270.6666666666667,"cosmic":435,"dbsnp":373},"known":{"1000genomes":4,"average":138.33333333333334,"cosmic":38,"dbsnp":373},"novel":{"1000genomes":0,"average":132.33333333333334,"cosmic":397,"dbsnp":0}},"value":435},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":6.45,"average":17.903333333333332,"cosmic":31.18,"dbsnp":16.08},"known":{"1000genomes":6.45,"average":14.073333333333332,"cosmic":19.69,"dbsnp":16.08},"novel":{"1000genomes":0.0,"average":11.01,"cosmic":33.03,"dbsnp":0.0}},"value":31.18},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":1.182e-05,"average":1.182e-05,"cosmic":1.182e-05,"dbsnp":1.182e-05},"known":{"1000genomes":7.4e-07,"average":7.4e-07,"cosmic":7.4e-07,"dbsnp":7.4e-07},"novel":{"1000genomes":1.109e-05,"average":1.1089999999999999e-05,"cosmic":1.109e-05,"dbsnp":1.109e-05}},"value":1.182e-05},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":84568.0,"average":84568.0,"cosmic":84568.0,"dbsnp":84568.0},"known":{"1000genomes":1352807.0,"average":1352807.0,"cosmic":1352807.0,"dbsnp":1352807.0},"novel":{"1000genomes":90207.0,"average":90207.0,"cosmic":90207.0,"dbsnp":90207.0}},"value":84568.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":740.92,"average":740.92,"cosmic":740.92,"dbsnp":740.92},"known":{"1000genomes":48.34,"average":48.34,"cosmic":48.34,"dbsnp":48.34},"novel":{"1000genomes":11591.33,"average":11591.33,"cosmic":11591.33,"dbsnp":11591.33}},"value":740.92},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"known":{"1000genomes":0.02,"average":0.02,"cosmic":0.02,"dbsnp":0.02},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-8/qc/varQC/Tissue-Resistant-8-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-8/qc/varQC/Tissue-Resistant-8-mutect.varQC.json"},{"sample":{"name":"Tissue-Resistant-9","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-9","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-9/Tissue-Resistant-9-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":68804,"average":68804.0,"cosmic":68804,"dbsnp":68804},"known":{"1000genomes":4333,"average":4333.0,"cosmic":4333,"dbsnp":4333},"novel":{"1000genomes":64471,"average":64471.0,"cosmic":64471,"dbsnp":64471}},"value":68804},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":68619,"average":68619.0,"cosmic":68619,"dbsnp":68619},"known":{"1000genomes":4178,"average":4178.0,"cosmic":4178,"dbsnp":4178},"novel":{"1000genomes":64441,"average":64441.0,"cosmic":64441,"dbsnp":64441}},"value":68619},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":70,"average":70.0,"cosmic":70,"dbsnp":70},"known":{"1000genomes":61,"average":61.0,"cosmic":61,"dbsnp":61},"novel":{"1000genomes":9,"average":9.0,"cosmic":9,"dbsnp":9}},"value":70},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":105,"average":105.0,"cosmic":105,"dbsnp":105},"known":{"1000genomes":84,"average":84.0,"cosmic":84,"dbsnp":84},"novel":{"1000genomes":21,"average":21.0,"cosmic":21,"dbsnp":21}},"value":105},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":119,"average":2321.0,"cosmic":2511,"dbsnp":4333},"known":{"1000genomes":119,"average":1592.0,"cosmic":324,"dbsnp":4333},"novel":{"1000genomes":0,"average":729.0,"cosmic":2187,"dbsnp":0}},"value":2511},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.17,"average":3.373333333333333,"cosmic":3.65,"dbsnp":6.3},"known":{"1000genomes":2.75,"average":36.74333333333333,"cosmic":7.48,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.1300000000000001,"cosmic":3.39,"dbsnp":0.0}},"value":3.65},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":12,"average":513.0,"cosmic":894,"dbsnp":633},"known":{"1000genomes":12,"average":241.0,"cosmic":78,"dbsnp":633},"novel":{"1000genomes":0,"average":272.0,"cosmic":816,"dbsnp":0}},"value":894},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":10.08,"average":20.096666666666668,"cosmic":35.6,"dbsnp":14.61},"known":{"1000genomes":10.08,"average":16.253333333333334,"cosmic":24.07,"dbsnp":14.61},"novel":{"1000genomes":0.0,"average":12.436666666666667,"cosmic":37.31,"dbsnp":0.0}},"value":35.6},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":2.193e-05,"average":2.193e-05,"cosmic":2.193e-05,"dbsnp":2.193e-05},"known":{"1000genomes":1.38e-06,"average":1.3799999999999997e-06,"cosmic":1.38e-06,"dbsnp":1.38e-06},"novel":{"1000genomes":2.055e-05,"average":2.055e-05,"cosmic":2.055e-05,"dbsnp":2.055e-05}},"value":2.193e-05},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":45595.0,"average":45595.0,"cosmic":45595.0,"dbsnp":45595.0},"known":{"1000genomes":724015.0,"average":724015.0,"cosmic":724015.0,"dbsnp":724015.0},"novel":{"1000genomes":48660.0,"average":48660.0,"cosmic":48660.0,"dbsnp":48660.0}},"value":45595.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":838.07,"average":838.07,"cosmic":838.07,"dbsnp":838.07},"known":{"1000genomes":58.36,"average":58.35999999999999,"cosmic":58.36,"dbsnp":58.36},"novel":{"1000genomes":7162.44,"average":7162.44,"cosmic":7162.44,"dbsnp":7162.44}},"value":838.07},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"known":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-9/qc/varQC/Tissue-Resistant-9-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-9/qc/varQC/Tissue-Resistant-9-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-10","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-10","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-10/Tissue-Sensitive-10-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1755,"average":1755.0,"cosmic":1755,"dbsnp":1755},"known":{"1000genomes":249,"average":249.0,"cosmic":249,"dbsnp":249},"novel":{"1000genomes":1506,"average":1506.0,"cosmic":1506,"dbsnp":1506}},"value":1755},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1594,"average":1594.0,"cosmic":1594,"dbsnp":1594},"known":{"1000genomes":111,"average":111.0,"cosmic":111,"dbsnp":111},"novel":{"1000genomes":1483,"average":1483.0,"cosmic":1483,"dbsnp":1483}},"value":1594},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":53,"average":53.0,"cosmic":53,"dbsnp":53},"known":{"1000genomes":49,"average":49.0,"cosmic":49,"dbsnp":49},"novel":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4}},"value":53},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":91,"average":91.0,"cosmic":91,"dbsnp":91},"known":{"1000genomes":74,"average":74.0,"cosmic":74,"dbsnp":74},"novel":{"1000genomes":17,"average":17.0,"cosmic":17,"dbsnp":17}},"value":91},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":2,"average":109.33333333333333,"cosmic":77,"dbsnp":249},"known":{"1000genomes":2,"average":88.0,"cosmic":13,"dbsnp":249},"novel":{"1000genomes":0,"average":21.333333333333332,"cosmic":64,"dbsnp":0}},"value":77},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.11,"average":6.2299999999999995,"cosmic":4.39,"dbsnp":14.19},"known":{"1000genomes":0.8,"average":35.339999999999996,"cosmic":5.22,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4166666666666667,"cosmic":4.25,"dbsnp":0.0}},"value":4.39},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":0,"average":56.0,"cosmic":30,"dbsnp":138},"known":{"1000genomes":0,"average":49.333333333333336,"cosmic":10,"dbsnp":138},"novel":{"1000genomes":0,"average":6.666666666666667,"cosmic":20,"dbsnp":0}},"value":30},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":0.0,"average":31.459999999999997,"cosmic":38.96,"dbsnp":55.42},"known":{"1000genomes":0.0,"average":44.11333333333334,"cosmic":76.92,"dbsnp":55.42},"novel":{"1000genomes":0.0,"average":10.416666666666666,"cosmic":31.25,"dbsnp":0.0}},"value":38.96},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":5.6e-07,"average":5.6e-07,"cosmic":5.6e-07,"dbsnp":5.6e-07},"known":{"1000genomes":8e-08,"average":8.000000000000001e-08,"cosmic":8e-08,"dbsnp":8e-08},"novel":{"1000genomes":4.8e-07,"average":4.8e-07,"cosmic":4.8e-07,"dbsnp":4.8e-07}},"value":5.6e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":1787556.0,"average":1787556.0,"cosmic":1787556.0,"dbsnp":1787556.0},"known":{"1000genomes":12599041.0,"average":12599041.0,"cosmic":12599041.0,"dbsnp":12599041.0},"novel":{"1000genomes":2083108.0,"average":2083108.0,"cosmic":2083108.0,"dbsnp":2083108.0}},"value":1787556.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":38.0,"average":38.0,"cosmic":38.0,"dbsnp":38.0},"known":{"1000genomes":4.93,"average":4.93,"cosmic":4.93,"dbsnp":4.93},"novel":{"1000genomes":501.0,"average":501.0,"cosmic":501.0,"dbsnp":501.0}},"value":38.0},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.05,"average":0.05000000000000001,"cosmic":0.05,"dbsnp":0.05},"known":{"1000genomes":0.11,"average":0.11,"cosmic":0.11,"dbsnp":0.11},"novel":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04}},"value":0.05}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-10/qc/varQC/Tissue-Sensitive-10-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-10/qc/varQC/Tissue-Sensitive-10-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-1","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-1","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-1/Tissue-Sensitive-1-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1756,"average":1756.0,"cosmic":1756,"dbsnp":1756},"known":{"1000genomes":351,"average":351.0,"cosmic":351,"dbsnp":351},"novel":{"1000genomes":1405,"average":1405.0,"cosmic":1405,"dbsnp":1405}},"value":1756},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1606,"average":1606.0,"cosmic":1606,"dbsnp":1606},"known":{"1000genomes":219,"average":219.0,"cosmic":219,"dbsnp":219},"novel":{"1000genomes":1387,"average":1387.0,"cosmic":1387,"dbsnp":1387}},"value":1606},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":65,"average":65.0,"cosmic":65,"dbsnp":65},"known":{"1000genomes":55,"average":55.0,"cosmic":55,"dbsnp":55},"novel":{"1000genomes":10,"average":10.0,"cosmic":10,"dbsnp":10}},"value":65},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":75,"average":75.0,"cosmic":75,"dbsnp":75},"known":{"1000genomes":67,"average":67.0,"cosmic":67,"dbsnp":67},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":75},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":75,"average":164.0,"cosmic":66,"dbsnp":351},"known":{"1000genomes":75,"average":148.33333333333334,"cosmic":19,"dbsnp":351},"novel":{"1000genomes":0,"average":15.666666666666666,"cosmic":47,"dbsnp":0}},"value":66},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":4.27,"average":9.339999999999998,"cosmic":3.76,"dbsnp":19.99},"known":{"1000genomes":21.37,"average":42.26,"cosmic":5.41,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.1166666666666667,"cosmic":3.35,"dbsnp":0.0}},"value":3.76},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":71,"average":121.33333333333333,"cosmic":29,"dbsnp":264},"known":{"1000genomes":71,"average":116.33333333333333,"cosmic":14,"dbsnp":264},"novel":{"1000genomes":0,"average":5.0,"cosmic":15,"dbsnp":0}},"value":29},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":94.67,"average":71.27333333333333,"cosmic":43.94,"dbsnp":75.21},"known":{"1000genomes":94.67,"average":81.18666666666667,"cosmic":73.68,"dbsnp":75.21},"novel":{"1000genomes":0.0,"average":10.636666666666667,"cosmic":31.91,"dbsnp":0.0}},"value":43.94},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":5.6e-07,"average":5.6e-07,"cosmic":5.6e-07,"dbsnp":5.6e-07},"known":{"1000genomes":1.1e-07,"average":1.1e-07,"cosmic":1.1e-07,"dbsnp":1.1e-07},"novel":{"1000genomes":4.5e-07,"average":4.5e-07,"cosmic":4.5e-07,"dbsnp":4.5e-07}},"value":5.6e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":1786538.0,"average":1786538.0,"cosmic":1786538.0,"dbsnp":1786538.0},"known":{"1000genomes":8937781.0,"average":8937781.0,"cosmic":8937781.0,"dbsnp":8937781.0},"novel":{"1000genomes":2232854.0,"average":2232854.0,"cosmic":2232854.0,"dbsnp":2232854.0}},"value":1786538.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":32.77,"average":32.77,"cosmic":32.77,"dbsnp":32.77},"known":{"1000genomes":5.88,"average":5.88,"cosmic":5.88,"dbsnp":5.88},"novel":{"1000genomes":1404.0,"average":1404.0,"cosmic":1404.0,"dbsnp":1404.0}},"value":32.77},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.13,"average":0.13,"cosmic":0.13,"dbsnp":0.13},"known":{"1000genomes":0.96,"average":0.96,"cosmic":0.96,"dbsnp":0.96},"novel":{"1000genomes":0.06,"average":0.06,"cosmic":0.06,"dbsnp":0.06}},"value":0.13}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-1/qc/varQC/Tissue-Sensitive-1-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-1/qc/varQC/Tissue-Sensitive-1-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-2","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-2","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-2/Tissue-Sensitive-2-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1177,"average":1177.0,"cosmic":1177,"dbsnp":1177},"known":{"1000genomes":307,"average":307.0,"cosmic":307,"dbsnp":307},"novel":{"1000genomes":870,"average":870.0,"cosmic":870,"dbsnp":870}},"value":1177},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":912,"average":912.0,"cosmic":912,"dbsnp":912},"known":{"1000genomes":84,"average":84.0,"cosmic":84,"dbsnp":84},"novel":{"1000genomes":828,"average":828.0,"cosmic":828,"dbsnp":828}},"value":912},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":108,"average":108.0,"cosmic":108,"dbsnp":108},"known":{"1000genomes":95,"average":95.0,"cosmic":95,"dbsnp":95},"novel":{"1000genomes":13,"average":13.0,"cosmic":13,"dbsnp":13}},"value":108},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":141,"average":141.0,"cosmic":141,"dbsnp":141},"known":{"1000genomes":113,"average":113.0,"cosmic":113,"dbsnp":113},"novel":{"1000genomes":28,"average":28.0,"cosmic":28,"dbsnp":28}},"value":141},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":13,"average":127.0,"cosmic":61,"dbsnp":307},"known":{"1000genomes":13,"average":114.33333333333333,"cosmic":23,"dbsnp":307},"novel":{"1000genomes":0,"average":12.666666666666666,"cosmic":38,"dbsnp":0}},"value":61},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":1.1,"average":10.786666666666667,"cosmic":5.18,"dbsnp":26.08},"known":{"1000genomes":4.23,"average":37.24,"cosmic":7.49,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4566666666666668,"cosmic":4.37,"dbsnp":0.0}},"value":5.18},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":12,"average":97.66666666666667,"cosmic":37,"dbsnp":244},"known":{"1000genomes":12,"average":91.66666666666667,"cosmic":19,"dbsnp":244},"novel":{"1000genomes":0,"average":6.0,"cosmic":18,"dbsnp":0}},"value":37},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":92.31,"average":77.48333333333333,"cosmic":60.66,"dbsnp":79.48},"known":{"1000genomes":92.31,"average":84.80000000000001,"cosmic":82.61,"dbsnp":79.48},"novel":{"1000genomes":0.0,"average":15.79,"cosmic":47.37,"dbsnp":0.0}},"value":60.66},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":3.8e-07,"average":3.8e-07,"cosmic":3.8e-07,"dbsnp":3.8e-07},"known":{"1000genomes":1e-07,"average":1e-07,"cosmic":1e-07,"dbsnp":1e-07},"novel":{"1000genomes":2.8e-07,"average":2.8e-07,"cosmic":2.8e-07,"dbsnp":2.8e-07}},"value":3.8e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":2665387.0,"average":2665387.0,"cosmic":2665387.0,"dbsnp":2665387.0},"known":{"1000genomes":10218766.0,"average":10218766.0,"cosmic":10218766.0,"dbsnp":10218766.0},"novel":{"1000genomes":3605932.0,"average":3605932.0,"cosmic":3605932.0,"dbsnp":3605932.0}},"value":2665387.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":7.59,"average":7.59,"cosmic":7.59,"dbsnp":7.59},"known":{"1000genomes":1.44,"average":1.4400000000000002,"cosmic":1.44,"dbsnp":1.44},"novel":{"1000genomes":78.09,"average":78.09,"cosmic":78.09,"dbsnp":78.09}},"value":7.59},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.13,"average":0.13,"cosmic":0.13,"dbsnp":0.13},"known":{"1000genomes":0.65,"average":0.65,"cosmic":0.65,"dbsnp":0.65},"novel":{"1000genomes":0.1,"average":0.10000000000000002,"cosmic":0.1,"dbsnp":0.1}},"value":0.13}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-2/qc/varQC/Tissue-Sensitive-2-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-2/qc/varQC/Tissue-Sensitive-2-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-3","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-3","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-3/Tissue-Sensitive-3-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1232,"average":1232.0,"cosmic":1232,"dbsnp":1232},"known":{"1000genomes":205,"average":205.0,"cosmic":205,"dbsnp":205},"novel":{"1000genomes":1027,"average":1027.0,"cosmic":1027,"dbsnp":1027}},"value":1232},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1074,"average":1074.0,"cosmic":1074,"dbsnp":1074},"known":{"1000genomes":68,"average":68.0,"cosmic":68,"dbsnp":68},"novel":{"1000genomes":1006,"average":1006.0,"cosmic":1006,"dbsnp":1006}},"value":1074},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":60,"average":60.0,"cosmic":60,"dbsnp":60},"known":{"1000genomes":56,"average":56.0,"cosmic":56,"dbsnp":56},"novel":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4}},"value":60},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":89,"average":89.0,"cosmic":89,"dbsnp":89},"known":{"1000genomes":72,"average":72.0,"cosmic":72,"dbsnp":72},"novel":{"1000genomes":17,"average":17.0,"cosmic":17,"dbsnp":17}},"value":89},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":2,"average":89.0,"cosmic":60,"dbsnp":205},"known":{"1000genomes":2,"average":74.0,"cosmic":15,"dbsnp":205},"novel":{"1000genomes":0,"average":15.0,"cosmic":45,"dbsnp":0}},"value":60},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.16,"average":7.223333333333334,"cosmic":4.87,"dbsnp":16.64},"known":{"1000genomes":0.98,"average":36.1,"cosmic":7.32,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.46,"cosmic":4.38,"dbsnp":0.0}},"value":4.87},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":0,"average":58.333333333333336,"cosmic":27,"dbsnp":148},"known":{"1000genomes":0,"average":52.0,"cosmic":8,"dbsnp":148},"novel":{"1000genomes":0,"average":6.333333333333333,"cosmic":19,"dbsnp":0}},"value":27},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":0.0,"average":39.06666666666667,"cosmic":45.0,"dbsnp":72.2},"known":{"1000genomes":0.0,"average":41.843333333333334,"cosmic":53.33,"dbsnp":72.2},"novel":{"1000genomes":0.0,"average":14.073333333333332,"cosmic":42.22,"dbsnp":0.0}},"value":45.0},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":3.9e-07,"average":3.9e-07,"cosmic":3.9e-07,"dbsnp":3.9e-07},"known":{"1000genomes":7e-08,"average":7e-08,"cosmic":7e-08,"dbsnp":7e-08},"novel":{"1000genomes":3.3e-07,"average":3.3e-07,"cosmic":3.3e-07,"dbsnp":3.3e-07}},"value":3.9e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":2546397.0,"average":2546397.0,"cosmic":2546397.0,"dbsnp":2546397.0},"known":{"1000genomes":15303225.0,"average":15303225.0,"cosmic":15303225.0,"dbsnp":15303225.0},"novel":{"1000genomes":3054684.0,"average":3054684.0,"cosmic":3054684.0,"dbsnp":3054684.0}},"value":2546397.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":19.88,"average":19.88,"cosmic":19.88,"dbsnp":19.88},"known":{"1000genomes":2.73,"average":2.73,"cosmic":2.73,"dbsnp":2.73},"novel":{"1000genomes":255.75,"average":255.75,"cosmic":255.75,"dbsnp":255.75}},"value":19.88},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04},"known":{"1000genomes":0.24,"average":0.24,"cosmic":0.24,"dbsnp":0.24},"novel":{"1000genomes":0.03,"average":0.03,"cosmic":0.03,"dbsnp":0.03}},"value":0.04}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-3/qc/varQC/Tissue-Sensitive-3-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-3/qc/varQC/Tissue-Sensitive-3-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-5","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-5","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-5/Tissue-Sensitive-5-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":3259,"average":3259.0,"cosmic":3259,"dbsnp":3259},"known":{"1000genomes":383,"average":383.0,"cosmic":383,"dbsnp":383},"novel":{"1000genomes":2876,"average":2876.0,"cosmic":2876,"dbsnp":2876}},"value":3259},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":3064,"average":3064.0,"cosmic":3064,"dbsnp":3064},"known":{"1000genomes":216,"average":216.0,"cosmic":216,"dbsnp":216},"novel":{"1000genomes":2848,"average":2848.0,"cosmic":2848,"dbsnp":2848}},"value":3064},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":87,"average":87.0,"cosmic":87,"dbsnp":87},"known":{"1000genomes":76,"average":76.0,"cosmic":76,"dbsnp":76},"novel":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11}},"value":87},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":93,"average":93.0,"cosmic":93,"dbsnp":93},"known":{"1000genomes":76,"average":76.0,"cosmic":76,"dbsnp":76},"novel":{"1000genomes":17,"average":17.0,"cosmic":17,"dbsnp":17}},"value":93},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":13,"average":177.66666666666666,"cosmic":137,"dbsnp":383},"known":{"1000genomes":13,"average":141.33333333333334,"cosmic":28,"dbsnp":383},"novel":{"1000genomes":0,"average":36.333333333333336,"cosmic":109,"dbsnp":0}},"value":137},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.4,"average":5.45,"cosmic":4.2,"dbsnp":11.75},"known":{"1000genomes":3.39,"average":36.9,"cosmic":7.31,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.2633333333333334,"cosmic":3.79,"dbsnp":0.0}},"value":4.2},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":8,"average":90.66666666666667,"cosmic":60,"dbsnp":204},"known":{"1000genomes":8,"average":76.0,"cosmic":16,"dbsnp":204},"novel":{"1000genomes":0,"average":14.666666666666666,"cosmic":44,"dbsnp":0}},"value":60},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":61.54,"average":52.86666666666667,"cosmic":43.8,"dbsnp":53.26},"known":{"1000genomes":61.54,"average":57.31333333333333,"cosmic":57.14,"dbsnp":53.26},"novel":{"1000genomes":0.0,"average":13.456666666666665,"cosmic":40.37,"dbsnp":0.0}},"value":43.8},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":1.04e-06,"average":1.04e-06,"cosmic":1.04e-06,"dbsnp":1.04e-06},"known":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07},"novel":{"1000genomes":9.2e-07,"average":9.2e-07,"cosmic":9.2e-07,"dbsnp":9.2e-07}},"value":1.04e-06},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":962614.0,"average":962614.0,"cosmic":962614.0,"dbsnp":962614.0},"known":{"1000genomes":8191021.0,"average":8191021.0,"cosmic":8191021.0,"dbsnp":8191021.0},"novel":{"1000genomes":1090807.0,"average":1090807.0,"cosmic":1090807.0,"dbsnp":1090807.0}},"value":962614.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":40.78,"average":40.78,"cosmic":40.78,"dbsnp":40.78},"known":{"1000genomes":3.97,"average":3.97,"cosmic":3.97,"dbsnp":3.97},"novel":{"1000genomes":2875.0,"average":2875.0,"cosmic":2875.0,"dbsnp":2875.0}},"value":40.78},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04},"known":{"1000genomes":0.14,"average":0.14,"cosmic":0.14,"dbsnp":0.14},"novel":{"1000genomes":0.03,"average":0.03,"cosmic":0.03,"dbsnp":0.03}},"value":0.04}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-5/qc/varQC/Tissue-Sensitive-5-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-5/qc/varQC/Tissue-Sensitive-5-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-6","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-6","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-6/Tissue-Sensitive-6-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1388,"average":1388.0,"cosmic":1388,"dbsnp":1388},"known":{"1000genomes":245,"average":245.0,"cosmic":245,"dbsnp":245},"novel":{"1000genomes":1143,"average":1143.0,"cosmic":1143,"dbsnp":1143}},"value":1388},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1193,"average":1193.0,"cosmic":1193,"dbsnp":1193},"known":{"1000genomes":85,"average":85.0,"cosmic":85,"dbsnp":85},"novel":{"1000genomes":1108,"average":1108.0,"cosmic":1108,"dbsnp":1108}},"value":1193},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":76,"average":76.0,"cosmic":76,"dbsnp":76},"known":{"1000genomes":66,"average":66.0,"cosmic":66,"dbsnp":66},"novel":{"1000genomes":10,"average":10.0,"cosmic":10,"dbsnp":10}},"value":76},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":110,"average":110.0,"cosmic":110,"dbsnp":110},"known":{"1000genomes":86,"average":86.0,"cosmic":86,"dbsnp":86},"novel":{"1000genomes":24,"average":24.0,"cosmic":24,"dbsnp":24}},"value":110},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":2,"average":104.66666666666667,"cosmic":67,"dbsnp":245},"known":{"1000genomes":2,"average":90.0,"cosmic":23,"dbsnp":245},"novel":{"1000genomes":0,"average":14.666666666666666,"cosmic":44,"dbsnp":0}},"value":67},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.14,"average":7.539999999999999,"cosmic":4.83,"dbsnp":17.65},"known":{"1000genomes":0.82,"average":36.736666666666665,"cosmic":9.39,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.2833333333333334,"cosmic":3.85,"dbsnp":0.0}},"value":4.83},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":1,"average":70.66666666666667,"cosmic":37,"dbsnp":174},"known":{"1000genomes":1,"average":64.66666666666667,"cosmic":19,"dbsnp":174},"novel":{"1000genomes":0,"average":6.0,"cosmic":18,"dbsnp":0}},"value":37},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":50.0,"average":58.74666666666667,"cosmic":55.22,"dbsnp":71.02},"known":{"1000genomes":50.0,"average":67.87666666666667,"cosmic":82.61,"dbsnp":71.02},"novel":{"1000genomes":0.0,"average":13.636666666666665,"cosmic":40.91,"dbsnp":0.0}},"value":55.22},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":4.4e-07,"average":4.4e-07,"cosmic":4.4e-07,"dbsnp":4.4e-07},"known":{"1000genomes":8e-08,"average":8.000000000000001e-08,"cosmic":8e-08,"dbsnp":8e-08},"novel":{"1000genomes":3.6e-07,"average":3.6e-07,"cosmic":3.6e-07,"dbsnp":3.6e-07}},"value":4.4e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":2260202.0,"average":2260202.0,"cosmic":2260202.0,"dbsnp":2260202.0},"known":{"1000genomes":12804739.0,"average":12804739.0,"cosmic":12804739.0,"dbsnp":12804739.0},"novel":{"1000genomes":2744673.0,"average":2744673.0,"cosmic":2744673.0,"dbsnp":2744673.0}},"value":2260202.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":14.77,"average":14.770000000000001,"cosmic":14.77,"dbsnp":14.77},"known":{"1000genomes":2.02,"average":2.02,"cosmic":2.02,"dbsnp":2.02},"novel":{"1000genomes":162.29,"average":162.29,"cosmic":162.29,"dbsnp":162.29}},"value":14.77},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.14,"average":0.14,"cosmic":0.14,"dbsnp":0.14},"known":{"1000genomes":0.49,"average":0.49,"cosmic":0.49,"dbsnp":0.49},"novel":{"1000genomes":0.12,"average":0.12,"cosmic":0.12,"dbsnp":0.12}},"value":0.14}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-6/qc/varQC/Tissue-Sensitive-6-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-6/qc/varQC/Tissue-Sensitive-6-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-7","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-7","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":2045,"average":2045.0,"cosmic":2045,"dbsnp":2045},"known":{"1000genomes":259,"average":259.0,"cosmic":259,"dbsnp":259},"novel":{"1000genomes":1786,"average":1786.0,"cosmic":1786,"dbsnp":1786}},"value":2045},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1891,"average":1891.0,"cosmic":1891,"dbsnp":1891},"known":{"1000genomes":136,"average":136.0,"cosmic":136,"dbsnp":136},"novel":{"1000genomes":1755,"average":1755.0,"cosmic":1755,"dbsnp":1755}},"value":1891},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":52,"average":52.0,"cosmic":52,"dbsnp":52},"known":{"1000genomes":42,"average":42.0,"cosmic":42,"dbsnp":42},"novel":{"1000genomes":10,"average":10.0,"cosmic":10,"dbsnp":10}},"value":52},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":88,"average":88.0,"cosmic":88,"dbsnp":88},"known":{"1000genomes":68,"average":68.0,"cosmic":68,"dbsnp":68},"novel":{"1000genomes":20,"average":20.0,"cosmic":20,"dbsnp":20}},"value":88},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":10,"average":119.0,"cosmic":88,"dbsnp":259},"known":{"1000genomes":10,"average":96.66666666666667,"cosmic":21,"dbsnp":259},"novel":{"1000genomes":0,"average":22.333333333333332,"cosmic":67,"dbsnp":0}},"value":88},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.49,"average":5.82,"cosmic":4.3,"dbsnp":12.67},"known":{"1000genomes":3.86,"average":37.32333333333333,"cosmic":8.11,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.25,"cosmic":3.75,"dbsnp":0.0}},"value":4.3},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":6,"average":63.0,"cosmic":38,"dbsnp":145},"known":{"1000genomes":6,"average":54.333333333333336,"cosmic":12,"dbsnp":145},"novel":{"1000genomes":0,"average":8.666666666666666,"cosmic":26,"dbsnp":0}},"value":38},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":60.0,"average":53.053333333333335,"cosmic":43.18,"dbsnp":55.98},"known":{"1000genomes":60.0,"average":57.70666666666667,"cosmic":57.14,"dbsnp":55.98},"novel":{"1000genomes":0.0,"average":12.936666666666667,"cosmic":38.81,"dbsnp":0.0}},"value":43.18},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":6.5e-07,"average":6.5e-07,"cosmic":6.5e-07,"dbsnp":6.5e-07},"known":{"1000genomes":8e-08,"average":8.000000000000001e-08,"cosmic":8e-08,"dbsnp":8e-08},"novel":{"1000genomes":5.7e-07,"average":5.7e-07,"cosmic":5.7e-07,"dbsnp":5.7e-07}},"value":6.5e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":1534064.0,"average":1534064.0,"cosmic":1534064.0,"dbsnp":1534064.0},"known":{"1000genomes":12112591.0,"average":12112591.0,"cosmic":12112591.0,"dbsnp":12112591.0},"novel":{"1000genomes":1756529.0,"average":1756529.0,"cosmic":1756529.0,"dbsnp":1756529.0}},"value":1534064.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":51.44,"average":51.44,"cosmic":51.44,"dbsnp":51.44},"known":{"1000genomes":6.85,"average":6.849999999999999,"cosmic":6.85,"dbsnp":6.85},"novel":{"1000genomes":296.67,"average":296.67,"cosmic":296.67,"dbsnp":296.67}},"value":51.44},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04},"known":{"1000genomes":0.23,"average":0.23,"cosmic":0.23,"dbsnp":0.23},"novel":{"1000genomes":0.03,"average":0.03,"cosmic":0.03,"dbsnp":0.03}},"value":0.04}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-7/qc/varQC/Tissue-Sensitive-7-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-7/qc/varQC/Tissue-Sensitive-7-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-8","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-8","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":2750,"average":2750.0,"cosmic":2750,"dbsnp":2750},"known":{"1000genomes":394,"average":394.0,"cosmic":394,"dbsnp":394},"novel":{"1000genomes":2356,"average":2356.0,"cosmic":2356,"dbsnp":2356}},"value":2750},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":2522,"average":2522.0,"cosmic":2522,"dbsnp":2522},"known":{"1000genomes":210,"average":210.0,"cosmic":210,"dbsnp":210},"novel":{"1000genomes":2312,"average":2312.0,"cosmic":2312,"dbsnp":2312}},"value":2522},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":94,"average":94.0,"cosmic":94,"dbsnp":94},"known":{"1000genomes":78,"average":78.0,"cosmic":78,"dbsnp":78},"novel":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16}},"value":94},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":122,"average":122.0,"cosmic":122,"dbsnp":122},"known":{"1000genomes":95,"average":95.0,"cosmic":95,"dbsnp":95},"novel":{"1000genomes":27,"average":27.0,"cosmic":27,"dbsnp":27}},"value":122},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":9,"average":175.66666666666666,"cosmic":124,"dbsnp":394},"known":{"1000genomes":9,"average":145.0,"cosmic":32,"dbsnp":394},"novel":{"1000genomes":0,"average":30.666666666666668,"cosmic":92,"dbsnp":0}},"value":124},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.33,"average":6.390000000000001,"cosmic":4.51,"dbsnp":14.33},"known":{"1000genomes":2.28,"average":36.800000000000004,"cosmic":8.12,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.3,"cosmic":3.9,"dbsnp":0.0}},"value":4.51},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":2,"average":94.33333333333333,"cosmic":66,"dbsnp":215},"known":{"1000genomes":2,"average":79.33333333333333,"cosmic":21,"dbsnp":215},"novel":{"1000genomes":0,"average":15.0,"cosmic":45,"dbsnp":0}},"value":66},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":22.22,"average":43.339999999999996,"cosmic":53.23,"dbsnp":54.57},"known":{"1000genomes":22.22,"average":47.47333333333333,"cosmic":65.63,"dbsnp":54.57},"novel":{"1000genomes":0.0,"average":16.30333333333333,"cosmic":48.91,"dbsnp":0.0}},"value":53.23},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":8.8e-07,"average":8.8e-07,"cosmic":8.8e-07,"dbsnp":8.8e-07},"known":{"1000genomes":1.3e-07,"average":1.3e-07,"cosmic":1.3e-07,"dbsnp":1.3e-07},"novel":{"1000genomes":7.5e-07,"average":7.5e-07,"cosmic":7.5e-07,"dbsnp":7.5e-07}},"value":8.8e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":1140785.0,"average":1140785.0,"cosmic":1140785.0,"dbsnp":1140785.0},"known":{"1000genomes":7962338.0,"average":7962338.0,"cosmic":7962338.0,"dbsnp":7962338.0},"novel":{"1000genomes":1331562.0,"average":1331562.0,"cosmic":1331562.0,"dbsnp":1331562.0}},"value":1140785.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":27.95,"average":27.95,"cosmic":27.95,"dbsnp":27.95},"known":{"1000genomes":3.58,"average":3.58,"cosmic":3.58,"dbsnp":3.58},"novel":{"1000genomes":260.78,"average":260.78,"cosmic":260.78,"dbsnp":260.78}},"value":27.95},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.05,"average":0.05000000000000001,"cosmic":0.05,"dbsnp":0.05},"known":{"1000genomes":0.19,"average":0.19000000000000003,"cosmic":0.19,"dbsnp":0.19},"novel":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04}},"value":0.05}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-8/qc/varQC/Tissue-Sensitive-8-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-8/qc/varQC/Tissue-Sensitive-8-mutect.varQC.json"},{"sample":{"name":"Tissue-Sensitive-9","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-9","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-9/Tissue-Sensitive-9-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1976,"average":1976.0,"cosmic":1976,"dbsnp":1976},"known":{"1000genomes":289,"average":289.0,"cosmic":289,"dbsnp":289},"novel":{"1000genomes":1687,"average":1687.0,"cosmic":1687,"dbsnp":1687}},"value":1976},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1787,"average":1787.0,"cosmic":1787,"dbsnp":1787},"known":{"1000genomes":128,"average":128.0,"cosmic":128,"dbsnp":128},"novel":{"1000genomes":1659,"average":1659.0,"cosmic":1659,"dbsnp":1659}},"value":1787},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":72,"average":72.0,"cosmic":72,"dbsnp":72},"known":{"1000genomes":61,"average":61.0,"cosmic":61,"dbsnp":61},"novel":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11}},"value":72},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":104,"average":104.0,"cosmic":104,"dbsnp":104},"known":{"1000genomes":88,"average":88.0,"cosmic":88,"dbsnp":88},"novel":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16}},"value":104},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":6,"average":131.33333333333334,"cosmic":99,"dbsnp":289},"known":{"1000genomes":6,"average":106.33333333333333,"cosmic":24,"dbsnp":289},"novel":{"1000genomes":0,"average":25.0,"cosmic":75,"dbsnp":0}},"value":99},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.3,"average":6.646666666666667,"cosmic":5.01,"dbsnp":14.63},"known":{"1000genomes":2.08,"average":36.79333333333333,"cosmic":8.3,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4833333333333334,"cosmic":4.45,"dbsnp":0.0}},"value":5.01},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":2,"average":76.66666666666667,"cosmic":50,"dbsnp":178},"known":{"1000genomes":2,"average":65.0,"cosmic":15,"dbsnp":178},"novel":{"1000genomes":0,"average":11.666666666666666,"cosmic":35,"dbsnp":0}},"value":50},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":33.33,"average":48.47666666666667,"cosmic":50.51,"dbsnp":61.59},"known":{"1000genomes":33.33,"average":52.473333333333336,"cosmic":62.5,"dbsnp":61.59},"novel":{"1000genomes":0.0,"average":15.556666666666667,"cosmic":46.67,"dbsnp":0.0}},"value":50.51},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":6.3e-07,"average":6.3e-07,"cosmic":6.3e-07,"dbsnp":6.3e-07},"known":{"1000genomes":9e-08,"average":9e-08,"cosmic":9e-08,"dbsnp":9e-08},"novel":{"1000genomes":5.4e-07,"average":5.4e-07,"cosmic":5.4e-07,"dbsnp":5.4e-07}},"value":6.3e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":1587632.0,"average":1587632.0,"cosmic":1587632.0,"dbsnp":1587632.0},"known":{"1000genomes":10855229.0,"average":10855229.0,"cosmic":10855229.0,"dbsnp":10855229.0},"novel":{"1000genomes":1859609.0,"average":1859609.0,"cosmic":1859609.0,"dbsnp":1859609.0}},"value":1587632.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":24.01,"average":24.01,"cosmic":24.01,"dbsnp":24.01},"known":{"1000genomes":2.8,"average":2.7999999999999994,"cosmic":2.8,"dbsnp":2.8},"novel":{"1000genomes":561.33,"average":561.33,"cosmic":561.33,"dbsnp":561.33}},"value":24.01},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.05,"average":0.05000000000000001,"cosmic":0.05,"dbsnp":0.05},"known":{"1000genomes":0.2,"average":0.20000000000000004,"cosmic":0.2,"dbsnp":0.2},"novel":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04}},"value":0.05}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-9/qc/varQC/Tissue-Sensitive-9-mutect.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-9/qc/varQC/Tissue-Sensitive-9-mutect.varQC.json"}],"date":"13 August 2014, Wednesday, 18:56:46"}
            </div>
        </div>
    </div>
</div>
</body>
</html>
