<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>
    <script type="text/javascript" src="html_aux/dragtable.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for vardict</h1>
        <p id='report_date'></p>
        <p id="extended_link"></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"report":[{"sample":{"name":"Tissue-Resistant-10","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-10","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-10/Tissue-Resistant-10-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1283,"average":1283.0,"cosmic":1283,"dbsnp":1283},"known":{"1000genomes":111,"average":111.0,"cosmic":111,"dbsnp":111},"novel":{"1000genomes":1172,"average":1172.0,"cosmic":1172,"dbsnp":1172}},"value":1283},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1256,"average":1256.0,"cosmic":1256,"dbsnp":1256},"known":{"1000genomes":99,"average":99.0,"cosmic":99,"dbsnp":99},"novel":{"1000genomes":1157,"average":1157.0,"cosmic":1157,"dbsnp":1157}},"value":1256},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8},"known":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3},"novel":{"1000genomes":5,"average":5.0,"cosmic":5,"dbsnp":5}},"value":8},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16},"known":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":16},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":6,"average":61.0,"cosmic":66,"dbsnp":111},"known":{"1000genomes":6,"average":42.333333333333336,"cosmic":10,"dbsnp":111},"novel":{"1000genomes":0,"average":18.666666666666668,"cosmic":56,"dbsnp":0}},"value":66},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.47,"average":4.753333333333334,"cosmic":5.14,"dbsnp":8.65},"known":{"1000genomes":5.41,"average":38.14,"cosmic":9.01,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.5933333333333335,"cosmic":4.78,"dbsnp":0.0}},"value":5.14},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":4,"average":32.333333333333336,"cosmic":32,"dbsnp":61},"known":{"1000genomes":4,"average":24.333333333333332,"cosmic":8,"dbsnp":61},"novel":{"1000genomes":0,"average":8.0,"cosmic":24,"dbsnp":0}},"value":32},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":66.67,"average":56.699999999999996,"cosmic":48.48,"dbsnp":54.95},"known":{"1000genomes":66.67,"average":67.20666666666666,"cosmic":80.0,"dbsnp":54.95},"novel":{"1000genomes":0.0,"average":14.286666666666667,"cosmic":42.86,"dbsnp":0.0}},"value":48.48},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":4.1e-07,"average":4.1000000000000004e-07,"cosmic":4.1e-07,"dbsnp":4.1e-07},"known":{"1000genomes":4e-08,"average":4.000000000000001e-08,"cosmic":4e-08,"dbsnp":4e-08},"novel":{"1000genomes":3.7e-07,"average":3.7e-07,"cosmic":3.7e-07,"dbsnp":3.7e-07}},"value":4.1e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":2445176.0,"average":2445176.0,"cosmic":2445176.0,"dbsnp":2445176.0},"known":{"1000genomes":28262714.0,"average":28262714.0,"cosmic":28262714.0,"dbsnp":28262714.0},"novel":{"1000genomes":2676758.0,"average":2676758.0,"cosmic":2676758.0,"dbsnp":2676758.0}},"value":2445176.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":324.75,"average":324.75,"cosmic":324.75,"dbsnp":324.75},"known":{"1000genomes":39.67,"average":39.67,"cosmic":39.67,"dbsnp":39.67},"novel":{"1000genomes":1180.0,"average":1180.0,"cosmic":1180.0,"dbsnp":1180.0}},"value":324.75},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.09,"average":0.09000000000000001,"cosmic":0.09,"dbsnp":0.09},"known":{"1000genomes":0.46,"average":0.46,"cosmic":0.46,"dbsnp":0.46},"novel":{"1000genomes":0.07,"average":0.07,"cosmic":0.07,"dbsnp":0.07}},"value":0.09}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-10/qc/varQC/Tissue-Resistant-10-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-10/qc/varQC/Tissue-Resistant-10-vardict.varQC.json"},{"sample":{"name":"Tissue-Resistant-1","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-1","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":12012,"average":12012.0,"cosmic":12012,"dbsnp":12012},"known":{"1000genomes":771,"average":771.0,"cosmic":771,"dbsnp":771},"novel":{"1000genomes":11241,"average":11241.0,"cosmic":11241,"dbsnp":11241}},"value":12012},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":11986,"average":11986.0,"cosmic":11986,"dbsnp":11986},"known":{"1000genomes":764,"average":764.0,"cosmic":764,"dbsnp":764},"novel":{"1000genomes":11222,"average":11222.0,"cosmic":11222,"dbsnp":11222}},"value":11986},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4},"known":{"1000genomes":1,"average":1.0,"cosmic":1,"dbsnp":1},"novel":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3}},"value":4},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":20,"average":20.0,"cosmic":20,"dbsnp":20},"known":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"novel":{"1000genomes":14,"average":14.0,"cosmic":14,"dbsnp":14}},"value":20},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":27,"average":471.6666666666667,"cosmic":617,"dbsnp":771},"known":{"1000genomes":27,"average":296.0,"cosmic":90,"dbsnp":771},"novel":{"1000genomes":0,"average":175.66666666666666,"cosmic":527,"dbsnp":0}},"value":617},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.22,"average":3.9266666666666663,"cosmic":5.14,"dbsnp":6.42},"known":{"1000genomes":3.5,"average":38.39,"cosmic":11.67,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.5633333333333335,"cosmic":4.69,"dbsnp":0.0}},"value":5.14},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":6,"average":111.0,"cosmic":201,"dbsnp":126},"known":{"1000genomes":6,"average":52.333333333333336,"cosmic":25,"dbsnp":126},"novel":{"1000genomes":0,"average":58.666666666666664,"cosmic":176,"dbsnp":0}},"value":201},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":22.22,"average":23.713333333333335,"cosmic":32.58,"dbsnp":16.34},"known":{"1000genomes":22.22,"average":22.113333333333333,"cosmic":27.78,"dbsnp":16.34},"novel":{"1000genomes":0.0,"average":11.133333333333333,"cosmic":33.4,"dbsnp":0.0}},"value":32.58},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":3.83e-06,"average":3.83e-06,"cosmic":3.83e-06,"dbsnp":3.83e-06},"known":{"1000genomes":2.5e-07,"average":2.5e-07,"cosmic":2.5e-07,"dbsnp":2.5e-07},"novel":{"1000genomes":3.58e-06,"average":3.58e-06,"cosmic":3.58e-06,"dbsnp":3.58e-06}},"value":3.83e-06},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":261168.0,"average":261168.0,"cosmic":261168.0,"dbsnp":261168.0},"known":{"1000genomes":4068951.0,"average":4068951.0,"cosmic":4068951.0,"dbsnp":4068951.0},"novel":{"1000genomes":279082.0,"average":279082.0,"cosmic":279082.0,"dbsnp":279082.0}},"value":261168.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":858.71,"average":858.71,"cosmic":858.71,"dbsnp":858.71},"known":{"1000genomes":155.0,"average":155.0,"cosmic":155.0,"dbsnp":155.0},"novel":{"1000genomes":1249.67,"average":1249.67,"cosmic":1249.67,"dbsnp":1249.67}},"value":858.71},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01},"known":{"1000genomes":0.03,"average":0.03,"cosmic":0.03,"dbsnp":0.03},"novel":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01}},"value":0.01}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-1/qc/varQC/Tissue-Resistant-1-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-1/qc/varQC/Tissue-Resistant-1-vardict.varQC.json"},{"sample":{"name":"Tissue-Resistant-2","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-2","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-2/Tissue-Resistant-2-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":15715,"average":15715.0,"cosmic":15715,"dbsnp":15715},"known":{"1000genomes":904,"average":904.0,"cosmic":904,"dbsnp":904},"novel":{"1000genomes":14811,"average":14811.0,"cosmic":14811,"dbsnp":14811}},"value":15715},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":15686,"average":15686.0,"cosmic":15686,"dbsnp":15686},"known":{"1000genomes":895,"average":895.0,"cosmic":895,"dbsnp":895},"novel":{"1000genomes":14791,"average":14791.0,"cosmic":14791,"dbsnp":14791}},"value":15686},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":9,"average":9.0,"cosmic":9,"dbsnp":9},"known":{"1000genomes":2,"average":2.0,"cosmic":2,"dbsnp":2},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":9},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":19,"average":19.0,"cosmic":19,"dbsnp":19},"known":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"novel":{"1000genomes":13,"average":13.0,"cosmic":13,"dbsnp":13}},"value":19},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":24,"average":561.0,"cosmic":755,"dbsnp":904},"known":{"1000genomes":24,"average":337.0,"cosmic":83,"dbsnp":904},"novel":{"1000genomes":0,"average":224.0,"cosmic":672,"dbsnp":0}},"value":755},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.15,"average":3.5666666666666664,"cosmic":4.8,"dbsnp":5.75},"known":{"1000genomes":2.65,"average":37.27666666666667,"cosmic":9.18,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.5133333333333334,"cosmic":4.54,"dbsnp":0.0}},"value":4.8},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":1,"average":128.33333333333334,"cosmic":266,"dbsnp":118},"known":{"1000genomes":1,"average":44.666666666666664,"cosmic":15,"dbsnp":118},"novel":{"1000genomes":0,"average":83.66666666666667,"cosmic":251,"dbsnp":0}},"value":266},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":4.17,"average":17.48333333333333,"cosmic":35.23,"dbsnp":13.05},"known":{"1000genomes":4.17,"average":11.763333333333334,"cosmic":18.07,"dbsnp":13.05},"novel":{"1000genomes":0.0,"average":12.450000000000001,"cosmic":37.35,"dbsnp":0.0}},"value":35.23},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":5.01e-06,"average":5.01e-06,"cosmic":5.01e-06,"dbsnp":5.01e-06},"known":{"1000genomes":2.9e-07,"average":2.9e-07,"cosmic":2.9e-07,"dbsnp":2.9e-07},"novel":{"1000genomes":4.72e-06,"average":4.72e-06,"cosmic":4.72e-06,"dbsnp":4.72e-06}},"value":5.01e-06},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":199628.0,"average":199628.0,"cosmic":199628.0,"dbsnp":199628.0},"known":{"1000genomes":3470311.0,"average":3470311.0,"cosmic":3470311.0,"dbsnp":3470311.0},"novel":{"1000genomes":211812.0,"average":211812.0,"cosmic":211812.0,"dbsnp":211812.0}},"value":199628.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":5243.0,"average":5243.0,"cosmic":5243.0,"dbsnp":5243.0},"known":{"1000genomes":301.67,"average":301.67,"cosmic":301.67,"dbsnp":301.67},"novel":{"1000genomes":14824.0,"average":14824.0,"cosmic":14824.0,"dbsnp":14824.0}},"value":5243.0},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0},"known":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01},"novel":{"1000genomes":0.0,"average":0.0,"cosmic":0.0,"dbsnp":0.0}},"value":0.0}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-2/qc/varQC/Tissue-Resistant-2-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-2/qc/varQC/Tissue-Resistant-2-vardict.varQC.json"},{"sample":{"name":"Tissue-Resistant-4","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-4","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":15606,"average":15606.0,"cosmic":15606,"dbsnp":15606},"known":{"1000genomes":983,"average":983.0,"cosmic":983,"dbsnp":983},"novel":{"1000genomes":14623,"average":14623.0,"cosmic":14623,"dbsnp":14623}},"value":15606},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":15576,"average":15576.0,"cosmic":15576,"dbsnp":15576},"known":{"1000genomes":977,"average":977.0,"cosmic":977,"dbsnp":977},"novel":{"1000genomes":14599,"average":14599.0,"cosmic":14599,"dbsnp":14599}},"value":15576},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":9,"average":9.0,"cosmic":9,"dbsnp":9},"known":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3},"novel":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6}},"value":9},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":18,"average":18.0,"cosmic":18,"dbsnp":18},"known":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3},"novel":{"1000genomes":15,"average":15.0,"cosmic":15,"dbsnp":15}},"value":18},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":30,"average":577.6666666666666,"cosmic":720,"dbsnp":983},"known":{"1000genomes":30,"average":367.6666666666667,"cosmic":90,"dbsnp":983},"novel":{"1000genomes":0,"average":210.0,"cosmic":630,"dbsnp":0}},"value":720},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.19,"average":3.7000000000000006,"cosmic":4.61,"dbsnp":6.3},"known":{"1000genomes":3.05,"average":37.40333333333333,"cosmic":9.16,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4366666666666665,"cosmic":4.31,"dbsnp":0.0}},"value":4.61},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":12,"average":118.66666666666667,"cosmic":208,"dbsnp":136},"known":{"1000genomes":12,"average":54.333333333333336,"cosmic":15,"dbsnp":136},"novel":{"1000genomes":0,"average":64.33333333333333,"cosmic":193,"dbsnp":0}},"value":208},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":40.0,"average":27.576666666666668,"cosmic":28.89,"dbsnp":13.84},"known":{"1000genomes":40.0,"average":23.503333333333334,"cosmic":16.67,"dbsnp":13.84},"novel":{"1000genomes":0.0,"average":10.209999999999999,"cosmic":30.63,"dbsnp":0.0}},"value":28.89},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":4.97e-06,"average":4.97e-06,"cosmic":4.97e-06,"dbsnp":4.97e-06},"known":{"1000genomes":3.1e-07,"average":3.1e-07,"cosmic":3.1e-07,"dbsnp":3.1e-07},"novel":{"1000genomes":4.66e-06,"average":4.66e-06,"cosmic":4.66e-06,"dbsnp":4.66e-06}},"value":4.97e-06},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":201022.0,"average":201022.0,"cosmic":201022.0,"dbsnp":201022.0},"known":{"1000genomes":3191415.0,"average":3191415.0,"cosmic":3191415.0,"dbsnp":3191415.0},"novel":{"1000genomes":214536.0,"average":214536.0,"cosmic":214536.0,"dbsnp":214536.0}},"value":201022.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":1200.54,"average":1200.54,"cosmic":1200.54,"dbsnp":1200.54},"known":{"1000genomes":163.5,"average":163.5,"cosmic":163.5,"dbsnp":163.5},"novel":{"1000genomes":2089.43,"average":2089.43,"cosmic":2089.43,"dbsnp":2089.43}},"value":1200.54},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01},"known":{"1000genomes":0.02,"average":0.02,"cosmic":0.02,"dbsnp":0.02},"novel":{"1000genomes":0.01,"average":0.01,"cosmic":0.01,"dbsnp":0.01}},"value":0.01}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-4/qc/varQC/Tissue-Resistant-4-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-4/qc/varQC/Tissue-Resistant-4-vardict.varQC.json"},{"sample":{"name":"Tissue-Resistant-5","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-5","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":148,"average":148.0,"cosmic":148,"dbsnp":148},"known":{"1000genomes":35,"average":35.0,"cosmic":35,"dbsnp":35},"novel":{"1000genomes":113,"average":113.0,"cosmic":113,"dbsnp":113}},"value":148},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":130,"average":130.0,"cosmic":130,"dbsnp":130},"known":{"1000genomes":29,"average":29.0,"cosmic":29,"dbsnp":29},"novel":{"1000genomes":101,"average":101.0,"cosmic":101,"dbsnp":101}},"value":130},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":2,"average":2.0,"cosmic":2,"dbsnp":2},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":2,"average":2.0,"cosmic":2,"dbsnp":2}},"value":2},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":12,"average":12.0,"cosmic":12,"dbsnp":12},"known":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"novel":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6}},"value":12},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":0,"average":16.0,"cosmic":13,"dbsnp":35},"known":{"1000genomes":0,"average":12.666666666666666,"cosmic":3,"dbsnp":35},"novel":{"1000genomes":0,"average":3.3333333333333335,"cosmic":10,"dbsnp":0}},"value":13},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.0,"average":10.81,"cosmic":8.78,"dbsnp":23.65},"known":{"1000genomes":0.0,"average":36.19,"cosmic":8.57,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":2.9499999999999997,"cosmic":8.85,"dbsnp":0.0}},"value":8.78},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":0,"average":14.666666666666666,"cosmic":11,"dbsnp":33},"known":{"1000genomes":0,"average":12.0,"cosmic":3,"dbsnp":33},"novel":{"1000genomes":0,"average":2.6666666666666665,"cosmic":8,"dbsnp":0}},"value":11},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":0.0,"average":59.63666666666668,"cosmic":84.62,"dbsnp":94.29},"known":{"1000genomes":0.0,"average":64.76333333333334,"cosmic":100.0,"dbsnp":94.29},"novel":{"1000genomes":0.0,"average":26.666666666666668,"cosmic":80.0,"dbsnp":0.0}},"value":84.62},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":5e-08,"average":5e-08,"cosmic":5e-08,"dbsnp":5e-08},"known":{"1000genomes":1e-08,"average":1.0000000000000002e-08,"cosmic":1e-08,"dbsnp":1e-08},"novel":{"1000genomes":4e-08,"average":4.000000000000001e-08,"cosmic":4e-08,"dbsnp":4e-08}},"value":5e-08},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":21197035.0,"average":21197035.0,"cosmic":21197035.0,"dbsnp":21197035.0},"known":{"1000genomes":89633178.0,"average":89633178.0,"cosmic":89633178.0,"dbsnp":89633178.0},"novel":{"1000genomes":27762489.0,"average":27762489.0,"cosmic":27762489.0,"dbsnp":27762489.0}},"value":21197035.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":23.14,"average":23.14,"cosmic":23.14,"dbsnp":23.14},"known":{"1000genomes":8.2,"average":8.2,"cosmic":8.2,"dbsnp":8.2},"novel":{"1000genomes":60.5,"average":60.5,"cosmic":60.5,"dbsnp":60.5}},"value":23.14},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.81,"average":0.81,"cosmic":0.81,"dbsnp":0.81},"known":{"1000genomes":2.63,"average":2.63,"cosmic":2.63,"dbsnp":2.63},"novel":{"1000genomes":0.58,"average":0.58,"cosmic":0.58,"dbsnp":0.58}},"value":0.81}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-5/qc/varQC/Tissue-Resistant-5-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-5/qc/varQC/Tissue-Resistant-5-vardict.varQC.json"},{"sample":{"name":"Tissue-Resistant-6","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-6","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":385,"average":385.0,"cosmic":385,"dbsnp":385},"known":{"1000genomes":63,"average":63.0,"cosmic":63,"dbsnp":63},"novel":{"1000genomes":322,"average":322.0,"cosmic":322,"dbsnp":322}},"value":385},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":360,"average":360.0,"cosmic":360,"dbsnp":360},"known":{"1000genomes":53,"average":53.0,"cosmic":53,"dbsnp":53},"novel":{"1000genomes":307,"average":307.0,"cosmic":307,"dbsnp":307}},"value":360},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11},"known":{"1000genomes":5,"average":5.0,"cosmic":5,"dbsnp":5},"novel":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6}},"value":11},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":13,"average":13.0,"cosmic":13,"dbsnp":13},"known":{"1000genomes":5,"average":5.0,"cosmic":5,"dbsnp":5},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":13},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":3,"average":26.666666666666668,"cosmic":14,"dbsnp":63},"known":{"1000genomes":3,"average":24.333333333333332,"cosmic":7,"dbsnp":63},"novel":{"1000genomes":0,"average":2.3333333333333335,"cosmic":7,"dbsnp":0}},"value":14},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.78,"average":6.926666666666667,"cosmic":3.64,"dbsnp":16.36},"known":{"1000genomes":4.76,"average":38.623333333333335,"cosmic":11.11,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.7233333333333333,"cosmic":2.17,"dbsnp":0.0}},"value":3.64},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":3,"average":20.666666666666668,"cosmic":9,"dbsnp":50},"known":{"1000genomes":3,"average":19.333333333333332,"cosmic":5,"dbsnp":50},"novel":{"1000genomes":0,"average":1.3333333333333333,"cosmic":4,"dbsnp":0}},"value":9},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":100.0,"average":81.22000000000001,"cosmic":64.29,"dbsnp":79.37},"known":{"1000genomes":100.0,"average":83.60000000000001,"cosmic":71.43,"dbsnp":79.37},"novel":{"1000genomes":0.0,"average":19.046666666666667,"cosmic":57.14,"dbsnp":0.0}},"value":64.29},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07},"known":{"1000genomes":2e-08,"average":2.0000000000000004e-08,"cosmic":2e-08,"dbsnp":2e-08},"novel":{"1000genomes":1e-07,"average":1e-07,"cosmic":1e-07,"dbsnp":1e-07}},"value":1.2e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":8148470.0,"average":8148470.0,"cosmic":8148470.0,"dbsnp":8148470.0},"known":{"1000genomes":49796210.0,"average":49796210.0,"cosmic":49796210.0,"dbsnp":49796210.0},"novel":{"1000genomes":9742736.0,"average":9742736.0,"cosmic":9742736.0,"dbsnp":9742736.0}},"value":8148470.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":35.27,"average":35.27,"cosmic":35.27,"dbsnp":35.27},"known":{"1000genomes":9.29,"average":9.29,"cosmic":9.29,"dbsnp":9.29},"novel":{"1000genomes":80.75,"average":80.75,"cosmic":80.75,"dbsnp":80.75}},"value":35.27},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.31,"average":0.31,"cosmic":0.31,"dbsnp":0.31},"known":{"1000genomes":1.21,"average":1.21,"cosmic":1.21,"dbsnp":1.21},"novel":{"1000genomes":0.23,"average":0.23,"cosmic":0.23,"dbsnp":0.23}},"value":0.31}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-6/qc/varQC/Tissue-Resistant-6-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-6/qc/varQC/Tissue-Resistant-6-vardict.varQC.json"},{"sample":{"name":"Tissue-Resistant-7","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-7","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-7/Tissue-Resistant-7-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1298,"average":1298.0,"cosmic":1298,"dbsnp":1298},"known":{"1000genomes":174,"average":174.0,"cosmic":174,"dbsnp":174},"novel":{"1000genomes":1124,"average":1124.0,"cosmic":1124,"dbsnp":1124}},"value":1298},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1137,"average":1137.0,"cosmic":1137,"dbsnp":1137},"known":{"1000genomes":167,"average":167.0,"cosmic":167,"dbsnp":167},"novel":{"1000genomes":970,"average":970.0,"cosmic":970,"dbsnp":970}},"value":1137},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":74,"average":74.0,"cosmic":74,"dbsnp":74},"known":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4},"novel":{"1000genomes":70,"average":70.0,"cosmic":70,"dbsnp":70}},"value":74},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":62,"average":62.0,"cosmic":62,"dbsnp":62},"known":{"1000genomes":2,"average":2.0,"cosmic":2,"dbsnp":2},"novel":{"1000genomes":60,"average":60.0,"cosmic":60,"dbsnp":60}},"value":62},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":6,"average":84.33333333333333,"cosmic":73,"dbsnp":174},"known":{"1000genomes":6,"average":66.66666666666667,"cosmic":20,"dbsnp":174},"novel":{"1000genomes":0,"average":17.666666666666668,"cosmic":53,"dbsnp":0}},"value":73},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.46,"average":6.496666666666667,"cosmic":5.62,"dbsnp":13.41},"known":{"1000genomes":3.45,"average":38.31333333333333,"cosmic":11.49,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.5733333333333333,"cosmic":4.72,"dbsnp":0.0}},"value":5.62},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":5,"average":64.66666666666667,"cosmic":53,"dbsnp":136},"known":{"1000genomes":5,"average":53.0,"cosmic":18,"dbsnp":136},"novel":{"1000genomes":0,"average":11.666666666666666,"cosmic":35,"dbsnp":0}},"value":53},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":83.33,"average":78.03,"cosmic":72.6,"dbsnp":78.16},"known":{"1000genomes":83.33,"average":83.83,"cosmic":90.0,"dbsnp":78.16},"novel":{"1000genomes":0.0,"average":22.013333333333335,"cosmic":66.04,"dbsnp":0.0}},"value":72.6},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":4.1e-07,"average":4.1000000000000004e-07,"cosmic":4.1e-07,"dbsnp":4.1e-07},"known":{"1000genomes":6e-08,"average":6e-08,"cosmic":6e-08,"dbsnp":6e-08},"novel":{"1000genomes":3.6e-07,"average":3.6e-07,"cosmic":3.6e-07,"dbsnp":3.6e-07}},"value":4.1e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":2416919.0,"average":2416919.0,"cosmic":2416919.0,"dbsnp":2416919.0},"known":{"1000genomes":18029662.0,"average":18029662.0,"cosmic":18029662.0,"dbsnp":18029662.0},"novel":{"1000genomes":2791068.0,"average":2791068.0,"cosmic":2791068.0,"dbsnp":2791068.0}},"value":2416919.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":163.75,"average":163.75,"cosmic":163.75,"dbsnp":163.75},"known":{"1000genomes":29.83,"average":29.83,"cosmic":29.83,"dbsnp":29.83},"novel":{"1000genomes":565.5,"average":565.5,"cosmic":565.5,"dbsnp":565.5}},"value":163.75},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.83,"average":0.83,"cosmic":0.83,"dbsnp":0.83},"known":{"1000genomes":1.46,"average":1.46,"cosmic":1.46,"dbsnp":1.46},"novel":{"1000genomes":0.75,"average":0.75,"cosmic":0.75,"dbsnp":0.75}},"value":0.83}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-7/qc/varQC/Tissue-Resistant-7-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-7/qc/varQC/Tissue-Resistant-7-vardict.varQC.json"},{"sample":{"name":"Tissue-Resistant-8","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-8","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-8/Tissue-Resistant-8-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":1494,"average":1494.0,"cosmic":1494,"dbsnp":1494},"known":{"1000genomes":103,"average":103.0,"cosmic":103,"dbsnp":103},"novel":{"1000genomes":1391,"average":1391.0,"cosmic":1391,"dbsnp":1391}},"value":1494},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":1475,"average":1475.0,"cosmic":1475,"dbsnp":1475},"known":{"1000genomes":98,"average":98.0,"cosmic":98,"dbsnp":98},"novel":{"1000genomes":1377,"average":1377.0,"cosmic":1377,"dbsnp":1377}},"value":1475},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4},"known":{"1000genomes":1,"average":1.0,"cosmic":1,"dbsnp":1},"novel":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3}},"value":4},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":12,"average":12.0,"cosmic":12,"dbsnp":12},"known":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":12},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":3,"average":56.666666666666664,"cosmic":64,"dbsnp":103},"known":{"1000genomes":3,"average":37.333333333333336,"cosmic":6,"dbsnp":103},"novel":{"1000genomes":0,"average":19.333333333333332,"cosmic":58,"dbsnp":0}},"value":64},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.2,"average":3.7900000000000005,"cosmic":4.28,"dbsnp":6.89},"known":{"1000genomes":2.91,"average":36.24666666666666,"cosmic":5.83,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.39,"cosmic":4.17,"dbsnp":0.0}},"value":4.28},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":2,"average":21.666666666666668,"cosmic":27,"dbsnp":36},"known":{"1000genomes":2,"average":14.333333333333334,"cosmic":5,"dbsnp":36},"novel":{"1000genomes":0,"average":7.333333333333333,"cosmic":22,"dbsnp":0}},"value":27},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":66.67,"average":47.93666666666667,"cosmic":42.19,"dbsnp":34.95},"known":{"1000genomes":66.67,"average":61.65,"cosmic":83.33,"dbsnp":34.95},"novel":{"1000genomes":0.0,"average":12.643333333333333,"cosmic":37.93,"dbsnp":0.0}},"value":42.19},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":4.8e-07,"average":4.8e-07,"cosmic":4.8e-07,"dbsnp":4.8e-07},"known":{"1000genomes":3e-08,"average":3e-08,"cosmic":3e-08,"dbsnp":3e-08},"novel":{"1000genomes":4.4e-07,"average":4.4e-07,"cosmic":4.4e-07,"dbsnp":4.4e-07}},"value":4.8e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":2099840.0,"average":2099840.0,"cosmic":2099840.0,"dbsnp":2099840.0},"known":{"1000genomes":30457876.0,"average":30457876.0,"cosmic":30457876.0,"dbsnp":30457876.0},"novel":{"1000genomes":2255328.0,"average":2255328.0,"cosmic":2255328.0,"dbsnp":2255328.0}},"value":2099840.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":751.5,"average":751.5,"cosmic":751.5,"dbsnp":751.5},"known":{"1000genomes":108.0,"average":108.0,"cosmic":108.0,"dbsnp":108.0},"novel":{"1000genomes":1395.0,"average":1395.0,"cosmic":1395.0,"dbsnp":1395.0}},"value":751.5},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.04,"average":0.04,"cosmic":0.04,"dbsnp":0.04},"known":{"1000genomes":0.31,"average":0.31,"cosmic":0.31,"dbsnp":0.31},"novel":{"1000genomes":0.03,"average":0.03,"cosmic":0.03,"dbsnp":0.03}},"value":0.04}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-8/qc/varQC/Tissue-Resistant-8-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-8/qc/varQC/Tissue-Resistant-8-vardict.varQC.json"},{"sample":{"name":"Tissue-Resistant-9","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-9","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-9/Tissue-Resistant-9-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":9527,"average":9527.0,"cosmic":9527,"dbsnp":9527},"known":{"1000genomes":591,"average":591.0,"cosmic":591,"dbsnp":591},"novel":{"1000genomes":8936,"average":8936.0,"cosmic":8936,"dbsnp":8936}},"value":9527},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":9487,"average":9487.0,"cosmic":9487,"dbsnp":9487},"known":{"1000genomes":584,"average":584.0,"cosmic":584,"dbsnp":584},"novel":{"1000genomes":8903,"average":8903.0,"cosmic":8903,"dbsnp":8903}},"value":9487},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":10,"average":10.0,"cosmic":10,"dbsnp":10},"known":{"1000genomes":2,"average":2.0,"cosmic":2,"dbsnp":2},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":10},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":23,"average":23.0,"cosmic":23,"dbsnp":23},"known":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4},"novel":{"1000genomes":19,"average":19.0,"cosmic":19,"dbsnp":19}},"value":23},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":16,"average":329.0,"cosmic":380,"dbsnp":591},"known":{"1000genomes":16,"average":217.66666666666666,"cosmic":46,"dbsnp":591},"novel":{"1000genomes":0,"average":111.33333333333333,"cosmic":334,"dbsnp":0}},"value":380},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.17,"average":3.453333333333333,"cosmic":3.99,"dbsnp":6.2},"known":{"1000genomes":2.71,"average":36.83,"cosmic":7.78,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.2466666666666668,"cosmic":3.74,"dbsnp":0.0}},"value":3.99},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":3,"average":90.33333333333333,"cosmic":140,"dbsnp":128},"known":{"1000genomes":3,"average":49.0,"cosmic":16,"dbsnp":128},"novel":{"1000genomes":0,"average":41.333333333333336,"cosmic":124,"dbsnp":0}},"value":140},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":18.75,"average":25.75,"cosmic":36.84,"dbsnp":21.66},"known":{"1000genomes":18.75,"average":25.063333333333333,"cosmic":34.78,"dbsnp":21.66},"novel":{"1000genomes":0.0,"average":12.376666666666667,"cosmic":37.13,"dbsnp":0.0}},"value":36.84},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":3.04e-06,"average":3.04e-06,"cosmic":3.04e-06,"dbsnp":3.04e-06},"known":{"1000genomes":1.9e-07,"average":1.9e-07,"cosmic":1.9e-07,"dbsnp":1.9e-07},"novel":{"1000genomes":2.85e-06,"average":2.85e-06,"cosmic":2.85e-06,"dbsnp":2.85e-06}},"value":3.04e-06},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":329291.0,"average":329291.0,"cosmic":329291.0,"dbsnp":329291.0},"known":{"1000genomes":5308225.0,"average":5308225.0,"cosmic":5308225.0,"dbsnp":5308225.0},"novel":{"1000genomes":351069.0,"average":351069.0,"cosmic":351069.0,"dbsnp":351069.0}},"value":329291.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":2384.5,"average":2384.5,"cosmic":2384.5,"dbsnp":2384.5},"known":{"1000genomes":148.0,"average":148.0,"cosmic":148.0,"dbsnp":148.0},"novel":{"1000genomes":8946.0,"average":8946.0,"cosmic":8946.0,"dbsnp":8946.0}},"value":2384.5},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.02,"average":0.02,"cosmic":0.02,"dbsnp":0.02},"known":{"1000genomes":0.07,"average":0.07,"cosmic":0.07,"dbsnp":0.07},"novel":{"1000genomes":0.02,"average":0.02,"cosmic":0.02,"dbsnp":0.02}},"value":0.02}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-9/qc/varQC/Tissue-Resistant-9-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Resistant-9/qc/varQC/Tissue-Resistant-9-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-10","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-10","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-10/Tissue-Sensitive-10-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":357,"average":357.0,"cosmic":357,"dbsnp":357},"known":{"1000genomes":54,"average":54.0,"cosmic":54,"dbsnp":54},"novel":{"1000genomes":303,"average":303.0,"cosmic":303,"dbsnp":303}},"value":357},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":324,"average":324.0,"cosmic":324,"dbsnp":324},"known":{"1000genomes":44,"average":44.0,"cosmic":44,"dbsnp":44},"novel":{"1000genomes":280,"average":280.0,"cosmic":280,"dbsnp":280}},"value":324},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"known":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4},"novel":{"1000genomes":2,"average":2.0,"cosmic":2,"dbsnp":2}},"value":6},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":22,"average":22.0,"cosmic":22,"dbsnp":22},"known":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"novel":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16}},"value":22},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":3,"average":22.666666666666668,"cosmic":11,"dbsnp":54},"known":{"1000genomes":3,"average":19.666666666666668,"cosmic":2,"dbsnp":54},"novel":{"1000genomes":0,"average":3.0,"cosmic":9,"dbsnp":0}},"value":11},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.84,"average":6.3500000000000005,"cosmic":3.08,"dbsnp":15.13},"known":{"1000genomes":5.56,"average":36.42,"cosmic":3.7,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.9900000000000001,"cosmic":2.97,"dbsnp":0.0}},"value":3.08},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":2,"average":14.0,"cosmic":6,"dbsnp":34},"known":{"1000genomes":2,"average":12.666666666666666,"cosmic":2,"dbsnp":34},"novel":{"1000genomes":0,"average":1.3333333333333333,"cosmic":4,"dbsnp":0}},"value":6},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":66.67,"average":61.39333333333334,"cosmic":54.55,"dbsnp":62.96},"known":{"1000genomes":66.67,"average":76.54333333333334,"cosmic":100.0,"dbsnp":62.96},"novel":{"1000genomes":0.0,"average":14.813333333333333,"cosmic":44.44,"dbsnp":0.0}},"value":54.55},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":1.1e-07,"average":1.1e-07,"cosmic":1.1e-07,"dbsnp":1.1e-07},"known":{"1000genomes":2e-08,"average":2.0000000000000004e-08,"cosmic":2e-08,"dbsnp":2e-08},"novel":{"1000genomes":1e-07,"average":1e-07,"cosmic":1e-07,"dbsnp":1e-07}},"value":1.1e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":8787566.0,"average":8787566.0,"cosmic":8787566.0,"dbsnp":8787566.0},"known":{"1000genomes":58095578.0,"average":58095578.0,"cosmic":58095578.0,"dbsnp":58095578.0},"novel":{"1000genomes":10353667.0,"average":10353667.0,"cosmic":10353667.0,"dbsnp":10353667.0}},"value":8787566.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":90.25,"average":90.25,"cosmic":90.25,"dbsnp":90.25},"known":{"1000genomes":18.33,"average":18.33,"cosmic":18.33,"dbsnp":18.33},"novel":{"1000genomes":306.0,"average":306.0,"cosmic":306.0,"dbsnp":306.0}},"value":90.25},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.29,"average":0.29,"cosmic":0.29,"dbsnp":0.29},"known":{"1000genomes":0.83,"average":0.83,"cosmic":0.83,"dbsnp":0.83},"novel":{"1000genomes":0.23,"average":0.23,"cosmic":0.23,"dbsnp":0.23}},"value":0.29}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-10/qc/varQC/Tissue-Sensitive-10-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-10/qc/varQC/Tissue-Sensitive-10-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-1","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-1","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-1/Tissue-Sensitive-1-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":219,"average":219.0,"cosmic":219,"dbsnp":219},"known":{"1000genomes":55,"average":55.0,"cosmic":55,"dbsnp":55},"novel":{"1000genomes":164,"average":164.0,"cosmic":164,"dbsnp":164}},"value":219},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":200,"average":200.0,"cosmic":200,"dbsnp":200},"known":{"1000genomes":50,"average":50.0,"cosmic":50,"dbsnp":50},"novel":{"1000genomes":150,"average":150.0,"cosmic":150,"dbsnp":150}},"value":200},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"known":{"1000genomes":0,"average":0.0,"cosmic":0,"dbsnp":0},"novel":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6}},"value":6},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":13,"average":13.0,"cosmic":13,"dbsnp":13},"known":{"1000genomes":5,"average":5.0,"cosmic":5,"dbsnp":5},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":13},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":13,"average":26.0,"cosmic":10,"dbsnp":55},"known":{"1000genomes":13,"average":23.666666666666668,"cosmic":3,"dbsnp":55},"novel":{"1000genomes":0,"average":2.3333333333333335,"cosmic":7,"dbsnp":0}},"value":10},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":5.94,"average":11.873333333333335,"cosmic":4.57,"dbsnp":25.11},"known":{"1000genomes":23.64,"average":43.03,"cosmic":5.45,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4233333333333331,"cosmic":4.27,"dbsnp":0.0}},"value":4.57},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":12,"average":21.333333333333332,"cosmic":4,"dbsnp":48},"known":{"1000genomes":12,"average":21.0,"cosmic":3,"dbsnp":48},"novel":{"1000genomes":0,"average":0.3333333333333333,"cosmic":1,"dbsnp":0}},"value":4},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":92.31,"average":73.19333333333333,"cosmic":40.0,"dbsnp":87.27},"known":{"1000genomes":92.31,"average":93.19333333333333,"cosmic":100.0,"dbsnp":87.27},"novel":{"1000genomes":0.0,"average":4.763333333333333,"cosmic":14.29,"dbsnp":0.0}},"value":40.0},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":7e-08,"average":7e-08,"cosmic":7e-08,"dbsnp":7e-08},"known":{"1000genomes":2e-08,"average":2.0000000000000004e-08,"cosmic":2e-08,"dbsnp":2e-08},"novel":{"1000genomes":5e-08,"average":5e-08,"cosmic":5e-08,"dbsnp":5e-08}},"value":7e-08},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":14324937.0,"average":14324937.0,"cosmic":14324937.0,"dbsnp":14324937.0},"known":{"1000genomes":57039295.0,"average":57039295.0,"cosmic":57039295.0,"dbsnp":57039295.0},"novel":{"1000genomes":19129032.0,"average":19129032.0,"cosmic":19129032.0,"dbsnp":19129032.0}},"value":14324937.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":55.0,"average":55.0,"cosmic":55.0,"dbsnp":55.0},"known":{"1000genomes":18.33,"average":18.33,"cosmic":18.33,"dbsnp":18.33},"novel":{"1000genomes":165.0,"average":165.0,"cosmic":165.0,"dbsnp":165.0}},"value":55.0},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.64,"average":0.64,"cosmic":0.64,"dbsnp":0.64},"known":{"1000genomes":1.63,"average":1.63,"cosmic":1.63,"dbsnp":1.63},"novel":{"1000genomes":0.46,"average":0.46,"cosmic":0.46,"dbsnp":0.46}},"value":0.64}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-1/qc/varQC/Tissue-Sensitive-1-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-1/qc/varQC/Tissue-Sensitive-1-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-2","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-2","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-2/Tissue-Sensitive-2-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":398,"average":398.0,"cosmic":398,"dbsnp":398},"known":{"1000genomes":82,"average":82.0,"cosmic":82,"dbsnp":82},"novel":{"1000genomes":316,"average":316.0,"cosmic":316,"dbsnp":316}},"value":398},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":356,"average":356.0,"cosmic":356,"dbsnp":356},"known":{"1000genomes":65,"average":65.0,"cosmic":65,"dbsnp":65},"novel":{"1000genomes":291,"average":291.0,"cosmic":291,"dbsnp":291}},"value":356},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11},"known":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7},"novel":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4}},"value":11},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":30,"average":30.0,"cosmic":30,"dbsnp":30},"known":{"1000genomes":10,"average":10.0,"cosmic":10,"dbsnp":10},"novel":{"1000genomes":20,"average":20.0,"cosmic":20,"dbsnp":20}},"value":30},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":8,"average":35.666666666666664,"cosmic":17,"dbsnp":82},"known":{"1000genomes":8,"average":33.0,"cosmic":9,"dbsnp":82},"novel":{"1000genomes":0,"average":2.6666666666666665,"cosmic":8,"dbsnp":0}},"value":17},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":2.01,"average":8.959999999999999,"cosmic":4.27,"dbsnp":20.6},"known":{"1000genomes":9.76,"average":40.24666666666667,"cosmic":10.98,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.8433333333333333,"cosmic":2.53,"dbsnp":0.0}},"value":4.27},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":8,"average":26.333333333333332,"cosmic":10,"dbsnp":61},"known":{"1000genomes":8,"average":24.333333333333332,"cosmic":4,"dbsnp":61},"novel":{"1000genomes":0,"average":2.0,"cosmic":6,"dbsnp":0}},"value":10},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":100.0,"average":77.73666666666666,"cosmic":58.82,"dbsnp":74.39},"known":{"1000genomes":100.0,"average":72.94333333333333,"cosmic":44.44,"dbsnp":74.39},"novel":{"1000genomes":0.0,"average":25.0,"cosmic":75.0,"dbsnp":0.0}},"value":58.82},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":1.3e-07,"average":1.3e-07,"cosmic":1.3e-07,"dbsnp":1.3e-07},"known":{"1000genomes":3e-08,"average":3e-08,"cosmic":3e-08,"dbsnp":3e-08},"novel":{"1000genomes":1e-07,"average":1e-07,"cosmic":1e-07,"dbsnp":1e-07}},"value":1.3e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":7882314.0,"average":7882314.0,"cosmic":7882314.0,"dbsnp":7882314.0},"known":{"1000genomes":38258064.0,"average":38258064.0,"cosmic":38258064.0,"dbsnp":38258064.0},"novel":{"1000genomes":9927725.0,"average":9927725.0,"cosmic":9927725.0,"dbsnp":9927725.0}},"value":7882314.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":19.29,"average":19.29,"cosmic":19.29,"dbsnp":19.29},"known":{"1000genomes":7.73,"average":7.73,"cosmic":7.73,"dbsnp":7.73},"novel":{"1000genomes":32.0,"average":32.0,"cosmic":32.0,"dbsnp":32.0}},"value":19.29},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.41,"average":0.41,"cosmic":0.41,"dbsnp":0.41},"known":{"1000genomes":1.17,"average":1.17,"cosmic":1.17,"dbsnp":1.17},"novel":{"1000genomes":0.31,"average":0.31,"cosmic":0.31,"dbsnp":0.31}},"value":0.41}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-2/qc/varQC/Tissue-Sensitive-2-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-2/qc/varQC/Tissue-Sensitive-2-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-3","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-3","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-3/Tissue-Sensitive-3-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":95,"average":95.0,"cosmic":95,"dbsnp":95},"known":{"1000genomes":34,"average":34.0,"cosmic":34,"dbsnp":34},"novel":{"1000genomes":61,"average":61.0,"cosmic":61,"dbsnp":61}},"value":95},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":73,"average":73.0,"cosmic":73,"dbsnp":73},"known":{"1000genomes":22,"average":22.0,"cosmic":22,"dbsnp":22},"novel":{"1000genomes":51,"average":51.0,"cosmic":51,"dbsnp":51}},"value":73},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"known":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3},"novel":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3}},"value":6},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16},"known":{"1000genomes":9,"average":9.0,"cosmic":9,"dbsnp":9},"novel":{"1000genomes":7,"average":7.0,"cosmic":7,"dbsnp":7}},"value":16},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":3,"average":15.0,"cosmic":8,"dbsnp":34},"known":{"1000genomes":3,"average":14.0,"cosmic":5,"dbsnp":34},"novel":{"1000genomes":0,"average":1.0,"cosmic":3,"dbsnp":0}},"value":8},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":3.16,"average":15.790000000000001,"cosmic":8.42,"dbsnp":35.79},"known":{"1000genomes":8.82,"average":41.17666666666667,"cosmic":14.71,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.64,"cosmic":4.92,"dbsnp":0.0}},"value":8.42},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":2,"average":12.333333333333334,"cosmic":6,"dbsnp":29},"known":{"1000genomes":2,"average":11.666666666666666,"cosmic":4,"dbsnp":29},"novel":{"1000genomes":0,"average":0.6666666666666666,"cosmic":2,"dbsnp":0}},"value":6},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":66.67,"average":75.65333333333334,"cosmic":75.0,"dbsnp":85.29},"known":{"1000genomes":66.67,"average":77.32000000000001,"cosmic":80.0,"dbsnp":85.29},"novel":{"1000genomes":0.0,"average":22.223333333333333,"cosmic":66.67,"dbsnp":0.0}},"value":75.0},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":3e-08,"average":3e-08,"cosmic":3e-08,"dbsnp":3e-08},"known":{"1000genomes":1e-08,"average":1.0000000000000002e-08,"cosmic":1e-08,"dbsnp":1e-08},"novel":{"1000genomes":2e-08,"average":2.0000000000000004e-08,"cosmic":2e-08,"dbsnp":2e-08}},"value":3e-08},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":33022750.0,"average":33022750.0,"cosmic":33022750.0,"dbsnp":33022750.0},"known":{"1000genomes":92269448.0,"average":92269448.0,"cosmic":92269448.0,"dbsnp":92269448.0},"novel":{"1000genomes":51428873.0,"average":51428873.0,"cosmic":51428873.0,"dbsnp":51428873.0}},"value":33022750.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":18.17,"average":18.17,"cosmic":18.17,"dbsnp":18.17},"known":{"1000genomes":6.6,"average":6.599999999999999,"cosmic":6.6,"dbsnp":6.6},"novel":{"1000genomes":76.0,"average":76.0,"cosmic":76.0,"dbsnp":76.0}},"value":18.17},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.35,"average":0.3499999999999999,"cosmic":0.35,"dbsnp":0.35},"known":{"1000genomes":1.75,"average":1.75,"cosmic":1.75,"dbsnp":1.75},"novel":{"1000genomes":0.11,"average":0.11,"cosmic":0.11,"dbsnp":0.11}},"value":0.35}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-3/qc/varQC/Tissue-Sensitive-3-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-3/qc/varQC/Tissue-Sensitive-3-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-5","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-5","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-5/Tissue-Sensitive-5-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":240,"average":240.0,"cosmic":240,"dbsnp":240},"known":{"1000genomes":61,"average":61.0,"cosmic":61,"dbsnp":61},"novel":{"1000genomes":179,"average":179.0,"cosmic":179,"dbsnp":179}},"value":240},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":212,"average":212.0,"cosmic":212,"dbsnp":212},"known":{"1000genomes":49,"average":49.0,"cosmic":49,"dbsnp":49},"novel":{"1000genomes":163,"average":163.0,"cosmic":163,"dbsnp":163}},"value":212},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":9,"average":9.0,"cosmic":9,"dbsnp":9},"known":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"novel":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3}},"value":9},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":14,"average":14.0,"cosmic":14,"dbsnp":14},"known":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":14},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":7,"average":27.666666666666668,"cosmic":15,"dbsnp":61},"known":{"1000genomes":7,"average":25.333333333333332,"cosmic":8,"dbsnp":61},"novel":{"1000genomes":0,"average":2.3333333333333335,"cosmic":7,"dbsnp":0}},"value":15},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":2.92,"average":11.530000000000001,"cosmic":6.25,"dbsnp":25.42},"known":{"1000genomes":11.48,"average":41.53,"cosmic":13.11,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.3033333333333335,"cosmic":3.91,"dbsnp":0.0}},"value":6.25},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":7,"average":23.333333333333332,"cosmic":10,"dbsnp":53},"known":{"1000genomes":7,"average":22.333333333333332,"cosmic":7,"dbsnp":53},"novel":{"1000genomes":0,"average":1.0,"cosmic":3,"dbsnp":0}},"value":10},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":100.0,"average":84.52,"cosmic":66.67,"dbsnp":86.89},"known":{"1000genomes":100.0,"average":91.46333333333332,"cosmic":87.5,"dbsnp":86.89},"novel":{"1000genomes":0.0,"average":14.286666666666667,"cosmic":42.86,"dbsnp":0.0}},"value":66.67},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":8e-08,"average":8.000000000000001e-08,"cosmic":8e-08,"dbsnp":8e-08},"known":{"1000genomes":2e-08,"average":2.0000000000000004e-08,"cosmic":2e-08,"dbsnp":2e-08},"novel":{"1000genomes":6e-08,"average":6e-08,"cosmic":6e-08,"dbsnp":6e-08}},"value":8e-08},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":13071505.0,"average":13071505.0,"cosmic":13071505.0,"dbsnp":13071505.0},"known":{"1000genomes":51428873.0,"average":51428873.0,"cosmic":51428873.0,"dbsnp":51428873.0},"novel":{"1000genomes":17526040.0,"average":17526040.0,"cosmic":17526040.0,"dbsnp":17526040.0}},"value":13071505.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":29.2,"average":29.2,"cosmic":29.2,"dbsnp":29.2},"known":{"1000genomes":9.71,"average":9.71,"cosmic":9.71,"dbsnp":9.71},"novel":{"1000genomes":74.67,"average":74.67,"cosmic":74.67,"dbsnp":74.67}},"value":29.2},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.7,"average":0.6999999999999998,"cosmic":0.7,"dbsnp":0.7},"known":{"1000genomes":1.33,"average":1.33,"cosmic":1.33,"dbsnp":1.33},"novel":{"1000genomes":0.57,"average":0.57,"cosmic":0.57,"dbsnp":0.57}},"value":0.7}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-5/qc/varQC/Tissue-Sensitive-5-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-5/qc/varQC/Tissue-Sensitive-5-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-6","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-6","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-6/Tissue-Sensitive-6-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":401,"average":401.0,"cosmic":401,"dbsnp":401},"known":{"1000genomes":62,"average":62.0,"cosmic":62,"dbsnp":62},"novel":{"1000genomes":339,"average":339.0,"cosmic":339,"dbsnp":339}},"value":401},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":361,"average":361.0,"cosmic":361,"dbsnp":361},"known":{"1000genomes":51,"average":51.0,"cosmic":51,"dbsnp":51},"novel":{"1000genomes":310,"average":310.0,"cosmic":310,"dbsnp":310}},"value":361},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":11,"average":11.0,"cosmic":11,"dbsnp":11},"known":{"1000genomes":5,"average":5.0,"cosmic":5,"dbsnp":5},"novel":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6}},"value":11},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":22,"average":22.0,"cosmic":22,"dbsnp":22},"known":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"novel":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16}},"value":22},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":6,"average":26.666666666666668,"cosmic":12,"dbsnp":62},"known":{"1000genomes":6,"average":24.0,"cosmic":4,"dbsnp":62},"novel":{"1000genomes":0,"average":2.6666666666666665,"cosmic":8,"dbsnp":0}},"value":12},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":1.5,"average":6.650000000000001,"cosmic":2.99,"dbsnp":15.46},"known":{"1000genomes":9.68,"average":38.71,"cosmic":6.45,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":0.7866666666666666,"cosmic":2.36,"dbsnp":0.0}},"value":2.99},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":6,"average":21.0,"cosmic":7,"dbsnp":50},"known":{"1000genomes":6,"average":20.0,"cosmic":4,"dbsnp":50},"novel":{"1000genomes":0,"average":1.0,"cosmic":3,"dbsnp":0}},"value":7},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":100.0,"average":79.66000000000001,"cosmic":58.33,"dbsnp":80.65},"known":{"1000genomes":100.0,"average":93.55,"cosmic":100.0,"dbsnp":80.65},"novel":{"1000genomes":0.0,"average":12.5,"cosmic":37.5,"dbsnp":0.0}},"value":58.33},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":1.3e-07,"average":1.3e-07,"cosmic":1.3e-07,"dbsnp":1.3e-07},"known":{"1000genomes":2e-08,"average":2.0000000000000004e-08,"cosmic":2e-08,"dbsnp":2e-08},"novel":{"1000genomes":1.1e-07,"average":1.1e-07,"cosmic":1.1e-07,"dbsnp":1.1e-07}},"value":1.3e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":7823344.0,"average":7823344.0,"cosmic":7823344.0,"dbsnp":7823344.0},"known":{"1000genomes":50599375.0,"average":50599375.0,"cosmic":50599375.0,"dbsnp":50599375.0},"novel":{"1000genomes":9254163.0,"average":9254163.0,"cosmic":9254163.0,"dbsnp":9254163.0}},"value":7823344.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":38.82,"average":38.82,"cosmic":38.82,"dbsnp":38.82},"known":{"1000genomes":8.13,"average":8.13,"cosmic":8.13,"dbsnp":8.13},"novel":{"1000genomes":120.67,"average":120.67,"cosmic":120.67,"dbsnp":120.67}},"value":38.82},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.68,"average":0.68,"cosmic":0.68,"dbsnp":0.68},"known":{"1000genomes":1.83,"average":1.83,"cosmic":1.83,"dbsnp":1.83},"novel":{"1000genomes":0.57,"average":0.57,"cosmic":0.57,"dbsnp":0.57}},"value":0.68}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-6/qc/varQC/Tissue-Sensitive-6-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-6/qc/varQC/Tissue-Sensitive-6-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-7","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-7","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":196,"average":196.0,"cosmic":196,"dbsnp":196},"known":{"1000genomes":37,"average":37.0,"cosmic":37,"dbsnp":37},"novel":{"1000genomes":159,"average":159.0,"cosmic":159,"dbsnp":159}},"value":196},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":174,"average":174.0,"cosmic":174,"dbsnp":174},"known":{"1000genomes":29,"average":29.0,"cosmic":29,"dbsnp":29},"novel":{"1000genomes":145,"average":145.0,"cosmic":145,"dbsnp":145}},"value":174},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8},"known":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3},"novel":{"1000genomes":5,"average":5.0,"cosmic":5,"dbsnp":5}},"value":8},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":13,"average":13.0,"cosmic":13,"dbsnp":13},"known":{"1000genomes":5,"average":5.0,"cosmic":5,"dbsnp":5},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":13},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":4,"average":17.0,"cosmic":10,"dbsnp":37},"known":{"1000genomes":4,"average":14.333333333333334,"cosmic":2,"dbsnp":37},"novel":{"1000genomes":0,"average":2.6666666666666665,"cosmic":8,"dbsnp":0}},"value":10},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":2.04,"average":8.673333333333332,"cosmic":5.1,"dbsnp":18.88},"known":{"1000genomes":10.81,"average":38.74,"cosmic":5.41,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.6766666666666667,"cosmic":5.03,"dbsnp":0.0}},"value":5.1},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":4,"average":13.0,"cosmic":5,"dbsnp":30},"known":{"1000genomes":4,"average":11.666666666666666,"cosmic":1,"dbsnp":30},"novel":{"1000genomes":0,"average":1.3333333333333333,"cosmic":4,"dbsnp":0}},"value":5},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":100.0,"average":77.02666666666666,"cosmic":50.0,"dbsnp":81.08},"known":{"1000genomes":100.0,"average":77.02666666666666,"cosmic":50.0,"dbsnp":81.08},"novel":{"1000genomes":0.0,"average":16.666666666666668,"cosmic":50.0,"dbsnp":0.0}},"value":50.0},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":6e-08,"average":6e-08,"cosmic":6e-08,"dbsnp":6e-08},"known":{"1000genomes":1e-08,"average":1.0000000000000002e-08,"cosmic":1e-08,"dbsnp":1e-08},"novel":{"1000genomes":5e-08,"average":5e-08,"cosmic":5e-08,"dbsnp":5e-08}},"value":6e-08},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":16005924.0,"average":16005924.0,"cosmic":16005924.0,"dbsnp":16005924.0},"known":{"1000genomes":84788142.0,"average":84788142.0,"cosmic":84788142.0,"dbsnp":84788142.0},"novel":{"1000genomes":19730573.0,"average":19730573.0,"cosmic":19730573.0,"dbsnp":19730573.0}},"value":16005924.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":207.0,"average":207.0,"cosmic":207.0,"dbsnp":207.0},"known":{"1000genomes":42.0,"average":42.0,"cosmic":42.0,"dbsnp":42.0},"novel":{"1000genomes":165.0,"average":165.0,"cosmic":165.0,"dbsnp":165.0}},"value":207.0},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.23,"average":0.23,"cosmic":0.23,"dbsnp":0.23},"known":{"1000genomes":1.23,"average":1.23,"cosmic":1.23,"dbsnp":1.23},"novel":{"1000genomes":0.13,"average":0.13,"cosmic":0.13,"dbsnp":0.13}},"value":0.23}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-7/qc/varQC/Tissue-Sensitive-7-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-7/qc/varQC/Tissue-Sensitive-7-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-8","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-8","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":438,"average":438.0,"cosmic":438,"dbsnp":438},"known":{"1000genomes":64,"average":64.0,"cosmic":64,"dbsnp":64},"novel":{"1000genomes":374,"average":374.0,"cosmic":374,"dbsnp":374}},"value":438},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":414,"average":414.0,"cosmic":414,"dbsnp":414},"known":{"1000genomes":53,"average":53.0,"cosmic":53,"dbsnp":53},"novel":{"1000genomes":361,"average":361.0,"cosmic":361,"dbsnp":361}},"value":414},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"known":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3},"novel":{"1000genomes":3,"average":3.0,"cosmic":3,"dbsnp":3}},"value":6},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":16,"average":16.0,"cosmic":16,"dbsnp":16},"known":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8},"novel":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8}},"value":16},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":4,"average":33.0,"cosmic":31,"dbsnp":64},"known":{"1000genomes":4,"average":26.333333333333332,"cosmic":11,"dbsnp":64},"novel":{"1000genomes":0,"average":6.666666666666667,"cosmic":20,"dbsnp":0}},"value":31},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.91,"average":7.533333333333334,"cosmic":7.08,"dbsnp":14.61},"known":{"1000genomes":6.25,"average":41.14666666666667,"cosmic":17.19,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.7833333333333332,"cosmic":5.35,"dbsnp":0.0}},"value":7.08},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":3,"average":24.333333333333332,"cosmic":21,"dbsnp":49},"known":{"1000genomes":3,"average":20.666666666666668,"cosmic":10,"dbsnp":49},"novel":{"1000genomes":0,"average":3.6666666666666665,"cosmic":11,"dbsnp":0}},"value":21},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":75.0,"average":73.10000000000001,"cosmic":67.74,"dbsnp":76.56},"known":{"1000genomes":75.0,"average":80.82333333333334,"cosmic":90.91,"dbsnp":76.56},"novel":{"1000genomes":0.0,"average":18.333333333333332,"cosmic":55.0,"dbsnp":0.0}},"value":67.74},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":1.4e-07,"average":1.4e-07,"cosmic":1.4e-07,"dbsnp":1.4e-07},"known":{"1000genomes":2e-08,"average":2.0000000000000004e-08,"cosmic":2e-08,"dbsnp":2e-08},"novel":{"1000genomes":1.2e-07,"average":1.2e-07,"cosmic":1.2e-07,"dbsnp":1.2e-07}},"value":1.4e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":7162468.0,"average":7162468.0,"cosmic":7162468.0,"dbsnp":7162468.0},"known":{"1000genomes":49018144.0,"average":49018144.0,"cosmic":49018144.0,"dbsnp":49018144.0},"novel":{"1000genomes":8388131.0,"average":8388131.0,"cosmic":8388131.0,"dbsnp":8388131.0}},"value":7162468.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":43.7,"average":43.70000000000001,"cosmic":43.7,"dbsnp":43.7},"known":{"1000genomes":8.57,"average":8.57,"cosmic":8.57,"dbsnp":8.57},"novel":{"1000genomes":125.67,"average":125.67,"cosmic":125.67,"dbsnp":125.67}},"value":43.7},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.28,"average":0.28,"cosmic":0.28,"dbsnp":0.28},"known":{"1000genomes":1.52,"average":1.5200000000000002,"cosmic":1.52,"dbsnp":1.52},"novel":{"1000genomes":0.2,"average":0.20000000000000004,"cosmic":0.2,"dbsnp":0.2}},"value":0.28}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-8/qc/varQC/Tissue-Sensitive-8-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-8/qc/varQC/Tissue-Sensitive-8-vardict.varQC.json"},{"sample":{"name":"Tissue-Sensitive-9","bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/SureSelect_Human_AllExon_V5.bed","vcf_by_caller":{"vardict":null,"mutect":null,"freebayes":null},"phenotype":"tumor","dirpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-9","bam":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-9/Tissue-Sensitive-9-ready.bam"},"records":[{"metric":{"unit":"","quality":"More is better","description":["Total variants evaluated"],"short_name":["Total"],"name":"nEvalVariants"},"meta":{"all":{"1000genomes":357,"average":357.0,"cosmic":357,"dbsnp":357},"known":{"1000genomes":57,"average":57.0,"cosmic":57,"dbsnp":57},"novel":{"1000genomes":300,"average":300.0,"cosmic":300,"dbsnp":300}},"value":357},{"metric":{"unit":"","quality":"More is better","description":["SNPs"],"short_name":["SNP"],"name":"nSNPs"},"meta":{"all":{"1000genomes":324,"average":324.0,"cosmic":324,"dbsnp":324},"known":{"1000genomes":44,"average":44.0,"cosmic":44,"dbsnp":44},"novel":{"1000genomes":280,"average":280.0,"cosmic":280,"dbsnp":280}},"value":324},{"metric":{"unit":"","quality":"More is better","description":["Insertions"],"short_name":["Ins"],"name":"nInsertions"},"meta":{"all":{"1000genomes":6,"average":6.0,"cosmic":6,"dbsnp":6},"known":{"1000genomes":2,"average":2.0,"cosmic":2,"dbsnp":2},"novel":{"1000genomes":4,"average":4.0,"cosmic":4,"dbsnp":4}},"value":6},{"metric":{"unit":"","quality":"More is better","description":["Deletions"],"short_name":["Del"],"name":"nDeletions"},"meta":{"all":{"1000genomes":22,"average":22.0,"cosmic":22,"dbsnp":22},"known":{"1000genomes":8,"average":8.0,"cosmic":8,"dbsnp":8},"novel":{"1000genomes":14,"average":14.0,"cosmic":14,"dbsnp":14}},"value":22},{"metric":{"unit":"","quality":"More is better","description":["Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"],"short_name":["At comp"],"name":"nVariantsAtComp"},"meta":{"all":{"1000genomes":3,"average":26.333333333333332,"cosmic":19,"dbsnp":57},"known":{"1000genomes":3,"average":22.0,"cosmic":6,"dbsnp":57},"novel":{"1000genomes":0,"average":4.333333333333333,"cosmic":13,"dbsnp":0}},"value":19},{"metric":{"unit":"","quality":"More is better","description":["Percentage of eval sites at comp sites"],"short_name":["Comp rate"],"name":"compRate"},"meta":{"all":{"1000genomes":0.84,"average":7.376666666666668,"cosmic":5.32,"dbsnp":15.97},"known":{"1000genomes":5.26,"average":38.59666666666667,"cosmic":10.53,"dbsnp":100.0},"novel":{"1000genomes":0.0,"average":1.4433333333333334,"cosmic":4.33,"dbsnp":0.0}},"value":5.32},{"metric":{"unit":"","quality":"More is better","description":["Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"],"short_name":["Concord"],"name":"nConcordant"},"meta":{"all":{"1000genomes":2,"average":17.333333333333332,"cosmic":14,"dbsnp":36},"known":{"1000genomes":2,"average":14.333333333333334,"cosmic":5,"dbsnp":36},"novel":{"1000genomes":0,"average":3.0,"cosmic":9,"dbsnp":0}},"value":14},{"metric":{"unit":"","quality":"More is better","description":["Concordance rate"],"short_name":["Conc rate"],"name":"concordantRate"},"meta":{"all":{"1000genomes":66.67,"average":67.83666666666666,"cosmic":73.68,"dbsnp":63.16},"known":{"1000genomes":66.67,"average":71.05333333333333,"cosmic":83.33,"dbsnp":63.16},"novel":{"1000genomes":0.0,"average":23.076666666666668,"cosmic":69.23,"dbsnp":0.0}},"value":73.68},{"metric":{"unit":"","quality":"More is better","description":["Variants per loci rate"],"short_name":["Var/loci"],"name":"variantRate"},"meta":{"all":{"1000genomes":1.1e-07,"average":1.1e-07,"cosmic":1.1e-07,"dbsnp":1.1e-07},"known":{"1000genomes":2e-08,"average":2.0000000000000004e-08,"cosmic":2e-08,"dbsnp":2e-08},"novel":{"1000genomes":1e-07,"average":1e-07,"cosmic":1e-07,"dbsnp":1e-07}},"value":1.1e-07},{"metric":{"unit":"","quality":"More is better","description":["Bases per variant rate"],"short_name":["Bp/var"],"name":"basesPerVariant"},"meta":{"all":{"1000genomes":8787566.0,"average":8787566.0,"cosmic":8787566.0,"dbsnp":8787566.0},"known":{"1000genomes":55037916.0,"average":55037916.0,"cosmic":55037916.0,"dbsnp":55037916.0},"novel":{"1000genomes":10457204.0,"average":10457204.0,"cosmic":10457204.0,"dbsnp":10457204.0}},"value":8787566.0},{"metric":{"unit":"","quality":"More is better","description":["Heterozygosity to homozygosity ratio"],"short_name":["Het/hom"],"name":"hetHomRatio"},"meta":{"all":{"1000genomes":53.29,"average":53.29,"cosmic":53.29,"dbsnp":53.29},"known":{"1000genomes":21.0,"average":21.0,"cosmic":21.0,"dbsnp":21.0},"novel":{"1000genomes":77.5,"average":77.5,"cosmic":77.5,"dbsnp":77.5}},"value":53.29},{"metric":{"unit":"","quality":"More is better","description":["Transition to transversion ratio"],"short_name":["Ti/Tv"],"name":"tiTvRatio"},"meta":{"all":{"1000genomes":0.29,"average":0.29,"cosmic":0.29,"dbsnp":0.29},"known":{"1000genomes":0.91,"average":0.91,"cosmic":0.91,"dbsnp":0.91},"novel":{"1000genomes":0.23,"average":0.23,"cosmic":0.23,"dbsnp":0.23}},"value":0.29}],"link":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-9/qc/varQC/Tissue-Sensitive-9-vardict.varQC.json","fpath":"/Users/vladsaveliev/vagrant/reporting_suite/test/NGS-99/final/Tissue-Sensitive-9/qc/varQC/Tissue-Sensitive-9-vardict.varQC.json"}],"date":"13 August 2014, Wednesday, 18:56:46"}
            </div>
        </div>
    </div>
</div>
</body>
</html>
