<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/table_sorter/style.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="html_aux/jquery.align-column.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>

    <script type="text/javascript" src="html_aux/dragtable.js"></script>
    <script type="text/javascript" src="html_aux/table_sorter/tsort.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });

            $('table.tableSorter').tableSort();
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for freebayes</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"report":{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"sample_reports":[{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-4","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-ready.bam","filtered_vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-mutect.anno.filt.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-freebayes.anno.filt.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-vardict.anno.filt.vcf"},"name":"Tissue-Resistant-4","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-4/PBMC-Resistant-4-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Resistant-4","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-4","vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-4/var"},"phenotype":"tumor","filtered_tsv_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-mutect.anno.filt.tsv","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-freebayes.anno.filt.tsv","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-vardict.anno.filt.tsv"},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/var/Tissue-Resistant-4-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/var/Tissue-Resistant-4-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/var/Tissue-Resistant-4-vardict.vcf"},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":7440,"dbsnp":7440,"cosmic":7440,"average":7440.0},"all":{"1000genomes":10440,"dbsnp":10440,"cosmic":10440,"average":10440.0},"novel":{"1000genomes":3000,"dbsnp":3000,"cosmic":3000,"average":3000.0}},"value":10440},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":7066,"dbsnp":7066,"cosmic":7066,"average":7066.0},"all":{"1000genomes":7693,"dbsnp":7693,"cosmic":7693,"average":7693.0},"novel":{"1000genomes":627,"dbsnp":627,"cosmic":627,"average":627.0}},"value":7693},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":4359,"dbsnp":7440,"cosmic":1226,"average":4341.666666666667},"all":{"1000genomes":4359,"dbsnp":7440,"cosmic":1263,"average":4354.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":37,"average":12.333333333333334}},"value":1263},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":58.59,"dbsnp":100.0,"cosmic":16.48,"average":58.35666666666666},"all":{"1000genomes":41.75,"dbsnp":71.26,"cosmic":12.1,"average":41.70333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":1.23,"average":0.41}},"value":12.1},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":4355,"dbsnp":7046,"cosmic":1029,"average":4143.333333333333},"all":{"1000genomes":4355,"dbsnp":7046,"cosmic":1047,"average":4149.333333333333},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":18,"average":6.0}},"value":1047},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":99.91,"dbsnp":94.7,"cosmic":83.93,"average":92.84666666666668},"all":{"1000genomes":99.91,"dbsnp":94.7,"cosmic":82.9,"average":92.50333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":48.65,"average":16.216666666666665}},"value":82.9},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.37e-06,"dbsnp":2.37e-06,"cosmic":2.37e-06,"average":2.37e-06},"all":{"1000genomes":3.33e-06,"dbsnp":3.33e-06,"cosmic":3.33e-06,"average":3.33e-06},"novel":{"1000genomes":9.6e-07,"dbsnp":9.6e-07,"cosmic":9.6e-07,"average":9.6e-07}},"value":3.33e-06},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":421661.0,"dbsnp":421661.0,"cosmic":421661.0,"average":421661.0},"all":{"1000genomes":300494.0,"dbsnp":300494.0,"cosmic":300494.0,"average":300494.0},"novel":{"1000genomes":1045720.0,"dbsnp":1045720.0,"cosmic":1045720.0,"average":1045720.0}},"value":300494.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.86,"dbsnp":1.86,"cosmic":1.86,"average":1.86},"all":{"1000genomes":1.8,"dbsnp":1.8,"cosmic":1.8,"average":1.8},"novel":{"1000genomes":1.65,"dbsnp":1.65,"cosmic":1.65,"average":1.6499999999999997}},"value":1.8},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.02,"dbsnp":2.02,"cosmic":2.02,"average":2.02},"all":{"1000genomes":1.73,"dbsnp":1.73,"cosmic":1.73,"average":1.7299999999999998},"novel":{"1000genomes":0.3,"dbsnp":0.3,"cosmic":0.3,"average":0.3}},"value":1.73}],"plots":[],"html_fpath":"../../../Tissue-Resistant-4/qc/varQC_postVarFilter/Tissue-Resistant-4-mutect.varQC_postVarFilter.html"},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Sensitive-7","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-ready.bam","filtered_vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-mutect.anno.filt.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-freebayes.anno.filt.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-vardict.anno.filt.vcf"},"name":"Tissue-Sensitive-7","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-7/PBMC-Sensitive-7-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Sensitive-7","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-7","vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-7/var"},"phenotype":"tumor","filtered_tsv_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-mutect.anno.filt.tsv","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-freebayes.anno.filt.tsv","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/Tissue-Sensitive-7-vardict.anno.filt.tsv"},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/var/Tissue-Sensitive-7-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/var/Tissue-Sensitive-7-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/var/Tissue-Sensitive-7-vardict.vcf"},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-7/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0}],"plots":[],"html_fpath":"../../../Tissue-Sensitive-7/qc/varQC_postVarFilter/Tissue-Sensitive-7-mutect.varQC_postVarFilter.html"},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Sensitive-8","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-ready.bam","filtered_vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-mutect.anno.filt.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-freebayes.anno.filt.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-vardict.anno.filt.vcf"},"name":"Tissue-Sensitive-8","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-8/PBMC-Sensitive-8-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Sensitive-8","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-8","vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Sensitive-8/var"},"phenotype":"tumor","filtered_tsv_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-mutect.anno.filt.tsv","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-freebayes.anno.filt.tsv","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/Tissue-Sensitive-8-vardict.anno.filt.tsv"},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/var/Tissue-Sensitive-8-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/var/Tissue-Sensitive-8-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/var/Tissue-Sensitive-8-vardict.vcf"},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-8/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"all":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":0.0,"average":0.0}},"value":0.0}],"plots":[],"html_fpath":"../../../Tissue-Sensitive-8/qc/varQC_postVarFilter/Tissue-Sensitive-8-mutect.varQC_postVarFilter.html"},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-1","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-ready.bam","filtered_vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-mutect.anno.filt.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-freebayes.anno.filt.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-vardict.anno.filt.vcf"},"name":"Tissue-Resistant-1","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-1/PBMC-Resistant-1-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Resistant-1","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-1","vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-1/var"},"phenotype":"tumor","filtered_tsv_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-mutect.anno.filt.tsv","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-freebayes.anno.filt.tsv","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/Tissue-Resistant-1-vardict.anno.filt.tsv"},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/var/Tissue-Resistant-1-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/var/Tissue-Resistant-1-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/var/Tissue-Resistant-1-vardict.vcf"},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-1/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":6972,"dbsnp":6972,"cosmic":6972,"average":6972.0},"all":{"1000genomes":9888,"dbsnp":9888,"cosmic":9888,"average":9888.0},"novel":{"1000genomes":2916,"dbsnp":2916,"cosmic":2916,"average":2916.0}},"value":9888},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":6629,"dbsnp":6629,"cosmic":6629,"average":6629.0},"all":{"1000genomes":7172,"dbsnp":7172,"cosmic":7172,"average":7172.0},"novel":{"1000genomes":543,"dbsnp":543,"cosmic":543,"average":543.0}},"value":7172},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":4006,"dbsnp":6972,"cosmic":1126,"average":4034.6666666666665},"all":{"1000genomes":4006,"dbsnp":6972,"cosmic":1170,"average":4049.3333333333335},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":44,"average":14.666666666666666}},"value":1170},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":57.46,"dbsnp":100.0,"cosmic":16.15,"average":57.870000000000005},"all":{"1000genomes":40.51,"dbsnp":70.51,"cosmic":11.83,"average":40.95},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":1.51,"average":0.5033333333333333}},"value":11.83},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":4004,"dbsnp":6612,"cosmic":954,"average":3856.6666666666665},"all":{"1000genomes":4004,"dbsnp":6612,"cosmic":971,"average":3862.3333333333335},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":17,"average":5.666666666666667}},"value":971},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":99.95,"dbsnp":94.84,"cosmic":84.72,"average":93.17},"all":{"1000genomes":99.95,"dbsnp":94.84,"cosmic":82.99,"average":92.59333333333335},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":38.64,"average":12.88}},"value":82.99},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.22e-06,"dbsnp":2.22e-06,"cosmic":2.22e-06,"average":2.22e-06},"all":{"1000genomes":3.15e-06,"dbsnp":3.15e-06,"cosmic":3.15e-06,"average":3.15e-06},"novel":{"1000genomes":9.3e-07,"dbsnp":9.3e-07,"cosmic":9.3e-07,"average":9.3e-07}},"value":3.15e-06},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":449965.0,"dbsnp":449965.0,"cosmic":449965.0,"average":449965.0},"all":{"1000genomes":317269.0,"dbsnp":317269.0,"cosmic":317269.0,"average":317269.0},"novel":{"1000genomes":1075844.0,"dbsnp":1075844.0,"cosmic":1075844.0,"average":1075844.0}},"value":317269.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.59,"dbsnp":1.59,"cosmic":1.59,"average":1.59},"all":{"1000genomes":1.52,"dbsnp":1.52,"cosmic":1.52,"average":1.5200000000000002},"novel":{"1000genomes":1.37,"dbsnp":1.37,"cosmic":1.37,"average":1.37}},"value":1.52},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.02,"dbsnp":2.02,"cosmic":2.02,"average":2.02},"all":{"1000genomes":1.68,"dbsnp":1.68,"cosmic":1.68,"average":1.68},"novel":{"1000genomes":0.14,"dbsnp":0.14,"cosmic":0.14,"average":0.14}},"value":1.68}],"plots":[],"html_fpath":"../../../Tissue-Resistant-1/qc/varQC_postVarFilter/Tissue-Resistant-1-freebayes.varQC_postVarFilter.html"},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-4","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-ready.bam","filtered_vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-mutect.anno.filt.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-freebayes.anno.filt.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-vardict.anno.filt.vcf"},"name":"Tissue-Resistant-4","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-4/PBMC-Resistant-4-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Resistant-4","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-4","vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-4/var"},"phenotype":"tumor","filtered_tsv_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-mutect.anno.filt.tsv","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-freebayes.anno.filt.tsv","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/Tissue-Resistant-4-vardict.anno.filt.tsv"},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/var/Tissue-Resistant-4-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/var/Tissue-Resistant-4-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/var/Tissue-Resistant-4-vardict.vcf"},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-4/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":7440,"dbsnp":7440,"cosmic":7440,"average":7440.0},"all":{"1000genomes":10440,"dbsnp":10440,"cosmic":10440,"average":10440.0},"novel":{"1000genomes":3000,"dbsnp":3000,"cosmic":3000,"average":3000.0}},"value":10440},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":7066,"dbsnp":7066,"cosmic":7066,"average":7066.0},"all":{"1000genomes":7693,"dbsnp":7693,"cosmic":7693,"average":7693.0},"novel":{"1000genomes":627,"dbsnp":627,"cosmic":627,"average":627.0}},"value":7693},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":4359,"dbsnp":7440,"cosmic":1226,"average":4341.666666666667},"all":{"1000genomes":4359,"dbsnp":7440,"cosmic":1263,"average":4354.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":37,"average":12.333333333333334}},"value":1263},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":58.59,"dbsnp":100.0,"cosmic":16.48,"average":58.35666666666666},"all":{"1000genomes":41.75,"dbsnp":71.26,"cosmic":12.1,"average":41.70333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":1.23,"average":0.41}},"value":12.1},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":4355,"dbsnp":7046,"cosmic":1029,"average":4143.333333333333},"all":{"1000genomes":4355,"dbsnp":7046,"cosmic":1047,"average":4149.333333333333},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":18,"average":6.0}},"value":1047},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":99.91,"dbsnp":94.7,"cosmic":83.93,"average":92.84666666666668},"all":{"1000genomes":99.91,"dbsnp":94.7,"cosmic":82.9,"average":92.50333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":48.65,"average":16.216666666666665}},"value":82.9},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.37e-06,"dbsnp":2.37e-06,"cosmic":2.37e-06,"average":2.37e-06},"all":{"1000genomes":3.33e-06,"dbsnp":3.33e-06,"cosmic":3.33e-06,"average":3.33e-06},"novel":{"1000genomes":9.6e-07,"dbsnp":9.6e-07,"cosmic":9.6e-07,"average":9.6e-07}},"value":3.33e-06},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":421661.0,"dbsnp":421661.0,"cosmic":421661.0,"average":421661.0},"all":{"1000genomes":300494.0,"dbsnp":300494.0,"cosmic":300494.0,"average":300494.0},"novel":{"1000genomes":1045720.0,"dbsnp":1045720.0,"cosmic":1045720.0,"average":1045720.0}},"value":300494.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.86,"dbsnp":1.86,"cosmic":1.86,"average":1.86},"all":{"1000genomes":1.8,"dbsnp":1.8,"cosmic":1.8,"average":1.8},"novel":{"1000genomes":1.65,"dbsnp":1.65,"cosmic":1.65,"average":1.6499999999999997}},"value":1.8},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.02,"dbsnp":2.02,"cosmic":2.02,"average":2.02},"all":{"1000genomes":1.73,"dbsnp":1.73,"cosmic":1.73,"average":1.7299999999999998},"novel":{"1000genomes":0.3,"dbsnp":0.3,"cosmic":0.3,"average":0.3}},"value":1.73}],"plots":[],"html_fpath":"../../../Tissue-Resistant-4/qc/varQC_postVarFilter/Tissue-Resistant-4-freebayes.varQC_postVarFilter.html"},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-5","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-ready.bam","filtered_vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-mutect.anno.filt.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-freebayes.anno.filt.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-vardict.anno.filt.vcf"},"name":"Tissue-Resistant-5","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-5/PBMC-Resistant-5-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Resistant-5","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-5","vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-5/var"},"phenotype":"tumor","filtered_tsv_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-mutect.anno.filt.tsv","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-freebayes.anno.filt.tsv","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/Tissue-Resistant-5-vardict.anno.filt.tsv"},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/var/Tissue-Resistant-5-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/var/Tissue-Resistant-5-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/var/Tissue-Resistant-5-vardict.vcf"},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-5/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":7972,"dbsnp":7972,"cosmic":7972,"average":7972.0},"all":{"1000genomes":9675,"dbsnp":9675,"cosmic":9675,"average":9675.0},"novel":{"1000genomes":1703,"dbsnp":1703,"cosmic":1703,"average":1703.0}},"value":9675},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":7610,"dbsnp":7610,"cosmic":7610,"average":7610.0},"all":{"1000genomes":7781,"dbsnp":7781,"cosmic":7781,"average":7781.0},"novel":{"1000genomes":171,"dbsnp":171,"cosmic":171,"average":171.0}},"value":7781},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":4771,"dbsnp":7972,"cosmic":1278,"average":4673.666666666667},"all":{"1000genomes":4771,"dbsnp":7972,"cosmic":1300,"average":4681.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":22,"average":7.333333333333333}},"value":1300},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":59.85,"dbsnp":100.0,"cosmic":16.03,"average":58.626666666666665},"all":{"1000genomes":49.31,"dbsnp":82.4,"cosmic":13.44,"average":48.38333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":1.29,"average":0.43}},"value":13.44},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":4769,"dbsnp":7606,"cosmic":1080,"average":4485.0},"all":{"1000genomes":4769,"dbsnp":7606,"cosmic":1091,"average":4488.666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":11,"average":3.6666666666666665}},"value":1091},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":99.96,"dbsnp":95.41,"cosmic":84.51,"average":93.29333333333334},"all":{"1000genomes":99.96,"dbsnp":95.41,"cosmic":83.92,"average":93.09666666666668},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":50.0,"average":16.666666666666668}},"value":83.92},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.54e-06,"dbsnp":2.54e-06,"cosmic":2.54e-06,"average":2.54e-06},"all":{"1000genomes":3.08e-06,"dbsnp":3.08e-06,"cosmic":3.08e-06,"average":3.0799999999999997e-06},"novel":{"1000genomes":5.4e-07,"dbsnp":5.4e-07,"cosmic":5.4e-07,"average":5.4e-07}},"value":3.08e-06},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":393522.0,"dbsnp":393522.0,"cosmic":393522.0,"average":393522.0},"all":{"1000genomes":324254.0,"dbsnp":324254.0,"cosmic":324254.0,"average":324254.0},"novel":{"1000genomes":1842138.0,"dbsnp":1842138.0,"cosmic":1842138.0,"average":1842138.0}},"value":324254.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.6,"dbsnp":1.6,"cosmic":1.6,"average":1.6000000000000003},"all":{"1000genomes":1.56,"dbsnp":1.56,"cosmic":1.56,"average":1.5599999999999998},"novel":{"1000genomes":1.4,"dbsnp":1.4,"cosmic":1.4,"average":1.3999999999999997}},"value":1.56},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.06,"dbsnp":2.06,"cosmic":2.06,"average":2.06},"all":{"1000genomes":2.01,"dbsnp":2.01,"cosmic":2.01,"average":2.01},"novel":{"1000genomes":0.71,"dbsnp":0.71,"cosmic":0.71,"average":0.71}},"value":2.01}],"plots":[],"html_fpath":"../../../Tissue-Resistant-5/qc/varQC_postVarFilter/Tissue-Resistant-5-freebayes.varQC_postVarFilter.html"},{"metric_storage":{"common_for_all_samples_section":{"metrics":[],"name":"common_for_all_samples_section","metrics_by_name":{},"title":""},"sections_by_name":{"basic":{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}},"sections":[{"metrics":[{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"}],"name":"basic","metrics_by_name":{"nSNPs":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"nDeletions":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"compRate":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"basesPerVariant":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"nEvalVariants":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"nConcordant":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"hetHomRatio":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"concordantRate":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"nVariantsAtComp":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"variantRate":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"tiTvRatio":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"nInsertions":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"}},"title":""}]},"display_name":"Tissue-Resistant-6","json_fpath":null,"report_name":"","sample":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-ready.bam","filtered_vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-mutect.anno.filt.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-freebayes.anno.filt.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-vardict.anno.filt.vcf"},"name":"Tissue-Resistant-6","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":{"bam":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-6/PBMC-Resistant-6-ready.bam","filtered_vcf_by_callername":{},"name":"PBMC-Resistant-6","filtered_maf_by_callername":{},"bed":"/ngs/reference_data/genomes/Hsapiens/hg19/bed/AZ_Exome.bed","normal_match":null,"phenotype":"normal","filtered_tsv_by_callername":{},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-6","vcf_by_callername":{},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/PBMC-Resistant-6/var"},"phenotype":"tumor","filtered_tsv_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-mutect.anno.filt.tsv","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-freebayes.anno.filt.tsv","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/Tissue-Resistant-6-vardict.anno.filt.tsv"},"dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6","vcf_by_callername":{"mutect":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/var/Tissue-Resistant-6-mutect.vcf","freebayes":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/var/Tissue-Resistant-6-freebayes.vcf","vardict":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/var/Tissue-Resistant-6-vardict.vcf"},"var_dirpath":"/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Resistant-6/var"},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":7316,"dbsnp":7316,"cosmic":7316,"average":7316.0},"all":{"1000genomes":9293,"dbsnp":9293,"cosmic":9293,"average":9293.0},"novel":{"1000genomes":1977,"dbsnp":1977,"cosmic":1977,"average":1977.0}},"value":9293},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":6958,"dbsnp":6958,"cosmic":6958,"average":6958.0},"all":{"1000genomes":7130,"dbsnp":7130,"cosmic":7130,"average":7130.0},"novel":{"1000genomes":172,"dbsnp":172,"cosmic":172,"average":172.0}},"value":7130},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":4318,"dbsnp":7316,"cosmic":1172,"average":4268.666666666667},"all":{"1000genomes":4318,"dbsnp":7316,"cosmic":1196,"average":4276.666666666667},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":24,"average":8.0}},"value":1196},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":59.02,"dbsnp":100.0,"cosmic":16.02,"average":58.34666666666667},"all":{"1000genomes":46.47,"dbsnp":78.73,"cosmic":12.87,"average":46.02333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":1.21,"average":0.4033333333333333}},"value":12.87},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":4312,"dbsnp":6955,"cosmic":995,"average":4087.3333333333335},"all":{"1000genomes":4312,"dbsnp":6955,"cosmic":1000,"average":4089.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":5,"average":1.6666666666666667}},"value":1000},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":99.86,"dbsnp":95.07,"cosmic":84.9,"average":93.27666666666669},"all":{"1000genomes":99.86,"dbsnp":95.07,"cosmic":83.61,"average":92.84666666666668},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":20.83,"average":6.9433333333333325}},"value":83.61},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":2.33e-06,"dbsnp":2.33e-06,"cosmic":2.33e-06,"average":2.33e-06},"all":{"1000genomes":2.96e-06,"dbsnp":2.96e-06,"cosmic":2.96e-06,"average":2.96e-06},"novel":{"1000genomes":6.3e-07,"dbsnp":6.3e-07,"cosmic":6.3e-07,"average":6.3e-07}},"value":2.96e-06},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":428808.0,"dbsnp":428808.0,"cosmic":428808.0,"average":428808.0},"all":{"1000genomes":337583.0,"dbsnp":337583.0,"cosmic":337583.0,"average":337583.0},"novel":{"1000genomes":1586829.0,"dbsnp":1586829.0,"cosmic":1586829.0,"average":1586829.0}},"value":337583.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":1.25,"dbsnp":1.25,"cosmic":1.25,"average":1.25},"all":{"1000genomes":1.2,"dbsnp":1.2,"cosmic":1.2,"average":1.2},"novel":{"1000genomes":1.03,"dbsnp":1.03,"cosmic":1.03,"average":1.03}},"value":1.2},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":2.03,"dbsnp":2.03,"cosmic":2.03,"average":2.03},"all":{"1000genomes":1.98,"dbsnp":1.98,"cosmic":1.98,"average":1.9799999999999998},"novel":{"1000genomes":0.72,"dbsnp":0.72,"cosmic":0.72,"average":0.7200000000000001}},"value":1.98}],"plots":[],"html_fpath":"../../../Tissue-Resistant-6/qc/varQC_postVarFilter/Tissue-Resistant-6-freebayes.varQC_postVarFilter.html"}],"name":""},"data_outside_reports":{},"date":"08 September 2014, Monday, 07:56:31"}
            </div>
        </div>
    </div>
</div>
</body>
</html>
