<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="html_aux/jquery.align-column.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>
    <script type="text/javascript" src="html_aux/dragtable.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for Tissue-Sensitive-3 (caller: vardict)</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"date":"22 August 2014, Friday, 08:38:45","common_records":[],"reports":[{"sample_reports":[{"sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"Tissue-Sensitive-3","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":null,"filtered_tsv_by_callername":{},"dirpath":null,"vcf_by_callername":{},"filtered_clean_vcf_by_callername":{},"var_dirpath":null},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":34,"dbsnp":34,"cosmic":34,"average":34.0},"all":{"1000genomes":95,"dbsnp":95,"cosmic":95,"average":95.0},"novel":{"1000genomes":61,"dbsnp":61,"cosmic":61,"average":61.0}},"value":95},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":22,"dbsnp":22,"cosmic":22,"average":22.0},"all":{"1000genomes":73,"dbsnp":73,"cosmic":73,"average":73.0},"novel":{"1000genomes":51,"dbsnp":51,"cosmic":51,"average":51.0}},"value":73},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":3,"dbsnp":3,"cosmic":3,"average":3.0},"all":{"1000genomes":6,"dbsnp":6,"cosmic":6,"average":6.0},"novel":{"1000genomes":3,"dbsnp":3,"cosmic":3,"average":3.0}},"value":6},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":9,"dbsnp":9,"cosmic":9,"average":9.0},"all":{"1000genomes":16,"dbsnp":16,"cosmic":16,"average":16.0},"novel":{"1000genomes":7,"dbsnp":7,"cosmic":7,"average":7.0}},"value":16},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":3,"dbsnp":34,"cosmic":5,"average":14.0},"all":{"1000genomes":3,"dbsnp":34,"cosmic":8,"average":15.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":3,"average":1.0}},"value":8},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":8.82,"dbsnp":100.0,"cosmic":14.71,"average":41.17666666666667},"all":{"1000genomes":3.16,"dbsnp":35.79,"cosmic":8.42,"average":15.790000000000001},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":4.92,"average":1.64}},"value":8.42},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":2,"dbsnp":29,"cosmic":4,"average":11.666666666666666},"all":{"1000genomes":2,"dbsnp":29,"cosmic":6,"average":12.333333333333334},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":2,"average":0.6666666666666666}},"value":6},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":66.67,"dbsnp":85.29,"cosmic":80.0,"average":77.32000000000001},"all":{"1000genomes":66.67,"dbsnp":85.29,"cosmic":75.0,"average":75.65333333333334},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":66.67,"average":22.223333333333333}},"value":75.0},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1e-08,"dbsnp":1e-08,"cosmic":1e-08,"average":1.0000000000000002e-08},"all":{"1000genomes":3e-08,"dbsnp":3e-08,"cosmic":3e-08,"average":3e-08},"novel":{"1000genomes":2e-08,"dbsnp":2e-08,"cosmic":2e-08,"average":2.0000000000000004e-08}},"value":3e-08},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":92269448.0,"dbsnp":92269448.0,"cosmic":92269448.0,"average":92269448.0},"all":{"1000genomes":33022750.0,"dbsnp":33022750.0,"cosmic":33022750.0,"average":33022750.0},"novel":{"1000genomes":51428873.0,"dbsnp":51428873.0,"cosmic":51428873.0,"average":51428873.0}},"value":33022750.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":6.6,"dbsnp":6.6,"cosmic":6.6,"average":6.599999999999999},"all":{"1000genomes":18.17,"dbsnp":18.17,"cosmic":18.17,"average":18.17},"novel":{"1000genomes":76.0,"dbsnp":76.0,"cosmic":76.0,"average":76.0}},"value":18.17},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":1.75,"dbsnp":1.75,"cosmic":1.75,"average":1.75},"all":{"1000genomes":0.35,"dbsnp":0.35,"cosmic":0.35,"average":0.3499999999999999},"novel":{"1000genomes":0.11,"dbsnp":0.11,"cosmic":0.11,"average":0.11}},"value":0.35}],"name":"","fpath":"","plots":["/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-3/qc/varQC/Tissue-Sensitive-3.substitutions.png","/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-3/qc/varQC/Tissue-Sensitive-3.indels.png"]}],"name":""}]}
            </div>
        </div>
    </div>
</div>
</body>
</html>
