<!--<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">-->
<!doctype html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
    <meta http-equiv="x-ua-compatible" content="IE=Edge"/>
    <meta http-equiv="Content-Type" content="text/html;charset=UTF-8">

    <link rel="stylesheet" type="text/css" href="bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="common.css" />
    <link rel="stylesheet" type="text/css" href="report.css" />
    <link rel="stylesheet" type="text/css" href="table_sorter/style.css" />

    <link rel="stylesheet" type="text/css" href="clinical_report.css" />
    <link rel="stylesheet" type="text/css" href="header_picture.css" />

    <!--[if IE]>
        <style>
            table.report_table {
                table-layout: fixed;
                width: 100%;
            }
            table.report_table td {
            }
        </style>
    <![endif]-->

    <script type="text/javascript" src="jquery-1.12.0.min.js"></script>
    <script type="text/javascript" src="flot/jquery.flot.js"></script>
    <script type="text/javascript" src="bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="scripts/utils.js"></script>
    <script type="text/javascript" src="table_sorter/tsort.js"></script>
    <script type="text/javascript" src="clinical_report.js"></script>
    <script type="text/javascript" src="draw_mutations_plot.js"></script>
    <script type="text/javascript" src="draw_substitutions_plot.js"></script>
    <script type="text/javascript" src="draw_genes_coverage_plot.js"></script>
    <script type="text/javascript" src="draw_seq2c_plot.js"></script>
    <script type="text/javascript" src="easyzoom.js"></script>
</head>

<body>

<div id="clinical_report_header_container">
    <!--<div style="width: 100%; height: 30px; padding: 5px;">Internet Explorer is not supported </div>-->
    <div id="clinical_report_header_image"></div>
</div>

<div class="content">
    <div style="margin-top: -150px; margin-bottom: 45px; color: white;">
        <h1 style="">Oncology NGS Report</h1>
        <span style="color: red;">For research use only</span>
    </div>

    <div id='report' class="clinical_report">
        <div style="width: 1127px">
            <div style="">
            {.section sample}
                <h1 class="report-header">Sample: {sample}</h1>
                <p style="line-height: 1.5em;">
                <b>Project: </b><a href="{project_report_rel_path}">{project_name}</a>, {genome_build}<br>
                {.section patient}<b>Patient: </b>{sex}<br>{.end}
                <!--<b>Study</b>: 9291 First line progressor. Progressed Feb 6 2015 after 13 months on 9291: L858R and T790M+ prior to this 1st line therapy<br>-->
                {.section sample_type}<b>Sample type: </b>{sample_type}<br>{.or}{.end}
                {.section target_section}
                    <!--{.section targqc_link}<b>QC metrics: </b><a href="{targqc_link}">TargQC report</a><br>{.or}{.end}-->
                    <b>Target: </b>{.section bed_path}<a href="{bed_path}">{panel}</a>{.or}{panel}{.end}
                    {.section target_fraction}covered {target_fraction} of all positions{.or}{.end}
                    {.section ave_depth}at {ave_depth} average coverage{.or}{.end}
                    <br>
                {.end}
                </p>

                <!--<div class="row placeholders">-->
                    <!--<div class="col-xs-6 col-sm-3 placeholder">-->
                        <!--<img data-src="holder.js/200x200/auto/sky" class="img-responsive" alt="Generic placeholder thumbnail">-->
                        <!--<h4 class="section_header">Gender</h4>-->
                        <!--{.section basecalls}-->
                        <!--<span class="text-muted">{contents}<br></span>-->
                        <!--{.end}-->
                        <!--{.section raw_fastqc}-->
                        <!--<span class="text-muted">{contents}<br></span>-->
                        <!--{.end}-->
                        <!--{.section fastqc}-->
                        <!--<span class="text-muted">{contents}</span>-->
                        <!--{.or}-->
                        <!--&lt;!&ndash;<span class="text-muted not-available">FastQC reports are not available</span>&ndash;&gt;-->
                        <!--{.end}-->
                    <!--</div>-->
                <!--</div>-->
                <!--<div class="space_30px"></div>-->
            {.end}
            </div>

            <!--<div style="float: left; margin-left: 10px;">-->
            <!--{.section approach}-->
                <!--<h1 class="report-header">Approach</h1>-->
                <!--<p style="max-width: 600px; line-height: 1.5em;">Sequencing of this individual's-->
                    <!--{target_type} was performed and covered {target_fraction}-->
                    <!--of all positions at {ave_depth} average coverage,-->
                    <!--resulting in {total_variants} variants compared to a reference genome.-->
                    <!--These data were analyzed to identify previously reported and novel variants-->
                    <!--in {total_key_genes} genes that have been previously implicated-->
                    <!--in various cancers (see Supplement for a list of genes and coverage information).-->
                    <!--&lt;!&ndash;All genes relevant to the indication had adequate coverage (>95% at 8X).&ndash;&gt;-->
                    <!--&lt;!&ndash;Please note that the presence of pathogenic variants in genes not analyzed&ndash;&gt;-->
                    <!--&lt;!&ndash;or with incomplete coverage cannot be fully excluded.&ndash;&gt;-->
                <!--</p>-->
            <!--{.end}-->
            <!--</div>-->
            <!--<div style="clear: both"></div>-->
        </div>

        <div class="space_8px"></div>
        <div class="space_8px"></div>
        <div>
            <a class="dotted-link" href="#variants">Mutations</a><span>&nbsp;&nbsp;&nbsp;</span>
            <a class="dotted-link" href="#actionable">Alterations in actionable genes</a><span>&nbsp;&nbsp;&nbsp;</span>
            {.section sv}<a class="dotted-link" href="#sv">Structural variants</a><span>&nbsp;&nbsp;&nbsp;</span>{.or}{.end}
            {.section circos}<a class="dotted-link" href="#circos">Circos plot</a><span>&nbsp;&nbsp;&nbsp;</span>{.or}{.end}
            <a class="dotted-link" href="#seq2c">Seq2C log2 ratios (CNV)</a>                  <span>&nbsp;&nbsp;&nbsp;</span>
            <a class="dotted-link" href="#coverage">Coverage analysis</a>        <span>&nbsp;&nbsp;&nbsp;</span>
        </div>
        <div class="space_4px"></div>

        <a name="variants"></a>
        <h1>Mutations in {total_key_genes} {key_or_target} genes</h1>
        {.section variants}
            <div id="variants_div">
                <p style="margin-top: -5px;">{total_variants} variants found compared to a reference genome using <a href="https://github.com/AstraZeneca-NGS/VarDict">VarDict variant caller</a>.
                    These data were analyzed to identify previously reported and novel variants
                    in {total_key_genes} {genes_description} (see <a class="dotted-link" href="#coverage">the list of genes and coverage information</a>).
                </p>
                <p>
                    Warning! There are many caveats to interpreting NGS results, and many sources of false positive and false negative results. Some factors and issues to be aware of:
                    <ol>
                    <li><b>Germline variants</b> – germline SNPs occur at approximately 100%, 50%, or 0% frequency and every effort is made to filter out germline variants (some exceptions like BRCA1/2).
                    <li><b>Somatic variants</b> – somatic variants occur at nearly any allele frequency but as allele frequencies approach 20% and lower, data become noisier and false positive variants increase.
                    <li><b>Sample types</b> - in general, focus on somatic variants above ~30% allele frequency for cell line and explant models, above ~5% and above for tumors, and above ~0.7% for ctDNA. Formalin treated tumors have the added complication of C>T and G>A de-amination artefacts.
                    <li><b>False positive mutations</b> – advance algorithms were employed to generate the highest quality variant calls, but false positive will still occur and are likely present in the candidate mutation output below. False positive variants are more found in homopolymer tracts, in genes with repetitive regions (within the gene or elsewhere in the genome), in very large genes, and in genes residing in repetitive or unstable regions of the genome.
                    </ol>
                </p>
                <div id="variants_table_controls">
                    <div style="float: left">
                        <p id="variants_switch">
                            <span>known, likely</span>
                            /
                            <a class="dotted-link" id="extend_link_variants" onclick="extendClick($(this))">
                                + unknown</a>
                            /
                            <a class="dotted-link" id="extend_link_crapome_variants" onclick="extendClick($(this))">
                                crapome</a>
                        </p>
                    </div>
                        <div style="float: left; margin-top: 10px; padding-left: 110px;">
                        Show only with freq ≥
                        <input type="range" id="mut_af_slider" min="0" max="100" value="{min_af}" style="vertical-align: middle;"
                               onchange="filterMutationsByAF(this.value)" oninput="mut_af_textbox.value=this.value" />
                        <input type="text" id="mut_af_textbox" min="0" max="100" value="{min_af}" maxlength="3" size="3"
                               onchange="filterMutationsByAF(this.value)" oninput="mut_af_slider.value=this.value" style="text-align: right"/>&nbsp%
                    </div>
                    <div style="float: right; margin-top: -20px; margin-bottom: 0;">
                        <p>
                            <a class="btn" id="download_mut_table" style="display: none;">Download for Excel</a>
                        </p>
                    </div>
                    <div style="clear: both"></div>
                </div>
                <div id="variants_table_div">
                    {table}
                </div>
            </div>
            <div class="js_plot" id="mut_plot" style="position: relative">
                <table>
                    <tr>
                        <td colspan="2">
                            <div class='plot_placeholder mut_plot_placeholder' id="mut_plot_placeholder">
                            </div>
                        </td>
                    </tr>
                    <tr>
                        <td>
                            <div class='legend_placeholder' id="mut_plot_legend_placeholder"></div>
                        </td>
                    </tr>
                </table>
                <div class="js_plot" id="substitutions_plot">
                    <table>
                        <tr>
                            <td colspan="2">
                                <div class='plot_placeholder substitutions_plot_placeholder' id="substitutions_plot_placeholder">
                                </div>
                            </td>
                        </tr>
                        <tr>
                            <td>
                                <div class='legend_placeholder' id="substitutions_plot_legend_placeholder"></div>
                            </td>
                        </tr>
                    </table>
                    <div class="json_code">
                        <div id="substitutions_plot_data_json">
                            {substitutions_plot_data}
                        </div>
                    </div>
                </div>
            </div>
            <div class="json_code">
                <div id="mut_plot_data_json">
                    {plot_data}
                </div>
            </div>
        {.or}
            <p>No variants detected in {total_key_genes} {key_or_target} regions.</p>
        {.end}

        <a name="actionable"></a>
        <h1>Actionable {total_key_genes} {key_or_target} genes - variants and larger alterations of interest</h1>
        {.section actionable_genes}
            <p style="margin-top: -5px;">Mutations annotated based on the Broad's database of actionable genes</p>
            <div class="actionable_table">
                {table}
            </div>
        {.or}
            <p>No mutations found in {total_key_genes} actionable genes.</p>
        {.end}

        {.section sv}
            <a name="sv"></a>
            <h1>Structural variants in {total_key_genes} {key_or_target} genes</h1>
            <p>Reported only and whole exon deletions and known fusions. See <a href="{sv_link}">the full list here</a>.</p>
            {.section report}
                <div id="sv_div">
                    <div id="sv_table_div">
                        {table}
                    </div>
                </div>
            {.or}
                <p>No whole exon deletion or known fusions detected in {total_key_genes} {key_or_target} regions.</p>
            {.end}
        {.or}
        {.end}

        {.section circos}
        <a name="seq2c"></a>
            <a name="circos"></a>
            <h1>Circos plot</h1>
            <p>Circos plot displays the following genomic data: CNVs, SNPs and structural variants.
                The outer ring indicates the chromosome ideograms and distances in Mb.
                The next ring indicates the copy number information. Color scheme correspond to Seq2C plot colors.
                The innermost ring represents the allele frequency of the SNPs (from 0 to 100%).
                Colored links represent chromosomal rearrangements.
                <a href="{circos_img}" target="_blank">Click</a> to open the image in a new tab.</p>
            <div id="circos_plot_div">
                <a href="{circos_img}" class="zoom" id="circos_zoom">
                    <img src="circos" class="circos_image">
                </a>
            </div>
        {.or}
        {.end}

        <div class="space_10px"></div>
        <a name="seq2c"></a>
        <h1>Seq2C log2 ratios (CNV)</h1>
        {.section seq2c}
            <p style="margin-top: -5px;">Gene coverage log ratios according to the Seq2C CNV caller.
                <span style="color: red;">Red</span> means amplification in a {key_or_target} gene,
                <span style="color: blue;">blue</span> means deletion in a {key_or_target} gene,
                <b>black</b> means no event.
                {.section description_for_whole_genomic_profile}
                <br><br>For non-{key_or_target} genes:
                <span style="color: #ff8080;">pink</span> means amplification,
                <span style="color: #67ddff">light blue</span> means deletion,
                <span style="color: gray;">gray</span> means no event.
                {.end}</p>
            <!--<div class="seq2c_plot" style="margin-left: -20px;">-->
                <!--<img src="seq2c_plot" width="1241px" alt="Seq2C log2 ratios"/>-->
            <!--</div>-->
            <div class="js_plot" id="seq2c_plot">
                <table>
                    <tr>
                        <td colspan="2">
                            <div class='plot_placeholder seq2c_plot_placeholder' id="seq2c_plot_placeholder">
                            </div>
                        </td>
                        <td>
                            <div class='legend_placeholder' id="seq2c_plot_legend_placeholder"></div>
                        </td>
                    </tr>
                </table>
                <div class="json_code">
                    <div id="seq2c_plot_data_json">
                        {plot_data}
                    </div>
                </div>
            </div>
            {.section amp_del}
                {.section seq2c_switch}
                <p id="seq2c_switch">
                    <span>{key_or_target} genes</span>
                    /
                    <a class="dotted-link" id="extend_link_seq2c" onclick="extendClick($(this))">
                        all genes</a>
                </p>
                {.or}
                {.end}
                <div id="seq2c_table_div">
                {.section short_table}
                    <table width="100%" cellspacing="0" class="table_short">
                        <tr style="vertical-align: top;">
                        <!--<div class="multi-col-layout">-->
                        {.repeated section columns}
                            <td style="">
                                <table class="report_table static_report"
                                       style="margin-right: 40px;">
                                    <thead>
                                        <tr class="top_row_tr">
                                            {.repeated section metric_names}
                                                {@}
                                            {.end}
                                        </tr>
                                    </thead>
                                    <tbody>
                                        {.repeated section rows}
                                            <tr>
                                                {.repeated section records}
                                                    {@}
                                                {.end}
                                            </tr>
                                        {.end}
                                    </tbody>
                                </table>
                            </td>
                        {.end}
                        </tr>
                    </table>
                    {.or}
                    <table width="100%" cellspacing="0" class="table_short">
                    </table>
                    {.end}
                    {.section full_table}
                    <table width="100%" cellspacing="0" class="table_full" style="display:none">
                        <tr style="vertical-align: top;">
                        <!--<div class="multi-col-layout">-->
                        {.repeated section columns}
                            <td style="">
                                <table class="report_table static_report"
                                       style="margin-right: 40px;">
                                    <thead>
                                        <tr class="top_row_tr">
                                            {.repeated section metric_names}
                                                {@}
                                            {.end}
                                        </tr>
                                    </thead>
                                    <tbody>
                                        {.repeated section rows}
                                            <tr>
                                                {.repeated section records}
                                                    {@}
                                                {.end}
                                            </tr>
                                        {.end}
                                    </tbody>
                                </table>
                            </td>
                        {.end}
                        </tr>
                    </table>
                    {.or}
                    {.end}
                </div>
            {.end}
        {.or}
            <p>No Seq2C events found in in {total_key_genes} {key_or_target} genes.</p>
        {.end}

        <a name="coverage"></a>
        {.section coverage}
        <h1>Coverage analysis for the {total_key_genes} {key_or_target} genes</h1>
        <div class="coverage_plot">
            <p style="margin-top: -5px;">
                Height indicates gene average coverage,
                <span style="color: green;">green</span> means that > 90% gene is covered at {depth_cutoff}x (half of global average coverage).
                More detailed plot for each gene is expanded by clicking on the corresponding point.
            </p>
            <div class="js_plot" id="gene_plot">
                <table>
                    <tr>
                        <td colspan="2">
                            <div class='plot_placeholder gene_plot_placeholder' id="gene_plot_placeholder">
                            </div>
                        </td>
                        <td>
                            <div class='legend_placeholder' id="gene_plot_legend_placeholder"></div>
                        </td>
                    </tr>
                    <tr>
                        <td colspan="2">
                               <span id="title_detailed_plot"><b></b></span>
                            <div class='plot_placeholder detailed_gene_plot_placeholder' id="detailed_gene_plot_placeholder">
                            </div>
                        </td>
                    </tr>
                </table>
                <div class="json_code">
                    <div id="gene_plot_data_json">
                        {plot_data}
                    </div>
                </div>
            </div>
        </div>

        <table width="100%">
            <tr style="vertical-align: top;">
            <!--<div class="multi-col-layout">-->
            {.repeated section columns}
                <td style="">
                    <table class="report_table static_report"
                           style="margin-right: 40px;">
                        <thead>
                            <tr class="top_row_tr">
                                {.repeated section metric_names}
                                    {@}
                                {.end}
                            </tr>
                        </thead>
                        <tbody>
                            {.repeated section rows}
                                <tr>
                                    {.repeated section records}
                                        {@}
                                    {.end}
                                </tr>
                            {.end}
                        </tbody>
                    </table>
                </td>
            {.end}
            <!--</div>-->
            </tr>
        </table>
        {.end}
    </div>
</div>
</body>

</html>