<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="html_aux/jquery.align-column.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>
    <script type="text/javascript" src="html_aux/dragtable.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for Tissue-Sensitive-2 (caller: vardict)</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"date":"22 August 2014, Friday, 08:41:49","common_records":[],"reports":[{"sample_reports":[{"sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"Tissue-Sensitive-2","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":null,"filtered_tsv_by_callername":{},"dirpath":null,"vcf_by_callername":{},"filtered_clean_vcf_by_callername":{},"var_dirpath":null},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":82,"dbsnp":82,"cosmic":82,"average":82.0},"all":{"1000genomes":398,"dbsnp":398,"cosmic":398,"average":398.0},"novel":{"1000genomes":316,"dbsnp":316,"cosmic":316,"average":316.0}},"value":398},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":65,"dbsnp":65,"cosmic":65,"average":65.0},"all":{"1000genomes":356,"dbsnp":356,"cosmic":356,"average":356.0},"novel":{"1000genomes":291,"dbsnp":291,"cosmic":291,"average":291.0}},"value":356},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":7,"dbsnp":7,"cosmic":7,"average":7.0},"all":{"1000genomes":11,"dbsnp":11,"cosmic":11,"average":11.0},"novel":{"1000genomes":4,"dbsnp":4,"cosmic":4,"average":4.0}},"value":11},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":10,"dbsnp":10,"cosmic":10,"average":10.0},"all":{"1000genomes":30,"dbsnp":30,"cosmic":30,"average":30.0},"novel":{"1000genomes":20,"dbsnp":20,"cosmic":20,"average":20.0}},"value":30},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":8,"dbsnp":82,"cosmic":9,"average":33.0},"all":{"1000genomes":8,"dbsnp":82,"cosmic":17,"average":35.666666666666664},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":8,"average":2.6666666666666665}},"value":17},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":9.76,"dbsnp":100.0,"cosmic":10.98,"average":40.24666666666667},"all":{"1000genomes":2.01,"dbsnp":20.6,"cosmic":4.27,"average":8.959999999999999},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":2.53,"average":0.8433333333333333}},"value":4.27},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":8,"dbsnp":61,"cosmic":4,"average":24.333333333333332},"all":{"1000genomes":8,"dbsnp":61,"cosmic":10,"average":26.333333333333332},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":6,"average":2.0}},"value":10},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":74.39,"cosmic":44.44,"average":72.94333333333333},"all":{"1000genomes":100.0,"dbsnp":74.39,"cosmic":58.82,"average":77.73666666666666},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":75.0,"average":25.0}},"value":58.82},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":3e-08,"dbsnp":3e-08,"cosmic":3e-08,"average":3e-08},"all":{"1000genomes":1.3e-07,"dbsnp":1.3e-07,"cosmic":1.3e-07,"average":1.3e-07},"novel":{"1000genomes":1e-07,"dbsnp":1e-07,"cosmic":1e-07,"average":1e-07}},"value":1.3e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":38258064.0,"dbsnp":38258064.0,"cosmic":38258064.0,"average":38258064.0},"all":{"1000genomes":7882314.0,"dbsnp":7882314.0,"cosmic":7882314.0,"average":7882314.0},"novel":{"1000genomes":9927725.0,"dbsnp":9927725.0,"cosmic":9927725.0,"average":9927725.0}},"value":7882314.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":7.73,"dbsnp":7.73,"cosmic":7.73,"average":7.73},"all":{"1000genomes":19.29,"dbsnp":19.29,"cosmic":19.29,"average":19.29},"novel":{"1000genomes":32.0,"dbsnp":32.0,"cosmic":32.0,"average":32.0}},"value":19.29},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":1.17,"dbsnp":1.17,"cosmic":1.17,"average":1.17},"all":{"1000genomes":0.41,"dbsnp":0.41,"cosmic":0.41,"average":0.41},"novel":{"1000genomes":0.31,"dbsnp":0.31,"cosmic":0.31,"average":0.31}},"value":0.41}],"name":"","fpath":"","plots":["/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-2/qc/varQC/Tissue-Sensitive-2.substitutions.png","/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-2/qc/varQC/Tissue-Sensitive-2.indels.png"]}],"name":""}]}
            </div>
        </div>
    </div>
</div>
</body>
</html>
