<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="html_aux/jquery.align-column.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>
    <script type="text/javascript" src="html_aux/dragtable.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for Tissue-Sensitive-1 (caller: mutect)</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"date":"22 August 2014, Friday, 08:46:54","common_records":[],"reports":[{"sample_reports":[{"sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"Tissue-Sensitive-1","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":null,"filtered_tsv_by_callername":{},"dirpath":null,"vcf_by_callername":{},"filtered_clean_vcf_by_callername":{},"var_dirpath":null},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":351,"dbsnp":351,"cosmic":351,"average":351.0},"all":{"1000genomes":1756,"dbsnp":1756,"cosmic":1756,"average":1756.0},"novel":{"1000genomes":1405,"dbsnp":1405,"cosmic":1405,"average":1405.0}},"value":1756},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":219,"dbsnp":219,"cosmic":219,"average":219.0},"all":{"1000genomes":1606,"dbsnp":1606,"cosmic":1606,"average":1606.0},"novel":{"1000genomes":1387,"dbsnp":1387,"cosmic":1387,"average":1387.0}},"value":1606},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":55,"dbsnp":55,"cosmic":55,"average":55.0},"all":{"1000genomes":65,"dbsnp":65,"cosmic":65,"average":65.0},"novel":{"1000genomes":10,"dbsnp":10,"cosmic":10,"average":10.0}},"value":65},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":67,"dbsnp":67,"cosmic":67,"average":67.0},"all":{"1000genomes":75,"dbsnp":75,"cosmic":75,"average":75.0},"novel":{"1000genomes":8,"dbsnp":8,"cosmic":8,"average":8.0}},"value":75},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":75,"dbsnp":351,"cosmic":19,"average":148.33333333333334},"all":{"1000genomes":75,"dbsnp":351,"cosmic":66,"average":164.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":47,"average":15.666666666666666}},"value":66},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":21.37,"dbsnp":100.0,"cosmic":5.41,"average":42.26},"all":{"1000genomes":4.27,"dbsnp":19.99,"cosmic":3.76,"average":9.339999999999998},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":3.35,"average":1.1166666666666667}},"value":3.76},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":71,"dbsnp":264,"cosmic":14,"average":116.33333333333333},"all":{"1000genomes":71,"dbsnp":264,"cosmic":29,"average":121.33333333333333},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":15,"average":5.0}},"value":29},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":94.67,"dbsnp":75.21,"cosmic":73.68,"average":81.18666666666667},"all":{"1000genomes":94.67,"dbsnp":75.21,"cosmic":43.94,"average":71.27333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":31.91,"average":10.636666666666667}},"value":43.94},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.1e-07,"dbsnp":1.1e-07,"cosmic":1.1e-07,"average":1.1e-07},"all":{"1000genomes":5.6e-07,"dbsnp":5.6e-07,"cosmic":5.6e-07,"average":5.6e-07},"novel":{"1000genomes":4.5e-07,"dbsnp":4.5e-07,"cosmic":4.5e-07,"average":4.5e-07}},"value":5.6e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":8937781.0,"dbsnp":8937781.0,"cosmic":8937781.0,"average":8937781.0},"all":{"1000genomes":1786538.0,"dbsnp":1786538.0,"cosmic":1786538.0,"average":1786538.0},"novel":{"1000genomes":2232854.0,"dbsnp":2232854.0,"cosmic":2232854.0,"average":2232854.0}},"value":1786538.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":5.88,"dbsnp":5.88,"cosmic":5.88,"average":5.88},"all":{"1000genomes":32.77,"dbsnp":32.77,"cosmic":32.77,"average":32.77},"novel":{"1000genomes":1404.0,"dbsnp":1404.0,"cosmic":1404.0,"average":1404.0}},"value":32.77},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":0.96,"dbsnp":0.96,"cosmic":0.96,"average":0.96},"all":{"1000genomes":0.13,"dbsnp":0.13,"cosmic":0.13,"average":0.13},"novel":{"1000genomes":0.06,"dbsnp":0.06,"cosmic":0.06,"average":0.06}},"value":0.13}],"name":"","fpath":"","plots":["/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-1/qc/varQC/Tissue-Sensitive-1.substitutions.png","/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-1/qc/varQC/Tissue-Sensitive-1.indels.png"]}],"name":""}]}
            </div>
        </div>
    </div>
</div>
</body>
</html>
