<!DOCTYPE html>
<html lang="en">
<head>
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="html_aux/report.css" />

    <script type="text/javascript">
        window.reporting = {};
    </script>

    <script type="text/javascript" src="html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>
    <script type="text/javascript" src="html_aux/jquery.align-column.js"></script>

    <script type="text/javascript" src="html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="html_aux/scripts/build_report.js"></script>
    <script type="text/javascript" src="html_aux/dragtable.js"></script>

    <script type="text/javascript">

        $(function() {
            reporting.buildReport();

            dragTable.init();

            $("[rel=tooltip]").tooltip({ animation: false, });
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'>Variant QC for Tissue-Sensitive-1 (caller: freebayes)</h1>
        <p id='report_date'></p>

        <!--<p id="data_set_p">-->
            <!--<span style="display: none;" id="reference_length">Genome:-->
                <!--<span class='val'></span><span class='rhs'>&nbsp;</span>bp</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_gc">,-->
                    <!--G+C content: <span class='val'></span><span class='rhs'>&nbsp;</span>%-->
                <!--</span>-->

                <!--<span style="display: none;" id="reference_genes"><br>-->
                    <!--<span class='val'></span> genes</span><span style='display: none;'>-->

                <!--</span><span style="display: none;" id="reference_operons">, <span class='val'></span> operons-->
            <!--</span>-->
        <!--</p>-->
    </div>

    <noscript>
        <div style="margin-left: 0px;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div id='report'>
        </div>

        <div id='plot'>
        </div>


        <!--<div class='plots'>-->
            <!--<table style='margin: -3px;'>-->
                <!--<tr id='all-switches'>-->
                    <!--<td>-->
                        <!--<div id='plots-switches'></div>-->
                    <!--</td>-->
                    <!--<td style='text-align: right; padding-right: 22px;'>-->
                        <!--<span id='scale-placeholder'></span>-->
                    <!--</td>-->
                <!--</tr>-->
                <!--<tr>-->
                    <!--<td colspan="2">-->
                        <!--<div id='plot-placeholder' style=''></div>-->
                    <!--</td>-->
                    <!--<td>-->
                        <!--<div id='legend-placeholder'></div>-->
                    <!--</td>-->
                <!--</tr>-->
            <!--</table>-->
        <!--</div>-->

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"date":"22 August 2014, Friday, 08:31:00","common_records":[],"reports":[{"sample_reports":[{"sample":{"bam":null,"filtered_vcf_by_callername":{},"name":"Tissue-Sensitive-1","filtered_maf_by_callername":{},"bed":null,"normal_match":null,"phenotype":null,"filtered_tsv_by_callername":{},"dirpath":null,"vcf_by_callername":{},"filtered_clean_vcf_by_callername":{},"var_dirpath":null},"records":[{"metric":{"name":"nEvalVariants","short_name":"Total","common":false,"quality":"More is better","unit":"","description":"Total variants evaluated"},"meta":{"known":{"1000genomes":391,"dbsnp":391,"cosmic":391,"average":391.0},"all":{"1000genomes":861,"dbsnp":861,"cosmic":861,"average":861.0},"novel":{"1000genomes":470,"dbsnp":470,"cosmic":470,"average":470.0}},"value":861},{"metric":{"name":"nSNPs","short_name":"SNP","common":false,"quality":"More is better","unit":"","description":"SNPs"},"meta":{"known":{"1000genomes":214,"dbsnp":214,"cosmic":214,"average":214.0},"all":{"1000genomes":440,"dbsnp":440,"cosmic":440,"average":440.0},"novel":{"1000genomes":226,"dbsnp":226,"cosmic":226,"average":226.0}},"value":440},{"metric":{"name":"nInsertions","short_name":"Ins","common":false,"quality":"More is better","unit":"","description":"Insertions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nDeletions","short_name":"Del","common":false,"quality":"More is better","unit":"","description":"Deletions"},"meta":{"known":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"all":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":0,"average":0.0}},"value":0},{"metric":{"name":"nVariantsAtComp","short_name":"At comp","common":false,"quality":"More is better","unit":"","description":"Number of eval sites at comp sites (that is, sharing the same locus as a variant in the comp track, regardless of whether the alternate allele is the same)"},"meta":{"known":{"1000genomes":13,"dbsnp":391,"cosmic":33,"average":145.66666666666666},"all":{"1000genomes":13,"dbsnp":391,"cosmic":41,"average":148.33333333333334},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":8,"average":2.6666666666666665}},"value":41},{"metric":{"name":"compRate","short_name":"Comp rate","common":false,"quality":"More is better","unit":"","description":"Percentage of eval sites at comp sites"},"meta":{"known":{"1000genomes":3.32,"dbsnp":100.0,"cosmic":8.44,"average":37.25333333333333},"all":{"1000genomes":1.51,"dbsnp":45.41,"cosmic":4.76,"average":17.226666666666663},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":1.7,"average":0.5666666666666667}},"value":4.76},{"metric":{"name":"nConcordant","short_name":"Concord","common":false,"quality":"More is better","unit":"","description":"Number of concordant sites (that is, for the sites that share the same locus as a variant in the comp track, those that have the same alternate allele)"},"meta":{"known":{"1000genomes":13,"dbsnp":211,"cosmic":26,"average":83.33333333333333},"all":{"1000genomes":13,"dbsnp":211,"cosmic":31,"average":85.0},"novel":{"1000genomes":0,"dbsnp":0,"cosmic":5,"average":1.6666666666666667}},"value":31},{"metric":{"name":"concordantRate","short_name":"Conc rate","common":false,"quality":"More is better","unit":"","description":"Concordance rate"},"meta":{"known":{"1000genomes":100.0,"dbsnp":53.96,"cosmic":78.79,"average":77.58333333333333},"all":{"1000genomes":100.0,"dbsnp":53.96,"cosmic":75.61,"average":76.52333333333333},"novel":{"1000genomes":0.0,"dbsnp":0.0,"cosmic":62.5,"average":20.833333333333332}},"value":75.61},{"metric":{"name":"variantRate","short_name":"Var/loci","common":false,"quality":"More is better","unit":"","description":"Variants per loci rate"},"meta":{"known":{"1000genomes":1.2e-07,"dbsnp":1.2e-07,"cosmic":1.2e-07,"average":1.2e-07},"all":{"1000genomes":2.7e-07,"dbsnp":2.7e-07,"cosmic":2.7e-07,"average":2.7e-07},"novel":{"1000genomes":1.5e-07,"dbsnp":1.5e-07,"cosmic":1.5e-07,"average":1.5e-07}},"value":2.7e-07},{"metric":{"name":"basesPerVariant","short_name":"Bp/var","common":false,"quality":"More is better","unit":"","description":"Bases per variant rate"},"meta":{"known":{"1000genomes":8023430.0,"dbsnp":8023430.0,"cosmic":8023430.0,"average":8023430.0},"all":{"1000genomes":3643625.0,"dbsnp":3643625.0,"cosmic":3643625.0,"average":3643625.0},"novel":{"1000genomes":6674811.0,"dbsnp":6674811.0,"cosmic":6674811.0,"average":6674811.0}},"value":3643625.0},{"metric":{"name":"hetHomRatio","short_name":"Het/hom","common":false,"quality":"More is better","unit":"","description":"Heterozygosity to homozygosity ratio"},"meta":{"known":{"1000genomes":7.76,"dbsnp":7.76,"cosmic":7.76,"average":7.760000000000001},"all":{"1000genomes":9.53,"dbsnp":9.53,"cosmic":9.53,"average":9.53},"novel":{"1000genomes":11.75,"dbsnp":11.75,"cosmic":11.75,"average":11.75}},"value":9.53},{"metric":{"name":"tiTvRatio","short_name":"Ti/Tv","common":false,"quality":"More is better","unit":"","description":"Transition to transversion ratio"},"meta":{"known":{"1000genomes":1.21,"dbsnp":1.21,"cosmic":1.21,"average":1.21},"all":{"1000genomes":0.74,"dbsnp":0.74,"cosmic":0.74,"average":0.7399999999999999},"novel":{"1000genomes":0.45,"dbsnp":0.45,"cosmic":0.45,"average":0.45}},"value":0.74}],"name":"","fpath":"","plots":["/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-1/qc/varQC/Tissue-Sensitive-1.substitutions.png","/gpfs/ngs/oncology/Analysis/translation/TS_0025_Platinum-Resistant-Ovarian/bcbio/NGS-99/final/Tissue-Sensitive-1/qc/varQC/Tissue-Sensitive-1.indels.png"]}],"name":""}]}
            </div>
        </div>
    </div>
</div>
</body>
</html>
